// @midnight-ntwrk/wallet-sdk-indexer-client/dist/effect/HttpQueryClient.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Effect, Layer } from 'effect';
import { print } from 'graphql';
import { createClient } from 'graphql-http';
import { QueryClient } from './QueryClient.js';
import { ClientError, ServerError, HttpURL, } from '@midnight-ntwrk/wallet-sdk-utilities/networking';
export const layer = (config) => Layer.effect(QueryClient, HttpURL.make(config.url).pipe(Effect.flatMap((url) => Effect.acquireRelease(Effect.sync(() => createClient({
    url: url.toString(),
    shouldRetry: (error, retries) => {
        const statusCode = error.response?.status ?? 500;
        return Promise.resolve(retries < 3 && statusCode >= 502 && statusCode <= 504);
    },
})), (client) => Effect.sync(() => client.dispose()))), Effect.map((client) => new HttpQueryClientImpl(client))));
class HttpQueryClientImpl {
    constructor(client) {
        this.client = client;
    }
    client;
    query(document, variables) {
        return Effect.async((resume) => {
            let result;
            const dispose = this.client.subscribe({ query: print(document), variables: variables }, {
                next: (data) => (result = data),
                error: (error) => {
                    const statusCode = error.response?.status ?? 500;
                    const message = error.response?.statusText ?? 'An unknown error occurred';
                    resume(Effect.fail(statusCode >= 400 && statusCode < 500
                        ? new ClientError({ message, cause: error })
                        : new ServerError({ message })));
                },
                complete: () => resume(result.errors
                    ? Effect.fail(new ClientError({ message: result.errors[0].message, cause: result.errors }))
                    : Effect.succeed(result.data)),
            });
            // Ensure we dispose of the query if the running Fiber is terminated.
            return Effect.sync(dispose);
        });
    }
}
