// @midnight-ntwrk/wallet-sdk-node-client/dist/effect/NodeClient.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Context, Effect, Option, Stream } from 'effect';
import * as SubmissionEvent from './SubmissionEvent.js';
import * as NodeClientError from './NodeClientError.js';
export class NodeClient extends Context.Tag('@midnight-ntwrk/wallet-node-client#NodeClient')() {
}
export const getGenesisTransactions = () => NodeClient.pipe(Effect.flatMap((client) => client.getGenesis()));
export const sendMidnightTransaction = (serializedTransaction) => NodeClient.pipe(Stream.fromEffect, Stream.flatMap((client) => client.sendMidnightTransaction(serializedTransaction)));
export function sendMidnightTransactionAndWait(serializedTransaction, waitFor) {
    return sendMidnightTransaction(serializedTransaction).pipe(Stream.find(SubmissionEvent.is(waitFor)), Stream.runHead, Effect.flatMap(Option.match({
        onNone: () => Effect.fail(new NodeClientError.TransactionProgressError({
            message: 'Transaction did not reach desired stage and no other error was reported',
            txData: serializedTransaction,
            desiredStage: waitFor,
        })),
        onSome: (event) => Effect.succeed(event),
    })));
}
