// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/storage/TransactionHistoryStorage.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { SafeBigInt } from '@midnight-ntwrk/wallet-sdk-utilities';
import { Schema } from 'effect';
const TransactionHashSchema = Schema.String;
const UtxoSchema = Schema.Struct({
    value: SafeBigInt.SafeBigInt,
    owner: Schema.String,
    tokenType: Schema.String,
    intentHash: Schema.String,
    outputIndex: Schema.Number,
});
export const TransactionHistoryEntrySchema = Schema.Struct({
    id: Schema.Number,
    hash: TransactionHashSchema,
    protocolVersion: Schema.Number,
    identifiers: Schema.Array(Schema.String),
    timestamp: Schema.Date,
    fees: Schema.NullOr(SafeBigInt.SafeBigInt),
    status: Schema.Literal('SUCCESS', 'FAILURE', 'PARTIAL_SUCCESS'),
    createdUtxos: Schema.Array(UtxoSchema),
    spentUtxos: Schema.Array(UtxoSchema),
});
