// @midnight-ntwrk/wallet-sdk-capabilities/dist/pendingTransactions/pendingTransactions.js
/*
 * This file is part of MIDNIGHT-WALLET-SDK.
 * Copyright (C) Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Array as Arr, Either, Order, pipe, Schema } from 'effect';
export const has = (transactions, transaction, txTrait) => {
    return transactions.all.some((item) => txTrait.areAllTxIdsIncluded(transaction, txTrait.ids(item.tx)));
};
export const all = (transactions) => {
    return transactions.all.map((item) => item.tx);
};
export const allFailed = (transactions) => {
    return transactions.all.filter((item) => 'result' in item && (item.result?.status === 'FAILURE' || item.result?.status === 'PARTIAL_SUCCESS'));
};
export const allPending = (state) => {
    return state.all.filter((item) => !('result' in item) || item.result === undefined);
};
export const empty = () => {
    return {
        all: [],
    };
};
export const addPendingTransaction = (state, tx, now, txTrait) => {
    const [rest, foundMatching] = pipe(state.all, Arr.partition((item) => txTrait.isOneIncludedInOther(tx, item.tx)));
    const allMatchingTransactions = Arr.append(foundMatching, { tx, creationTime: now });
    const theBiggestMatchingTx = Arr.max(allMatchingTransactions, pipe(Order.number, Order.mapInput((input) => txTrait.ids(input).length), Order.mapInput((input) => input.tx)));
    return {
        ...state,
        all: Arr.append(rest, theBiggestMatchingTx),
    };
};
export const clear = (state, tx, txTrait) => {
    return {
        ...state,
        all: Arr.filter(state.all, (item) => !txTrait.areAllTxIdsIncluded(item.tx, txTrait.ids(tx))),
    };
};
export const saveResult = (state, tx, result, txTrait) => {
    return {
        ...state,
        all: Arr.map(state.all, (item) => {
            return txTrait.areAllTxIdsIncluded(item.tx, txTrait.ids(tx)) ? { ...item, result } : item;
        }),
    };
};
export const SerializedSchema = (txTrait) => {
    const TxSchema = Schema.declare((tx) => txTrait.isTx(tx));
    const TxFromHex = Schema.transform(Schema.Uint8ArrayFromHex, TxSchema, {
        encode: (tx) => txTrait.serialize(tx),
        decode: (bytes) => txTrait.deserialize(bytes),
    });
    const TxItemSchema = Schema.Struct({
        tx: TxFromHex,
        creationTime: Schema.DateTimeUtc,
    });
    return Schema.Struct({
        version: Schema.Literal('v1'),
        transactions: Schema.Array(TxItemSchema),
    });
};
export const serialize = (state, txTrait) => pipe(state, toSerialized, Schema.encodeSync(SerializedSchema(txTrait)), JSON.stringify);
export const deserialize = (serialized, txTrait) => {
    return pipe(serialized, Schema.decodeUnknownEither(Schema.parseJson(SerializedSchema(txTrait))), Either.map((data) => fromSerialized(data)));
};
export const toSerialized = (pendingTransactions) => {
    return {
        version: 'v1',
        transactions: pendingTransactions.all,
    };
};
export const fromSerialized = (serialized) => {
    return {
        all: serialized.transactions,
    };
};
