// @midnight-ntwrk/wallet-sdk-capabilities/dist/pendingTransactions/pendingTransactionsService.js
/*
 * This file is part of MIDNIGHT-WALLET-SDK.
 * Copyright (C) Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DateTime, DefaultServices, Duration, Effect, Exit, Iterable, Option, pipe, Schedule, Scope, Stream, SubscriptionRef, } from 'effect';
import * as PendingTransactions from './pendingTransactions.js';
import { HttpQueryClient } from '@midnight-ntwrk/wallet-sdk-indexer-client/effect';
import { TransactionStatus } from '@midnight-ntwrk/wallet-sdk-indexer-client';
import { EitherOps, ObservableOps } from '@midnight-ntwrk/wallet-sdk-utilities';
export class PendingTransactionsServiceImpl {
    static init(initParams) {
        return PendingTransactionsServiceImpl.initEffect(initParams).pipe(Effect.runPromise);
    }
    static restore(data, txTrait, configuration) {
        return pipe(PendingTransactions.deserialize(data, txTrait), EitherOps.toEffect, Effect.andThen((state) => PendingTransactionsServiceImpl.initEffect({ txTrait, initialState: state, configuration })), Effect.runPromise);
    }
    static initEffect(initParams) {
        return Effect.gen(function* () {
            const service = new PendingTransactionsServiceEffectImpl(initParams.txTrait, initParams.initialState);
            const scope = yield* Scope.make();
            return new PendingTransactionsServiceImpl(service, scope, initParams.configuration);
        });
    }
    #effectService;
    #scope;
    #configuration;
    constructor(effectService, scope, configuration) {
        this.#effectService = effectService;
        this.#scope = scope;
        this.#configuration = configuration;
    }
    addPendingTransaction(tx) {
        return this.#effectService.addPendingTransaction(tx).pipe(Effect.runPromise);
    }
    clear(tx) {
        return this.#effectService.clear(tx).pipe(Effect.runPromise);
    }
    start() {
        return this.#effectService.startPolling(Stream.tick(Duration.seconds(1))).pipe(Effect.provide(HttpQueryClient.layer({
            url: this.#configuration.indexerClientConnection.indexerHttpUrl,
        })), Effect.provideService(Scope.Scope, this.#scope), Effect.provide(DefaultServices.liveServices), Effect.runFork, () => Promise.resolve());
    }
    state() {
        return this.#effectService.state().pipe(ObservableOps.fromStream);
    }
    stop() {
        return pipe(Scope.close(this.#scope, Exit.succeed(undefined)), Effect.runPromise);
    }
}
export class PendingTransactionsServiceEffectImpl {
    #state;
    #txTrait;
    static restore(data, txTrait) {
        return pipe(data, (data) => PendingTransactions.deserialize(data, txTrait), EitherOps.toEffect, Effect.map((state) => new PendingTransactionsServiceEffectImpl(txTrait, state)));
    }
    constructor(txTrait, initialState) {
        this.#txTrait = txTrait;
        this.#state = SubscriptionRef.make(initialState ?? PendingTransactions.empty()).pipe(Effect.runSync); // Should not be here, but otherwise initialization would be too involved
    }
    state() {
        return Stream.concat(Stream.fromEffect(SubscriptionRef.get(this.#state)), this.#state.changes);
    }
    startPolling(ticks) {
        return ticks.pipe(Stream.mapEffect(() => SubscriptionRef.get(this.#state)), Stream.mapConcat(PendingTransactions.allPending), Stream.mapConcatEffect((item) => {
            return Effect.gen(this, function* () {
                const now = yield* DateTime.now;
                const result = yield* this.queryForStatus(item.tx);
                return Option.match(result, {
                    onSome: (status) => [{ ...item, result: status }],
                    onNone: () => {
                        const failedResult = {
                            status: 'FAILURE',
                            segments: [],
                        };
                        return this.#txTrait.hasTTLExpired(item.tx, item.creationTime, now)
                            ? [{ ...item, result: failedResult }]
                            : [];
                    },
                });
            });
        }), Stream.retry(pipe(Schedule.exponential(Duration.millis(1)), Schedule.jitteredWith({ min: 0.1, max: 1.2 }), Schedule.resetAfter(Duration.minutes(5)))), Stream.catchAll((error) => {
            return Stream.execute(Effect.logWarning(error, 'Caught error in PendingTransactionsService'));
        }), Stream.runForEachScoped((item) => {
            return this.saveResult(item.tx, item.result);
        }));
    }
    addPendingTransaction(tx) {
        return SubscriptionRef.updateEffect(this.#state, (state) => {
            return DateTime.now.pipe(Effect.andThen((now) => PendingTransactions.addPendingTransaction(state, tx, now, this.#txTrait)));
        });
    }
    clear(tx) {
        return SubscriptionRef.update(this.#state, (state) => {
            return PendingTransactions.clear(state, tx, this.#txTrait);
        });
    }
    saveResult(tx, result) {
        switch (result.status) {
            case 'SUCCESS':
                return this.clear(tx);
            case 'FAILURE':
            case 'PARTIAL_SUCCESS':
                return SubscriptionRef.update(this.#state, (state) => {
                    return PendingTransactions.saveResult(state, tx, result, this.#txTrait);
                });
        }
    }
    queryForStatus(tx) {
        return Effect.gen(this, function* () {
            const statusQuery = yield* TransactionStatus;
            const result = yield* statusQuery({ transactionId: this.#txTrait.firstId(tx) }).pipe(Effect.catchAll((error) => {
                const fallback = { transactions: [] };
                return pipe(Effect.logWarning(error, 'Observed error in PendingTransactionsService, retrying'), Effect.as(fallback));
            }));
            return pipe(result.transactions, Iterable.filterMap((res) => {
                if (res.__typename == 'SystemTransaction') {
                    return Option.none();
                }
                if (res.transactionResult.status != 'SUCCESS' &&
                    res.transactionResult.status != 'FAILURE' &&
                    res.transactionResult.status != 'PARTIAL_SUCCESS') {
                    return Option.none();
                }
                if (this.#txTrait.areAllTxIdsIncluded(tx, res.identifiers)) {
                    return Option.some({
                        status: res.transactionResult.status,
                        segments: res.transactionResult.segments ?? [],
                    });
                }
                else {
                    return Option.none();
                }
            }), Iterable.head);
        });
    }
}
