// @midnight-ntwrk/midnight-js-level-private-state-provider/dist/index.mjs
import { MAX_EXPORT_SIGNING_KEYS, InvalidExportFormatError, ExportDecryptionError, ImportConflictError, SigningKeyExportError, MAX_EXPORT_STATES, PrivateStateExportError } from '@midnight-ntwrk/midnight-js-types';
import { Buffer } from 'buffer';
import * as crypto from 'crypto';
import { randomBytes, pbkdf2Sync, createCipheriv, createDecipheriv, createHash } from 'crypto';
import { Level } from 'level';
import * as superjson from 'superjson';

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const ALGORITHM = 'aes-256-gcm';
const KEY_LENGTH = 32;
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;
const SALT_LENGTH = 32;
const PBKDF2_ITERATIONS_V1 = 100000;
const PBKDF2_ITERATIONS_V2 = 600000;
const ENCRYPTION_VERSION_V1 = 1;
const ENCRYPTION_VERSION_V2 = 2;
const CURRENT_ENCRYPTION_VERSION = ENCRYPTION_VERSION_V2;
const VERSION_PREFIX_LENGTH = 1;
const HEADER_LENGTH = VERSION_PREFIX_LENGTH + SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH;
const HAS_NATIVE_TIMINGSAFEEQUAL = 'timingSafeEqual' in crypto && typeof crypto.timingSafeEqual === 'function';
const extractEncryptedComponents = (data) => {
    if (data.length < HEADER_LENGTH) {
        throw new Error('Invalid encrypted data: too short');
    }
    const version = data[0];
    if (version !== ENCRYPTION_VERSION_V1 && version !== ENCRYPTION_VERSION_V2) {
        throw new Error(`Unsupported encryption version: ${version}`);
    }
    return {
        version,
        salt: data.subarray(VERSION_PREFIX_LENGTH, VERSION_PREFIX_LENGTH + SALT_LENGTH),
        iv: data.subarray(VERSION_PREFIX_LENGTH + SALT_LENGTH, VERSION_PREFIX_LENGTH + SALT_LENGTH + IV_LENGTH),
        authTag: data.subarray(VERSION_PREFIX_LENGTH + SALT_LENGTH + IV_LENGTH, VERSION_PREFIX_LENGTH + SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH),
        encrypted: data.subarray(HEADER_LENGTH)
    };
};
const getIterationsForVersion = (version) => {
    switch (version) {
        case ENCRYPTION_VERSION_V1:
            return PBKDF2_ITERATIONS_V1;
        case ENCRYPTION_VERSION_V2:
            return PBKDF2_ITERATIONS_V2;
        default:
            throw new Error(`Unsupported encryption version: ${version}`);
    }
};
const hashPassword = (password) => {
    return createHash('sha256').update(password).digest('hex');
};
const constantTimeBufferEqual = (aBuf, bBuf) => {
    if (aBuf.length !== bBuf.length) {
        throw new RangeError('Input buffers must have the same byte length');
    }
    let result = 0;
    for (let i = 0; i < aBuf.length; i++) {
        result |= aBuf[i] ^ bBuf[i];
    }
    return result === 0;
};
/**
 * Compares two Buffers or Uint8Arrays in constant time.
 *
 * @param a - First buffer to compare.
 * @param b - Second buffer to compare.
 * @returns `true` if the buffers are equal, `false` otherwise.
 *
 * @remarks
 * If the inputs differ in length, an error is thrown (not constant-time for length mismatch).
 * This matches the Node.js native timingSafeEqual behavior (which throws on length mismatch).
 *
 * For fixed-length buffers (e.g., hashes), this is safe. For variable-length buffers, callers should be
 * aware of potential timing leakage.
 */
const timingSafeEqual = (a, b) => {
    const aBuf = Buffer.isBuffer(a) ? a : Buffer.from(a);
    const bBuf = Buffer.isBuffer(b) ? b : Buffer.from(b);
    if (HAS_NATIVE_TIMINGSAFEEQUAL) {
        // Use the native `timingSafeEqual` function and let any errors propagate.
        return crypto.timingSafeEqual(aBuf, bBuf);
    }
    return constantTimeBufferEqual(aBuf, bBuf);
};
class StorageEncryption {
    encryptionKey;
    salt;
    passwordHash;
    constructor(password, existingSalt) {
        this.salt = existingSalt ?? randomBytes(SALT_LENGTH);
        this.encryptionKey = this.deriveKey(password, this.salt, PBKDF2_ITERATIONS_V2);
        this.passwordHash = hashPassword(password);
    }
    deriveKey(password, salt, iterations) {
        return pbkdf2Sync(password, salt, iterations, KEY_LENGTH, 'sha256');
    }
    verifyPassword(password) {
        const inputHash = Buffer.from(hashPassword(password), 'hex');
        const storedHash = Buffer.from(this.passwordHash, 'hex');
        return timingSafeEqual(inputHash, storedHash);
    }
    encrypt(data) {
        const plaintext = Buffer.from(data, 'utf-8');
        const iv = randomBytes(IV_LENGTH);
        const cipher = createCipheriv(ALGORITHM, this.encryptionKey, iv);
        const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()]);
        const authTag = cipher.getAuthTag();
        const version = Buffer.from([CURRENT_ENCRYPTION_VERSION]);
        const result = Buffer.concat([version, this.salt, iv, authTag, encrypted]);
        return result.toString('base64');
    }
    decrypt(encryptedData) {
        const data = Buffer.from(encryptedData, 'base64');
        const { version, salt, iv, authTag, encrypted } = extractEncryptedComponents(data);
        if (version === ENCRYPTION_VERSION_V1) {
            throw new Error('V1 encrypted data requires password for decryption. Use decryptWithPassword() instead.');
        }
        if (!this.salt.equals(salt)) {
            throw new Error('Salt mismatch: data was encrypted with a different password');
        }
        const decipher = createDecipheriv(ALGORITHM, this.encryptionKey, iv, { authTagLength: AUTH_TAG_LENGTH });
        decipher.setAuthTag(authTag);
        const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);
        return decrypted.toString('utf-8');
    }
    decryptWithPassword(encryptedData, password) {
        const data = Buffer.from(encryptedData, 'base64');
        const { version, salt, iv, authTag, encrypted } = extractEncryptedComponents(data);
        if (!this.salt.equals(salt)) {
            throw new Error('Salt mismatch: data was encrypted with a different password');
        }
        const iterations = getIterationsForVersion(version);
        const decryptionKey = version === CURRENT_ENCRYPTION_VERSION
            ? this.encryptionKey
            : this.deriveKey(password, salt, iterations);
        const decipher = createDecipheriv(ALGORITHM, decryptionKey, iv, { authTagLength: AUTH_TAG_LENGTH });
        decipher.setAuthTag(authTag);
        const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);
        return decrypted.toString('utf-8');
    }
    static isEncrypted(data) {
        try {
            const buffer = Buffer.from(data, 'base64');
            const version = buffer[0];
            return buffer.length >= HEADER_LENGTH &&
                (version === ENCRYPTION_VERSION_V1 || version === ENCRYPTION_VERSION_V2);
        }
        catch {
            return false;
        }
    }
    static getVersion(encryptedData) {
        const buffer = Buffer.from(encryptedData, 'base64');
        if (buffer.length < 1) {
            throw new Error('Invalid encrypted data: too short');
        }
        return buffer[0];
    }
    getSalt() {
        return this.salt;
    }
}
const MIN_PASSWORD_LENGTH$1 = 16;
const MIN_CHARACTER_CLASSES = 3;
const MAX_CONSECUTIVE_REPEATED = 3;
const MIN_SEQUENTIAL_LENGTH = 4;
const countCharacterClasses = (password) => {
    let count = 0;
    if (/[a-z]/.test(password))
        count++;
    if (/[A-Z]/.test(password))
        count++;
    if (/[0-9]/.test(password))
        count++;
    if (/[^a-zA-Z0-9]/.test(password))
        count++;
    return count;
};
const hasRepeatedCharacters = (password) => {
    let consecutiveCount = 1;
    for (let i = 1; i < password.length; i++) {
        if (password[i] === password[i - 1]) {
            consecutiveCount++;
            if (consecutiveCount > MAX_CONSECUTIVE_REPEATED) {
                return true;
            }
        }
        else {
            consecutiveCount = 1;
        }
    }
    return false;
};
const hasSequentialPattern = (password) => {
    const lowerPassword = password.toLowerCase();
    for (let i = 0; i <= lowerPassword.length - MIN_SEQUENTIAL_LENGTH; i++) {
        let ascendingCount = 1;
        let descendingCount = 1;
        for (let j = 1; j < MIN_SEQUENTIAL_LENGTH; j++) {
            const currentCode = lowerPassword.charCodeAt(i + j);
            const prevCode = lowerPassword.charCodeAt(i + j - 1);
            if (currentCode === prevCode + 1) {
                ascendingCount++;
            }
            else {
                ascendingCount = 1;
            }
            if (currentCode === prevCode - 1) {
                descendingCount++;
            }
            else {
                descendingCount = 1;
            }
            if (ascendingCount >= MIN_SEQUENTIAL_LENGTH || descendingCount >= MIN_SEQUENTIAL_LENGTH) {
                return true;
            }
        }
    }
    return false;
};
const validatePassword = (password) => {
    if (!password) {
        throw new Error('Password is required for private state encryption.\n' +
            'Please provide a password via privateStoragePasswordProvider in the configuration.');
    }
    if (password.length < MIN_PASSWORD_LENGTH$1) {
        throw new Error(`Password must be at least ${MIN_PASSWORD_LENGTH$1} characters long. Current length: ${password.length}`);
    }
    if (hasRepeatedCharacters(password)) {
        throw new Error(`Password contains too many repeated characters (more than ${MAX_CONSECUTIVE_REPEATED} identical in a row)`);
    }
    const characterClasses = countCharacterClasses(password);
    if (characterClasses < MIN_CHARACTER_CLASSES) {
        throw new Error(`Password must contain at least ${MIN_CHARACTER_CLASSES} of: uppercase letters, lowercase letters, digits, special characters. Found: ${characterClasses}`);
    }
    if (hasSequentialPattern(password)) {
        throw new Error("Password contains sequential patterns (e.g., '1234', 'abcd'). Use a more random password");
    }
};
const getPasswordFromProvider = async (provider) => {
    const password = await provider();
    validatePassword(password);
    return password;
};
const decryptValue = (encryptedValue, encryption, password) => {
    if (!StorageEncryption.isEncrypted(encryptedValue)) {
        console.debug('MIDNIGHT: Encountered unencrypted data during decryption - passing through as-is');
        return encryptedValue;
    }
    const version = StorageEncryption.getVersion(encryptedValue);
    if (version === 1) {
        return encryption.decryptWithPassword(encryptedValue, password);
    }
    return encryption.decrypt(encryptedValue);
};

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * The default name of the indexedDB database for Midnight.
 */
const MN_LDB_DEFAULT_DB_NAME = 'midnight-level-db';
/**
 * The default name of the private state store.
 */
const MN_LDB_DEFAULT_PRIS_STORE_NAME = 'private-states';
/**
 * The default name of the signing key store.
 */
const MN_LDB_DEFAULT_KEY_STORE_NAME = 'signing-keys';
/**
 * The default configuration for the level database.
 */
const DEFAULT_CONFIG = {
    /**
     * The name of the database.
     */
    midnightDbName: MN_LDB_DEFAULT_DB_NAME,
    /**
     * The name of the "level" on which to store private state.
     */
    privateStateStoreName: MN_LDB_DEFAULT_PRIS_STORE_NAME,
    /**
     * The name of the "level" on which to store signing keys.
     */
    signingKeyStoreName: MN_LDB_DEFAULT_KEY_STORE_NAME
};
superjson.registerCustom({
    isApplicable: (v) => v instanceof Buffer,
    serialize: (v) => v.toString('hex'),
    deserialize: (v) => Buffer.from(v, 'hex')
}, 'buffer');
const ACCOUNT_ID_HASH_LENGTH = 32;
const hashAccountId = (accountId) => {
    return createHash('sha256').update(accountId).digest('hex').substring(0, ACCOUNT_ID_HASH_LENGTH);
};
const getScopedLevelName = (baseLevelName, accountId) => {
    const hashedAccountId = hashAccountId(accountId);
    return `${baseLevelName}:${hashedAccountId}`;
};
const withSubLevel = async (dbName, levelName, thunk) => {
    const level = new Level(dbName, {
        createIfMissing: true
    });
    const subLevel = level.sublevel(levelName, {
        valueEncoding: 'utf-8'
    });
    try {
        await level.open();
        await subLevel.open();
        return await thunk(subLevel);
    }
    finally {
        await subLevel.close();
        await level.close();
    }
};
const METADATA_KEY = '__midnight_encryption_metadata__';
const DEFAULT_MAX_ROTATION_ENTRIES = 10000;
const encryptionInitPromises = new Map();
const passwordRotationLocks = new Map();
/**
 * Module-level cache for StorageEncryption instances, keyed by `${dbName}:${levelName}`.
 * This cache avoids repeated PBKDF2 key derivation (600,000 iterations) on each operation.
 *
 * Note: This cache has no size limit. For typical usage with a small number of
 * database/level combinations, this is acceptable. If using dynamic db/level names,
 * call `invalidateEncryptionCache()` to prevent unbounded growth.
 */
const encryptionCache = new Map();
const getOrCreateSalt = async (dbName, levelName) => {
    const lockKey = `${dbName}:${levelName}`;
    const existingPromise = encryptionInitPromises.get(lockKey);
    if (existingPromise) {
        return existingPromise;
    }
    const initPromise = withSubLevel(dbName, levelName, async (subLevel) => {
        try {
            const metadataJson = await subLevel.get(METADATA_KEY);
            if (metadataJson) {
                const metadata = JSON.parse(metadataJson);
                return Buffer.from(metadata.salt, 'hex');
            }
        }
        catch (error) {
            if (!(error && typeof error === 'object' && 'code' in error && error.code === 'LEVEL_NOT_FOUND')) {
                throw error;
            }
        }
        const salt = randomBytes(32);
        const metadata = {
            salt: salt.toString('hex'),
            version: 1
        };
        await subLevel.put(METADATA_KEY, JSON.stringify(metadata));
        return salt;
    });
    encryptionInitPromises.set(lockKey, initPromise);
    try {
        return await initPromise;
    }
    finally {
        encryptionInitPromises.delete(lockKey);
    }
};
const getOrCreateEncryption = async (dbName, levelName, passwordProvider) => {
    const cacheKey = `${dbName}:${levelName}`;
    const salt = await getOrCreateSalt(dbName, levelName);
    const saltHex = salt.toString('hex');
    const cached = encryptionCache.get(cacheKey);
    if (cached && cached.saltHex === saltHex) {
        const password = await getPasswordFromProvider(passwordProvider);
        if (cached.encryption.verifyPassword(password)) {
            return cached.encryption;
        }
        const encryption = new StorageEncryption(password, salt);
        encryptionCache.set(cacheKey, { encryption, saltHex });
        return encryption;
    }
    const password = await getPasswordFromProvider(passwordProvider);
    const encryption = new StorageEncryption(password, salt);
    encryptionCache.set(cacheKey, { encryption, saltHex });
    return encryption;
};
const invalidateEncryptionCacheForDb = (dbName, privateStateStoreName, signingKeyStoreName) => {
    const privateStateKey = `${dbName}:${privateStateStoreName}`;
    const signingKeyKey = `${dbName}:${signingKeyStoreName}`;
    encryptionCache.delete(privateStateKey);
    encryptionCache.delete(signingKeyKey);
};
const DEFAULT_LOCK_TIMEOUT_MS = 300000; // 5 minutes
const withPasswordRotationLock = async (lockKey, operation, timeoutMs = DEFAULT_LOCK_TIMEOUT_MS) => {
    const startWait = Date.now();
    while (passwordRotationLocks.has(lockKey)) {
        if (Date.now() - startWait > timeoutMs) {
            throw new Error(`Timed out waiting for password rotation lock on "${lockKey}". ` +
                `Another rotation may be stuck or taking longer than ${timeoutMs}ms.`);
        }
        await passwordRotationLocks.get(lockKey);
    }
    let resolve;
    const lockPromise = new Promise((r) => {
        resolve = r;
    });
    passwordRotationLocks.set(lockKey, lockPromise);
    try {
        return await operation();
    }
    finally {
        passwordRotationLocks.delete(lockKey);
        resolve();
    }
};
const waitForRotationLock = async (dbName, levelName, timeoutMs = DEFAULT_LOCK_TIMEOUT_MS) => {
    const lockKey = `${dbName}:${levelName}`;
    const startWait = Date.now();
    while (passwordRotationLocks.has(lockKey)) {
        if (Date.now() - startWait > timeoutMs) {
            throw new Error(`Timed out waiting for password rotation to complete on "${lockKey}". ` +
                `The rotation may be stuck or taking longer than ${timeoutMs}ms.`);
        }
        await passwordRotationLocks.get(lockKey);
    }
};
const isDecryptionError = (error) => {
    if (!(error instanceof Error))
        return false;
    const message = error.message.toLowerCase();
    return (message.includes('unsupported state') ||
        message.includes('salt mismatch') ||
        message.includes('invalid encrypted data') ||
        message.includes('bad decrypt') ||
        message.includes('unable to authenticate'));
};
const rotateStorePassword = async (params) => {
    const { dbName, storeName, oldPasswordProvider, newPasswordProvider, maxEntries, shouldProceed } = params;
    const oldPassword = await getPasswordFromProvider(oldPasswordProvider);
    const newPassword = await getPasswordFromProvider(newPasswordProvider);
    const salt = await getOrCreateSalt(dbName, storeName);
    const oldEncryption = new StorageEncryption(oldPassword, salt);
    const newEncryption = new StorageEncryption(newPassword);
    const newSalt = newEncryption.getSalt();
    return withSubLevel(dbName, storeName, async (subLevel) => {
        const entriesToMigrate = [];
        let hasMatchingData = false;
        let firstEntryValidated = false;
        for await (const [key, encryptedValue] of subLevel.iterator()) {
            if (key === METADATA_KEY)
                continue;
            if (entriesToMigrate.length >= maxEntries) {
                throw new Error(`Entry count exceeds maximum allowed (${maxEntries}). ` +
                    `Use the maxEntries option to increase the limit if needed.`);
            }
            if (shouldProceed && shouldProceed(key)) {
                hasMatchingData = true;
            }
            if (!firstEntryValidated) {
                try {
                    decryptValue(encryptedValue, oldEncryption, oldPassword);
                }
                catch (error) {
                    if (isDecryptionError(error)) {
                        throw new Error('Old password is incorrect: failed to decrypt existing data', { cause: error });
                    }
                    throw error;
                }
                firstEntryValidated = true;
            }
            try {
                const decryptedValue = decryptValue(encryptedValue, oldEncryption, oldPassword);
                entriesToMigrate.push({ key, decryptedValue });
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : 'Unknown error';
                throw new Error(`Failed to decrypt entry "${key}": ${errorMessage}. ` +
                    `Successfully processed ${entriesToMigrate.length} entries before failure.`, { cause: error });
            }
        }
        if (entriesToMigrate.length === 0) {
            return { entriesMigrated: 0 };
        }
        if (shouldProceed && !hasMatchingData) {
            return { entriesMigrated: 0 };
        }
        const operations = [];
        for (const { key, decryptedValue } of entriesToMigrate) {
            try {
                const encryptedValue = newEncryption.encrypt(decryptedValue);
                operations.push({ type: 'put', key, value: encryptedValue });
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : 'Unknown error';
                throw new Error(`Failed to re-encrypt entry "${key}": ${errorMessage}. ` +
                    `Original data is still encrypted with old password.`, { cause: error });
            }
        }
        const metadata = {
            salt: newSalt.toString('hex'),
            version: 1
        };
        operations.push({ type: 'put', key: METADATA_KEY, value: JSON.stringify(metadata) });
        try {
            await subLevel.batch(operations);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Failed to write re-encrypted data: ${errorMessage}. ` +
                `Your data may be in an inconsistent state. ` +
                `Keep both old and new passwords until you can verify data integrity.`, { cause: error });
        }
        return { entriesMigrated: entriesToMigrate.length };
    });
};
const subLevelMaybeGet = async (dbName, levelName, key, passwordProvider) => {
    await waitForRotationLock(dbName, levelName);
    const encryption = await getOrCreateEncryption(dbName, levelName, passwordProvider);
    return withSubLevel(dbName, levelName, async (subLevel) => {
        try {
            const encryptedValue = await subLevel.get(key);
            if (encryptedValue === undefined) {
                return null;
            }
            let decryptedValue;
            if (StorageEncryption.isEncrypted(encryptedValue)) {
                const version = StorageEncryption.getVersion(encryptedValue);
                if (version === 1) {
                    const password = await getPasswordFromProvider(passwordProvider);
                    decryptedValue = encryption.decryptWithPassword(encryptedValue, password);
                    const reEncrypted = encryption.encrypt(decryptedValue);
                    await subLevel.put(key, reEncrypted);
                }
                else {
                    decryptedValue = encryption.decrypt(encryptedValue);
                }
            }
            else {
                decryptedValue = encryptedValue;
                const reEncrypted = encryption.encrypt(encryptedValue);
                await subLevel.put(key, reEncrypted);
            }
            const value = superjson.parse(decryptedValue);
            if (value === undefined) {
                return null;
            }
            return value;
        }
        catch (error) {
            if (error && typeof error === 'object' && 'code' in error && error.code === 'LEVEL_NOT_FOUND') {
                return null;
            }
            throw error;
        }
    });
};
/**
 * Iterate all key-value pairs in a sublevel, excluding metadata keys.
 */
const getAllEntries = async (dbName, levelName, passwordProvider) => {
    await waitForRotationLock(dbName, levelName);
    const encryption = await getOrCreateEncryption(dbName, levelName, passwordProvider);
    return withSubLevel(dbName, levelName, async (subLevel) => {
        const entries = new Map();
        let password = null;
        for await (const [key, encryptedValue] of subLevel.iterator()) {
            if (key === METADATA_KEY) {
                continue;
            }
            let decryptedValue;
            let needsReEncryption = false;
            if (StorageEncryption.isEncrypted(encryptedValue)) {
                const version = StorageEncryption.getVersion(encryptedValue);
                if (version === 1) {
                    if (password === null) {
                        password = await getPasswordFromProvider(passwordProvider);
                    }
                    decryptedValue = encryption.decryptWithPassword(encryptedValue, password);
                    needsReEncryption = true;
                }
                else {
                    decryptedValue = encryption.decrypt(encryptedValue);
                }
            }
            else {
                decryptedValue = encryptedValue;
                needsReEncryption = true;
            }
            if (needsReEncryption) {
                const reEncrypted = encryption.encrypt(decryptedValue);
                await subLevel.put(key, reEncrypted);
            }
            const value = superjson.parse(decryptedValue);
            entries.set(key, value);
        }
        return entries;
    });
};
const CURRENT_EXPORT_VERSION = 1;
const SUPPORTED_EXPORT_VERSIONS = [1];
const EXPECTED_SALT_LENGTH = 64; // 32 bytes as hex
const MIN_PASSWORD_LENGTH = 16;
/**
 * Validates a custom password meets minimum requirements.
 */
const validateExportPassword = (password) => {
    if (password.length < MIN_PASSWORD_LENGTH) {
        throw new PrivateStateExportError(`Password must be at least ${MIN_PASSWORD_LENGTH} characters`);
    }
};
/**
 * Validates the salt format and length.
 */
const validateSalt = (salt) => {
    if (salt.length !== EXPECTED_SALT_LENGTH) {
        throw new InvalidExportFormatError('Invalid salt length');
    }
    if (!/^[0-9a-fA-F]+$/.test(salt)) {
        throw new InvalidExportFormatError('Invalid salt format');
    }
};
/**
 * Validates a custom signing key export password meets minimum requirements.
 */
const validateSigningKeyExportPassword = (password) => {
    if (password.length < MIN_PASSWORD_LENGTH) {
        throw new SigningKeyExportError(`Password must be at least ${MIN_PASSWORD_LENGTH} characters`);
    }
};
const BROWSER_WARNING_KEY = '__midnight_browser_warning_shown__';
const isBrowserEnvironment = () => {
    const global = globalThis;
    return typeof globalThis !== 'undefined' &&
        'window' in globalThis &&
        global.window !== undefined &&
        'document' in globalThis &&
        global.document !== undefined;
};
const getSessionStorage = () => {
    if (typeof globalThis !== 'undefined' && 'sessionStorage' in globalThis) {
        return globalThis.sessionStorage;
    }
    return undefined;
};
/**
 * Shows a warning about browser storage risks.
 * Only shows once per session using sessionStorage.
 */
const showBrowserWarning = () => {
    if (!isBrowserEnvironment()) {
        return;
    }
    try {
        const storage = getSessionStorage();
        if (storage) {
            if (storage.getItem(BROWSER_WARNING_KEY)) {
                return;
            }
            storage.setItem(BROWSER_WARNING_KEY, 'true');
        }
    }
    catch (error) {
        console.debug('MIDNIGHT: Could not access sessionStorage for warning deduplication:', error instanceof Error ? error.message : 'Unknown error');
    }
    console.warn(`⚠️ MIDNIGHT: Private state and signing keys are stored in browser storage.\n` +
        `Clearing browser cache or storage will permanently destroy this data.\n` +
        `For assets with real-world value, this may result in irreversible financial loss.\n` +
        `Use exportPrivateStates() and exportSigningKeys() to create backups.`);
};
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * Constructs an instance of {@link PrivateStateProvider} based on {@link Level} database.
 *
 * ⚠️ WARNING
 *
 * RISK: This provider lacks a recovery mechanism.
 * Clearing browser cache or deleting local files permanently destroys the private state (contract state/keys).
 * For assets with real-world value, this may result in irreversible financial loss.
 * DO NOT use for production applications requiring data persistence.
 *
 * @param config Database configuration options.
 */
const levelPrivateStateProvider = (config) => {
    const fullConfig = { ...DEFAULT_CONFIG, ...config };
    if (!config.privateStoragePasswordProvider) {
        throw new Error('privateStoragePasswordProvider is required.\n' +
            'Provide a function that returns a strong, secret password (minimum 16 characters).');
    }
    if (!config.accountId || config.accountId.trim().length === 0) {
        throw new Error('accountId is required.\n' +
            'Provide an account identifier (e.g., wallet address) to scope storage and prevent cross-account data access.');
    }
    const passwordProvider = config.privateStoragePasswordProvider;
    const scopedPrivateStateLevelName = getScopedLevelName(fullConfig.privateStateStoreName, config.accountId);
    const scopedSigningKeyLevelName = getScopedLevelName(fullConfig.signingKeyStoreName, config.accountId);
    showBrowserWarning();
    let contractAddress = null;
    const getScopedKey = (privateStateId) => {
        if (contractAddress === null) {
            throw new Error('Contract address not set. Call setContractAddress() before accessing private state.');
        }
        return `${contractAddress}:${privateStateId}`;
    };
    return {
        setContractAddress(address) {
            contractAddress = address;
        },
        async get(privateStateId) {
            const scopedKey = getScopedKey(privateStateId);
            return subLevelMaybeGet(fullConfig.midnightDbName, scopedPrivateStateLevelName, scopedKey, passwordProvider);
        },
        async remove(privateStateId) {
            await waitForRotationLock(fullConfig.midnightDbName, scopedPrivateStateLevelName);
            const scopedKey = getScopedKey(privateStateId);
            return withSubLevel(fullConfig.midnightDbName, scopedPrivateStateLevelName, (subLevel) => subLevel.del(scopedKey));
        },
        async set(privateStateId, state) {
            await waitForRotationLock(fullConfig.midnightDbName, scopedPrivateStateLevelName);
            const scopedKey = getScopedKey(privateStateId);
            const encryption = await getOrCreateEncryption(fullConfig.midnightDbName, scopedPrivateStateLevelName, passwordProvider);
            const serialized = superjson.stringify(state);
            const encrypted = encryption.encrypt(serialized);
            return withSubLevel(fullConfig.midnightDbName, scopedPrivateStateLevelName, (subLevel) => subLevel.put(scopedKey, encrypted));
        },
        async clear() {
            if (contractAddress === null) {
                throw new Error('Contract address not set. Call setContractAddress() before accessing private state.');
            }
            return withSubLevel(fullConfig.midnightDbName, scopedPrivateStateLevelName, (subLevel) => subLevel.clear());
        },
        getSigningKey(address) {
            return subLevelMaybeGet(fullConfig.midnightDbName, scopedSigningKeyLevelName, address, passwordProvider);
        },
        async removeSigningKey(address) {
            await waitForRotationLock(fullConfig.midnightDbName, scopedSigningKeyLevelName);
            return withSubLevel(fullConfig.midnightDbName, scopedSigningKeyLevelName, (subLevel) => subLevel.del(address));
        },
        async setSigningKey(address, signingKey) {
            await waitForRotationLock(fullConfig.midnightDbName, scopedSigningKeyLevelName);
            const encryption = await getOrCreateEncryption(fullConfig.midnightDbName, scopedSigningKeyLevelName, passwordProvider);
            const serialized = superjson.stringify(signingKey);
            const encrypted = encryption.encrypt(serialized);
            return withSubLevel(fullConfig.midnightDbName, scopedSigningKeyLevelName, (subLevel) => subLevel.put(address, encrypted));
        },
        clearSigningKeys() {
            return withSubLevel(fullConfig.midnightDbName, scopedSigningKeyLevelName, (subLevel) => subLevel.clear());
        },
        async exportPrivateStates(options) {
            if (contractAddress === null) {
                throw new Error('Contract address not set. Call setContractAddress() before exporting private states.');
            }
            const maxStates = options?.maxStates ?? MAX_EXPORT_STATES;
            // Validate custom password if provided
            if (options?.password !== undefined) {
                validateExportPassword(options.password);
            }
            // Determine export password - use provided password or storage password
            const exportPassword = options?.password ?? await getPasswordFromProvider(passwordProvider);
            // Get all private states (not signing keys)
            const allStates = await getAllEntries(fullConfig.midnightDbName, scopedPrivateStateLevelName, passwordProvider);
            // Filter and extract only states for the current contract address
            const prefix = `${contractAddress}:`;
            const states = new Map();
            for (const [scopedKey, value] of allStates.entries()) {
                if (scopedKey.startsWith(prefix)) {
                    const rawStateId = scopedKey.slice(prefix.length);
                    states.set(rawStateId, value);
                }
            }
            if (states.size === 0) {
                throw new PrivateStateExportError('No private states to export');
            }
            if (states.size > maxStates) {
                throw new PrivateStateExportError(`Too many states to export (${states.size}). Maximum allowed: ${maxStates}`);
            }
            // Serialize states using superjson (to preserve types like BigInt, Buffer, etc.)
            // Include metadata in the encrypted payload to ensure it's authenticated
            const payload = {
                version: CURRENT_EXPORT_VERSION,
                exportedAt: new Date().toISOString(),
                stateCount: states.size,
                states: Object.fromEntries(Array.from(states.entries()).map(([key, value]) => [key, superjson.stringify(value)]))
            };
            // Create new encryption instance for export (different salt from storage)
            const exportEncryption = new StorageEncryption(exportPassword);
            const encryptedPayload = exportEncryption.encrypt(JSON.stringify(payload));
            return {
                format: 'midnight-private-state-export',
                encryptedPayload,
                salt: exportEncryption.getSalt().toString('hex')
            };
        },
        async importPrivateStates(exportData, options) {
            if (contractAddress === null) {
                throw new Error('Contract address not set. Call setContractAddress() before importing private states.');
            }
            const conflictStrategy = options?.conflictStrategy ?? 'error';
            const maxStates = options?.maxStates ?? MAX_EXPORT_STATES;
            // Validate format identifier
            if (exportData.format !== 'midnight-private-state-export') {
                throw new InvalidExportFormatError('Unrecognized export format');
            }
            // Validate structure
            if (!exportData.encryptedPayload || !exportData.salt) {
                throw new InvalidExportFormatError('Missing required fields');
            }
            // Validate salt format
            validateSalt(exportData.salt);
            // Validate custom password if provided
            if (options?.password !== undefined) {
                validateExportPassword(options.password);
            }
            // Determine import password - use provided password or storage password
            const importPassword = options?.password ?? await getPasswordFromProvider(passwordProvider);
            // Decrypt the payload - use single generic error to prevent oracle attacks
            let payload;
            try {
                const salt = Buffer.from(exportData.salt, 'hex');
                const importEncryption = new StorageEncryption(importPassword, salt);
                const decryptedJson = importEncryption.decrypt(exportData.encryptedPayload);
                payload = JSON.parse(decryptedJson);
            }
            catch {
                // Single generic error - don't reveal whether password was wrong or data was corrupted
                throw new ExportDecryptionError();
            }
            // Validate payload structure (metadata is now inside encrypted payload)
            if (!payload.states ||
                typeof payload.states !== 'object' ||
                typeof payload.version !== 'number' ||
                typeof payload.stateCount !== 'number') {
                throw new ExportDecryptionError();
            }
            // Validate version from authenticated payload
            if (!SUPPORTED_EXPORT_VERSIONS.includes(payload.version)) {
                throw new InvalidExportFormatError(`Export version ${payload.version} is not supported. Supported versions: ${SUPPORTED_EXPORT_VERSIONS.join(', ')}`);
            }
            // stateIds are raw state IDs (not scoped with contract address)
            const stateIds = Object.keys(payload.states);
            // Validate state count matches and is within limits
            if (stateIds.length !== payload.stateCount) {
                throw new ExportDecryptionError();
            }
            if (stateIds.length > maxStates) {
                throw new InvalidExportFormatError(`Too many states in export (${stateIds.length}). Maximum allowed: ${maxStates}`);
            }
            // Check for conflicts if strategy is 'error'
            // Use this.get() which properly scopes the state IDs
            if (conflictStrategy === 'error') {
                let conflictCount = 0;
                for (const stateId of stateIds) {
                    const existing = await this.get(stateId);
                    if (existing !== null) {
                        conflictCount++;
                    }
                }
                if (conflictCount > 0) {
                    throw new ImportConflictError(conflictCount);
                }
            }
            // Import states
            let imported = 0;
            let skipped = 0;
            let overwritten = 0;
            for (const stateId of stateIds) {
                const serializedState = payload.states[stateId];
                const existingState = await this.get(stateId);
                if (existingState !== null) {
                    if (conflictStrategy === 'skip') {
                        skipped++;
                        continue;
                    }
                    else if (conflictStrategy === 'overwrite') {
                        overwritten++;
                    }
                }
                // Deserialize and store the state
                const state = superjson.parse(serializedState);
                await this.set(stateId, state);
                if (existingState === null) {
                    imported++;
                }
            }
            return { imported, skipped, overwritten };
        },
        async exportSigningKeys(options) {
            const maxKeys = options?.maxKeys ?? MAX_EXPORT_SIGNING_KEYS;
            if (options?.password !== undefined) {
                validateSigningKeyExportPassword(options.password);
            }
            const exportPassword = options?.password ?? await getPasswordFromProvider(passwordProvider);
            const allKeys = await getAllEntries(fullConfig.midnightDbName, scopedSigningKeyLevelName, passwordProvider);
            if (allKeys.size === 0) {
                throw new SigningKeyExportError('No signing keys to export');
            }
            if (allKeys.size > maxKeys) {
                throw new SigningKeyExportError(`Too many keys to export (${allKeys.size}). Maximum allowed: ${maxKeys}`);
            }
            const payload = {
                version: CURRENT_EXPORT_VERSION,
                exportedAt: new Date().toISOString(),
                keyCount: allKeys.size,
                keys: Object.fromEntries(allKeys.entries())
            };
            const exportEncryption = new StorageEncryption(exportPassword);
            const encryptedPayload = exportEncryption.encrypt(JSON.stringify(payload));
            return {
                format: 'midnight-signing-key-export',
                encryptedPayload,
                salt: exportEncryption.getSalt().toString('hex')
            };
        },
        async importSigningKeys(exportData, options) {
            const conflictStrategy = options?.conflictStrategy ?? 'error';
            const maxKeys = options?.maxKeys ?? MAX_EXPORT_SIGNING_KEYS;
            if (exportData.format !== 'midnight-signing-key-export') {
                throw new InvalidExportFormatError('Unrecognized export format');
            }
            if (!exportData.encryptedPayload || !exportData.salt) {
                throw new InvalidExportFormatError('Missing required fields');
            }
            validateSalt(exportData.salt);
            if (options?.password !== undefined) {
                validateSigningKeyExportPassword(options.password);
            }
            const importPassword = options?.password ?? await getPasswordFromProvider(passwordProvider);
            let payload;
            try {
                const salt = Buffer.from(exportData.salt, 'hex');
                const importEncryption = new StorageEncryption(importPassword, salt);
                const decryptedJson = importEncryption.decrypt(exportData.encryptedPayload);
                payload = JSON.parse(decryptedJson);
            }
            catch {
                throw new ExportDecryptionError();
            }
            if (!payload.keys ||
                typeof payload.keys !== 'object' ||
                typeof payload.version !== 'number' ||
                typeof payload.keyCount !== 'number') {
                throw new ExportDecryptionError();
            }
            if (!SUPPORTED_EXPORT_VERSIONS.includes(payload.version)) {
                throw new InvalidExportFormatError(`Export version ${payload.version} is not supported. Supported versions: ${SUPPORTED_EXPORT_VERSIONS.join(', ')}`);
            }
            const addresses = Object.keys(payload.keys);
            if (addresses.length !== payload.keyCount) {
                throw new ExportDecryptionError();
            }
            if (addresses.length > maxKeys) {
                throw new InvalidExportFormatError(`Too many keys in export (${addresses.length}). Maximum allowed: ${maxKeys}`);
            }
            if (conflictStrategy === 'error') {
                let conflictCount = 0;
                for (const address of addresses) {
                    const existing = await this.getSigningKey(address);
                    if (existing !== null) {
                        conflictCount++;
                    }
                }
                if (conflictCount > 0) {
                    throw new ImportConflictError(conflictCount, 'signing key');
                }
            }
            let imported = 0;
            let skipped = 0;
            let overwritten = 0;
            for (const address of addresses) {
                const signingKey = payload.keys[address];
                const existingKey = await this.getSigningKey(address);
                if (existingKey !== null) {
                    if (conflictStrategy === 'skip') {
                        skipped++;
                        continue;
                    }
                    else if (conflictStrategy === 'overwrite') {
                        overwritten++;
                    }
                }
                await this.setSigningKey(address, signingKey);
                if (existingKey === null) {
                    imported++;
                }
            }
            return { imported, skipped, overwritten };
        },
        async changePassword(oldPasswordProvider, newPasswordProvider, options) {
            if (contractAddress === null) {
                throw new Error('Contract address not set. Call setContractAddress() before changing password.');
            }
            const lockKey = `${fullConfig.midnightDbName}:${scopedPrivateStateLevelName}`;
            const prefix = `${contractAddress}:`;
            return withPasswordRotationLock(lockKey, async () => {
                const result = await rotateStorePassword({
                    dbName: fullConfig.midnightDbName,
                    storeName: scopedPrivateStateLevelName,
                    oldPasswordProvider,
                    newPasswordProvider,
                    maxEntries: options?.maxEntries ?? DEFAULT_MAX_ROTATION_ENTRIES,
                    shouldProceed: (key) => key.startsWith(prefix)
                });
                invalidateEncryptionCacheForDb(fullConfig.midnightDbName, scopedPrivateStateLevelName, scopedSigningKeyLevelName);
                return result;
            });
        },
        async changeSigningKeysPassword(oldPasswordProvider, newPasswordProvider, options) {
            const lockKey = `${fullConfig.midnightDbName}:${scopedSigningKeyLevelName}`;
            return withPasswordRotationLock(lockKey, async () => {
                const result = await rotateStorePassword({
                    dbName: fullConfig.midnightDbName,
                    storeName: scopedSigningKeyLevelName,
                    oldPasswordProvider,
                    newPasswordProvider,
                    maxEntries: options?.maxEntries ?? DEFAULT_MAX_ROTATION_ENTRIES
                });
                invalidateEncryptionCacheForDb(fullConfig.midnightDbName, scopedPrivateStateLevelName, scopedSigningKeyLevelName);
                return result;
            });
        },
        invalidateEncryptionCache() {
            invalidateEncryptionCacheForDb(fullConfig.midnightDbName, scopedPrivateStateLevelName, scopedSigningKeyLevelName);
        }
    };
};
const migrateSublevel = async (dbName, oldLevelName, newLevelName) => {
    const level = new Level(dbName, { createIfMissing: true });
    try {
        await level.open();
        const oldSubLevel = level.sublevel(oldLevelName, {
            valueEncoding: 'utf-8'
        });
        const newSubLevel = level.sublevel(newLevelName, {
            valueEncoding: 'utf-8'
        });
        try {
            await oldSubLevel.open();
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Failed to open source sublevel "${oldLevelName}": ${errorMessage}. ` +
                `Ensure no other process is accessing the database.`, { cause: error });
        }
        try {
            await newSubLevel.open();
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Failed to open target sublevel "${newLevelName}": ${errorMessage}. ` +
                `Ensure no other process is accessing the database.`, { cause: error });
        }
        let count = 0;
        const operations = [];
        try {
            for await (const [key, value] of oldSubLevel.iterator()) {
                operations.push({ type: 'put', key, value });
                count++;
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : 'Unknown error';
            throw new Error(`Failed to read data from source sublevel "${oldLevelName}" after ${count} entries: ${errorMessage}. ` +
                `Migration incomplete. Source data is unchanged.`, { cause: error });
        }
        if (operations.length > 0) {
            try {
                await newSubLevel.batch(operations);
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : 'Unknown error';
                throw new Error(`Failed to write ${operations.length} entries to target sublevel "${newLevelName}": ${errorMessage}. ` +
                    `Migration incomplete. Target sublevel may contain partial data. ` +
                    `Source data at "${oldLevelName}" is unchanged.`, { cause: error });
            }
        }
        await newSubLevel.close();
        await oldSubLevel.close();
        return count;
    }
    finally {
        try {
            await level.close();
        }
        catch {
            // Don't mask the original error - just ignore the close failure
        }
    }
};
/**
 * Migrates existing unscoped private state and signing key data to account-scoped sublevels.
 *
 * This function copies data from the legacy unscoped locations to the new account-scoped
 * locations. The original data is preserved (not deleted) to allow for safe rollback if needed.
 * To remove old data after successful migration, manually clear the unscoped sublevels.
 *
 * Note: Running this function multiple times is safe but will re-copy all data, overwriting
 * any changes made in the scoped location since the last migration.
 */
const migrateToAccountScoped = async (config) => {
    const fullConfig = { ...DEFAULT_CONFIG, ...config };
    if (!config.accountId || config.accountId.trim().length === 0) {
        throw new Error('accountId is required for migration');
    }
    const scopedPrivateStateLevelName = getScopedLevelName(fullConfig.privateStateStoreName, config.accountId);
    const scopedSigningKeyLevelName = getScopedLevelName(fullConfig.signingKeyStoreName, config.accountId);
    let privateStatesMigrated;
    try {
        privateStatesMigrated = await migrateSublevel(fullConfig.midnightDbName, fullConfig.privateStateStoreName, scopedPrivateStateLevelName);
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Migration failed during private states copy: ${errorMessage}. ` +
            `No data has been migrated. Source data is unchanged.`, { cause: error });
    }
    let signingKeysMigrated;
    try {
        signingKeysMigrated = await migrateSublevel(fullConfig.midnightDbName, fullConfig.signingKeyStoreName, scopedSigningKeyLevelName);
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        throw new Error(`Migration failed during signing keys copy: ${errorMessage}. ` +
            `WARNING: ${privateStatesMigrated} private states were already migrated to scoped location. ` +
            `Signing keys remain at original location. Manual intervention may be required.`, { cause: error });
    }
    return { privateStatesMigrated, signingKeysMigrated };
};

export { DEFAULT_CONFIG, StorageEncryption, decryptValue, levelPrivateStateProvider, migrateToAccountScoped };
