// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/Serialization.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Either, pipe, Schema } from 'effect';
import { OtherWalletError } from './WalletError.js';
import { CoreWallet } from './CoreWallet.js';
import { ProtocolVersion } from '@midnight-ntwrk/wallet-sdk-abstractions';
import { UnshieldedState } from './UnshieldedState.js';
export const makeDefaultV1SerializationCapability = () => {
    const UtxoWithMetaSchema = Schema.Struct({
        utxo: Schema.Struct({
            value: Schema.BigInt,
            owner: Schema.String,
            type: Schema.String,
            intentHash: Schema.String,
            outputNo: Schema.Number,
        }),
        meta: Schema.Struct({
            ctime: Schema.Date,
            registeredForDustGeneration: Schema.Boolean,
        }),
    });
    const SnapshotSchema = Schema.Struct({
        publicKey: Schema.Struct({
            publicKey: Schema.String,
            addressHex: Schema.String,
            address: Schema.String,
        }),
        state: Schema.Struct({
            availableUtxos: Schema.Array(UtxoWithMetaSchema),
            pendingUtxos: Schema.Array(UtxoWithMetaSchema),
        }),
        protocolVersion: Schema.BigInt,
        appliedId: Schema.optional(Schema.BigInt),
        networkId: Schema.String,
    });
    return {
        serialize: (wallet) => {
            const buildSnapshot = (w) => ({
                publicKey: w.publicKey,
                state: UnshieldedState.toArrays(w.state),
                protocolVersion: w.protocolVersion,
                networkId: w.networkId,
                appliedId: w.progress?.appliedId,
            });
            return pipe(wallet, buildSnapshot, Schema.encodeSync(SnapshotSchema), JSON.stringify);
        },
        deserialize: (serialized) => pipe(serialized, Schema.decodeUnknownEither(Schema.parseJson(SnapshotSchema)), Either.mapLeft((err) => new OtherWalletError(err)), Either.map((snapshot) => {
            return CoreWallet.restore(UnshieldedState.restore(snapshot.state.availableUtxos, snapshot.state.pendingUtxos), snapshot.publicKey, {
                highestTransactionId: snapshot.appliedId ?? 0n,
                appliedId: snapshot.appliedId ?? 0n,
            }, ProtocolVersion.ProtocolVersion(snapshot.protocolVersion), snapshot.networkId);
        })),
    };
};
