// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/SyncSchema.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Schema } from 'effect';
import { SafeBigInt } from '@midnight-ntwrk/wallet-sdk-utilities';
const DateFromMillis = Schema.transform(Schema.Number, Schema.DateFromSelf, {
    strict: true,
    decode: (millis) => new Date(millis),
    encode: (date) => date.getTime(),
});
const WireUtxoSchema = Schema.Struct({
    value: SafeBigInt.SafeBigInt,
    owner: Schema.String,
    tokenType: Schema.String,
    intentHash: Schema.String,
    outputIndex: Schema.Number,
    ctime: Schema.Number,
    registeredForDustGeneration: Schema.Boolean,
});
const UtxoWithMetaSchema = Schema.transform(WireUtxoSchema, Schema.typeSchema(Schema.Struct({
    utxo: Schema.Struct({
        value: SafeBigInt.BigIntSchema,
        owner: Schema.String,
        type: Schema.String,
        intentHash: Schema.String,
        outputNo: Schema.Number,
    }),
    meta: Schema.Struct({
        ctime: Schema.DateFromSelf,
        registeredForDustGeneration: Schema.Boolean,
    }),
})), {
    strict: true,
    decode: (wire) => ({
        utxo: {
            value: wire.value,
            owner: wire.owner,
            type: wire.tokenType,
            intentHash: wire.intentHash,
            outputNo: wire.outputIndex,
        },
        meta: {
            ctime: new Date(wire.ctime * 1000),
            registeredForDustGeneration: wire.registeredForDustGeneration,
        },
    }),
    encode: (domain) => ({
        value: domain.utxo.value,
        owner: domain.utxo.owner,
        tokenType: domain.utxo.type,
        intentHash: domain.utxo.intentHash,
        outputIndex: domain.utxo.outputNo,
        ctime: Math.floor(domain.meta.ctime.getTime() / 1000),
        registeredForDustGeneration: domain.meta.registeredForDustGeneration,
    }),
});
export const UnshieldedTransactionSchema = Schema.Data(Schema.Struct({
    id: Schema.Number,
    hash: Schema.String,
    type: Schema.Literal('RegularTransaction', 'SystemTransaction'),
    protocolVersion: Schema.Number,
    identifiers: Schema.optional(Schema.Array(Schema.String)),
    block: Schema.Struct({
        timestamp: DateFromMillis,
    }),
    fees: Schema.optional(Schema.Struct({
        paidFees: SafeBigInt.SafeBigInt,
        estimatedFees: SafeBigInt.SafeBigInt,
    })),
    transactionResult: Schema.optional(Schema.Struct({
        status: Schema.Literal('SUCCESS', 'FAILURE', 'PARTIAL_SUCCESS'),
        segments: Schema.NullOr(Schema.Array(Schema.Struct({
            id: Schema.Number,
            success: Schema.Boolean,
        }))),
    })),
}));
const UnshieldedUpdateWireSchema = Schema.Struct({
    type: Schema.Literal('UnshieldedTransaction'),
    transaction: UnshieldedTransactionSchema,
    createdUtxos: Schema.Array(UtxoWithMetaSchema),
    spentUtxos: Schema.Array(UtxoWithMetaSchema),
});
export const UnshieldedUpdateSchema = Schema.transform(UnshieldedUpdateWireSchema, Schema.typeSchema(Schema.Struct({
    type: Schema.Literal('UnshieldedTransaction'),
    transaction: UnshieldedTransactionSchema,
    createdUtxos: Schema.Array(Schema.typeSchema(UtxoWithMetaSchema)),
    spentUtxos: Schema.Array(Schema.typeSchema(UtxoWithMetaSchema)),
    status: Schema.Literal('SUCCESS', 'FAILURE', 'PARTIAL_SUCCESS'),
})), {
    strict: true,
    decode: (wire) => {
        const isSystemTransaction = wire.transaction.type === 'SystemTransaction';
        return {
            ...wire,
            status: isSystemTransaction ? 'SUCCESS' : wire.transaction.transactionResult.status,
        };
    },
    encode: ({ status: _status, ...rest }) => rest,
});
export const ProgressSchema = Schema.Struct({
    type: Schema.Literal('UnshieldedTransactionsProgress'),
    highestTransactionId: Schema.Number,
});
export const WalletSyncUpdateSchema = Schema.Union(UnshieldedUpdateSchema, ProgressSchema);
