// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/TransactionOps.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Either, pipe, Array as Arr } from 'effect';
import { Imbalances } from '@midnight-ntwrk/wallet-sdk-capabilities';
import { TransactingError } from './WalletError.js';
export const TransactionOps = {
    getSignatureData(tx, segment) {
        if (!tx.intents) {
            return Either.left(new TransactingError({ message: 'Transaction has no intents' }));
        }
        const intent = tx.intents.get(segment);
        if (!intent) {
            return Either.left(new TransactingError({ message: `Intent with segment ${segment} was not found` }));
        }
        return Either.try({
            try: () => intent.signatureData(segment),
            catch: (error) => new TransactingError({ message: 'Failed to get offer signature data', cause: error }),
        });
    },
    getSegments(transaction) {
        return transaction.intents?.keys().toArray() ?? [];
    },
    // @TODO - https://shielded.atlassian.net/browse/PM-21260
    addSignature(transaction, signature, segment) {
        return Either.gen(function* () {
            if (!transaction.intents || transaction.intents.size === 0) {
                return yield* Either.left(new TransactingError({ message: 'No intents found in the transaction' }));
            }
            const intent = transaction.intents?.get(segment);
            if (!intent) {
                return yield* Either.left(new TransactingError({ message: `Intent with id ${segment} was not found` }));
            }
            if (TransactionOps.isIntentBound(intent)) {
                return yield* Either.left(new TransactingError({ message: `Intent at segment ${segment} is already bound` }));
            }
            if (intent.fallibleUnshieldedOffer) {
                intent.fallibleUnshieldedOffer = yield* TransactionOps.addSignaturesToOffer(intent.fallibleUnshieldedOffer, signature, segment, 'fallible');
            }
            if (intent.guaranteedUnshieldedOffer) {
                intent.guaranteedUnshieldedOffer = yield* TransactionOps.addSignaturesToOffer(intent.guaranteedUnshieldedOffer, signature, segment, 'guaranteed');
            }
            transaction.intents = transaction.intents.set(segment, intent);
            return transaction;
        });
    },
    getImbalances(transaction, segment) {
        const imbalances = transaction
            .imbalances(segment)
            .entries()
            .filter(([token, value]) => token.tag === 'unshielded' && value !== 0n)
            .map(([token, value]) => [token.raw.toString(), value])
            .toArray();
        return Imbalances.fromEntries(imbalances);
    },
    addSignaturesToOffer(offer, signature, segment, offerType) {
        return pipe(offer.inputs, Arr.map((_, i) => offer.signatures.at(i) ?? signature), (signatures) => Either.try({
            try: () => offer.addSignatures(signatures),
            catch: (error) => new TransactingError({
                message: `Failed to add ${offerType} signature at segment ${segment}`,
                cause: error,
            }),
        }));
    },
    isIntentBound(intent) {
        return intent.binding.instance === 'binding';
    },
    extractOwnInputs(transaction, signatureVerifyingKey) {
        const segments = TransactionOps.getSegments(transaction);
        return pipe(segments, Arr.flatMap((segment) => {
            const intent = transaction.intents?.get(segment);
            if (!intent) {
                return [];
            }
            const { guaranteedUnshieldedOffer, fallibleUnshieldedOffer } = intent;
            const ownedInputsfromGuaranteedSection = guaranteedUnshieldedOffer?.inputs
                ? guaranteedUnshieldedOffer.inputs.filter((input) => input.owner === signatureVerifyingKey)
                : [];
            const ownedInputsfromFallibleSection = fallibleUnshieldedOffer
                ? fallibleUnshieldedOffer.inputs.filter((input) => input.owner === signatureVerifyingKey)
                : [];
            return [...ownedInputsfromGuaranteedSection, ...ownedInputsfromFallibleSection];
        }));
    },
};
