// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/V1Builder.js
import { Effect, Scope } from 'effect';
import { WalletSeed } from '@midnight-ntwrk/wallet-sdk-abstractions';
import { RunningV1Variant, V1Tag } from './RunningV1Variant.js';
import { makeDefaultV1SerializationCapability } from './Serialization.js';
import { makeDefaultSyncService, makeDefaultSyncCapability, } from './Sync.js';
import { makeDefaultTransactingCapability, } from './Transacting.js';
import { makeDefaultCoinsAndBalancesCapability } from './CoinsAndBalances.js';
import { makeDefaultKeysCapability } from './Keys.js';
import { chooseCoin } from '@midnight-ntwrk/wallet-sdk-capabilities';
import { CoreWallet } from './CoreWallet.js';
import { makeDefaultTransactionHistoryService, } from './TransactionHistory.js';
import { createKeystore, PublicKey } from '../KeyStore.js';
const V1BuilderSymbol = {
    typeId: Symbol('@midnight-ntwrk/unshielded-wallet#V1Builder'),
};
export class V1Builder {
    #buildState;
    constructor(buildState = {}) {
        this.#buildState = buildState;
    }
    withDefaults() {
        return this.withSyncDefaults()
            .withSerializationDefaults()
            .withTransactingDefaults()
            .withCoinsAndBalancesDefaults()
            .withTransactionHistoryDefaults()
            .withKeysDefaults()
            .withCoinSelectionDefaults();
    }
    withSyncDefaults() {
        return this.withSync(makeDefaultSyncService, makeDefaultSyncCapability);
    }
    withSync(syncService, syncCapability) {
        return new V1Builder({
            ...this.#buildState,
            syncService,
            syncCapability,
        });
    }
    withSerializationDefaults() {
        return this.withSerialization(makeDefaultV1SerializationCapability);
    }
    withSerialization(serializationCapability) {
        return new V1Builder({
            ...this.#buildState,
            serializationCapability,
        });
    }
    withTransactingDefaults() {
        return this.withTransacting(makeDefaultTransactingCapability);
    }
    withTransacting(transactingCapability) {
        return new V1Builder({
            ...this.#buildState,
            transactingCapability,
        });
    }
    withCoinSelection(coinSelection) {
        return new V1Builder({
            ...this.#buildState,
            coinSelection,
        });
    }
    withCoinSelectionDefaults() {
        return this.withCoinSelection(() => chooseCoin);
    }
    withCoinsAndBalancesDefaults() {
        return this.withCoinsAndBalances(makeDefaultCoinsAndBalancesCapability);
    }
    withCoinsAndBalances(coinsAndBalancesCapability) {
        return new V1Builder({
            ...this.#buildState,
            coinsAndBalancesCapability,
        });
    }
    withTransactionHistoryDefaults() {
        return this.withTransactionHistory(makeDefaultTransactionHistoryService);
    }
    withTransactionHistory(transactionHistoryService) {
        return new V1Builder({
            ...this.#buildState,
            transactionHistoryService,
        });
    }
    withKeysDefaults() {
        return this.withKeys(makeDefaultKeysCapability);
    }
    withKeys(keysCapability) {
        return new V1Builder({
            ...this.#buildState,
            keysCapability,
        });
    }
    build(configuration) {
        const v1Context = this.#buildContextFromBuildState(configuration);
        const { networkId } = configuration;
        return {
            __polyTag__: V1Tag,
            coinsAndBalances: v1Context.coinsAndBalancesCapability,
            keys: v1Context.keysCapability,
            serialization: v1Context.serializationCapability,
            transactionHistory: v1Context.transactionHistoryService,
            start(context) {
                return Effect.gen(function* () {
                    const scope = yield* Scope.Scope;
                    return new RunningV1Variant(scope, context, v1Context);
                });
            },
            migrateState(_previousState) {
                const seed = WalletSeed.fromString('0000000000000000000000000000000000000000000000000000000000000001');
                return Effect.succeed(CoreWallet.init(PublicKey.fromKeyStore(createKeystore(seed, networkId)), networkId));
            },
            deserializeState: (serialized) => {
                return v1Context.serializationCapability.deserialize(serialized);
            },
        };
    }
    #buildContextFromBuildState(configuration) {
        if (!isBuildStateFull(this.#buildState)) {
            throw new Error('Not all components are configured in V1 Builder');
        }
        const { syncCapability, syncService, transactingCapability, serializationCapability, coinSelection, coinsAndBalancesCapability, keysCapability, transactionHistoryService, } = this.#buildState;
        const getContext = () => context;
        const context = {
            serializationCapability: serializationCapability(configuration, getContext),
            syncCapability: syncCapability(configuration, getContext),
            syncService: syncService(configuration, getContext),
            transactingCapability: transactingCapability(configuration, getContext),
            coinsAndBalancesCapability: coinsAndBalancesCapability(configuration, getContext),
            keysCapability: keysCapability(configuration, getContext),
            coinSelection: coinSelection(configuration, getContext),
            transactionHistoryService: transactionHistoryService(configuration, getContext),
        };
        return context;
    }
}
const isBuildStateFull = (buildState) => {
    const allBuildStateKeys = [
        'syncService',
        'syncCapability',
        'transactingCapability',
        'coinSelection',
        'serializationCapability',
        'coinsAndBalancesCapability',
        'keysCapability',
        'transactionHistoryService',
    ];
    return allBuildStateKeys.every((key) => typeof buildState[key] == 'function');
};
