// @midnight-ntwrk/wallet-sdk-utilities/dist/ArrayOps.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { reduce, match } from 'effect/Array';
import { dual } from 'effect/Function';
export const fold = dual(2, (arr, folder) => arr.reduce(folder));
export const generalSum = dual(2, (arr, monoid) => reduce(arr, monoid.empty, monoid.combine));
const numberAdditionMonoid = {
    empty: 0,
    combine: (a, b) => a + b,
};
const bigintAdditionMonoid = {
    empty: 0n,
    combine: (a, b) => a + b,
};
export const sumNumber = generalSum(numberAdditionMonoid);
export const sumBigInt = generalSum(bigintAdditionMonoid);
export const assertNonEmpty = (arr) => {
    return match(arr, {
        onNonEmpty: (refined) => refined,
        onEmpty: () => {
            throw new Error('Expected non-empty array');
        },
    });
};
