// @midnight-ntwrk/wallet-sdk-utilities/dist/hlist.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
/* eslint-disable @typescript-eslint/no-explicit-any */
import { dispatch } from './polyFunction.js';
export const empty = [];
export const prepend = (list, element) => {
    return [element, ...list];
};
export const append = (list, element) => {
    return [...list, element];
};
export function headOr(list, def) {
    if (list.length == 0) {
        return def();
    }
    else {
        return list.at(0);
    }
}
export const head = (list) => {
    return headOr(list, () => {
        throw new Error('Cannot get head from empty hlist');
    });
};
export const tail = (list) => {
    return list.toSpliced(0, 1);
};
export const reverse = (list) => {
    return list.toReversed();
};
export const find = (list, predicate) => {
    return list.find(predicate);
};
export const foldLeft = (list, acc, folder) => {
    return list.reduce((acc, item) => dispatch(item, folder(acc)), acc);
};
export const foldRight = (list, acc, folder) => {
    return list.reduceRight((acc, item) => dispatch(item, folder(acc)), acc);
};
