// @midnight-ntwrk/wallet-sdk-prover-client/dist/effect/WasmProver.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import Worker from 'web-worker';
import { Effect, Layer, Schema, pipe } from 'effect';
import { ClientError } from '@midnight-ntwrk/wallet-sdk-utilities/networking';
import { ProverClient } from './ProverClient.js';
/**
 * Creates a layer for a {@link ProverClient} that sends requests to a Wasm Prover.
 *
 * @param config The Key Material Provider to use when configuring the prover's elements of the layer.
 * @returns A `Layer` for {@link ProverClient} that sends requests to a configured Wasm Prover.
 */
export const layer = (config) => Layer.effect(ProverClient, Effect.succeed(new WasmProverImpl(config.keyMaterialProvider)));
export const create = (config) => {
    return Effect.succeed(new WasmProverImpl(config.keyMaterialProvider));
};
const MAX_TIME_TO_PROCESS = 10 * 60 * 1000;
export const CheckOperationSchema = Schema.Struct({
    op: Schema.Literal('check'),
    args: Schema.Tuple(Schema.Uint8Array),
});
export const ProveOperationSchema = Schema.Struct({
    op: Schema.Literal('prove'),
    args: Schema.Tuple(Schema.Uint8Array, Schema.Union(Schema.BigIntFromSelf, Schema.Undefined)),
});
export const LookupKeyRequestSchema = Schema.Struct({
    op: Schema.Literal('lookupKey'),
    keyLocation: Schema.String,
});
export const GetParamsRequestSchema = Schema.Struct({
    op: Schema.Literal('getParams'),
    k: Schema.Number,
});
export const ResponseFromWorkerSchema = Schema.Struct({
    op: Schema.Literal('result'),
    value: Schema.Union(Schema.Uint8Array, Schema.Array(Schema.Union(Schema.BigIntFromSelf, Schema.Undefined))),
});
const ProvingKeyMaterialSchema = Schema.Struct({
    proverKey: Schema.Uint8Array,
    verifierKey: Schema.Uint8Array,
    ir: Schema.Uint8Array,
});
export const LookupKeyOperationResultSchema = Schema.Struct({
    op: Schema.Literal('lookupKey'),
    keyLocation: Schema.String,
    result: Schema.optional(ProvingKeyMaterialSchema),
});
export const GetParamsOperationResultSchema = Schema.Struct({
    op: Schema.Literal('getParams'),
    k: Schema.Number,
    result: Schema.Uint8Array,
});
const MessageDataSchema = Schema.Union(LookupKeyRequestSchema, GetParamsRequestSchema, ResponseFromWorkerSchema);
const callProverWorker = ({ kmProvider, op, args }) => {
    return new Promise((resolve, reject) => {
        const currentFile = import.meta.url;
        const worker = new Worker(new URL(`../../dist/proof-worker.js`, currentFile), { type: 'module' });
        // initialize worker
        worker.postMessage(op === 'check'
            ? Schema.encodeSync(CheckOperationSchema)({ op, args: [args[0]] })
            : Schema.encodeSync(ProveOperationSchema)({ op, args: [args[0], args[1]] }));
        // a message from the worker
        worker.addEventListener('message', ({ data }) => {
            const decoded = Schema.decodeUnknownSync(MessageDataSchema)(data);
            const { op } = decoded;
            if (op === 'lookupKey') {
                const { keyLocation } = decoded;
                kmProvider
                    .lookupKey(keyLocation)
                    .then((result) => {
                    worker.postMessage(Schema.encodeSync(LookupKeyOperationResultSchema)({ op, keyLocation, result }));
                })
                    .catch((e) => {
                    worker.terminate();
                    reject(e);
                });
            }
            else if (op === 'getParams') {
                const { k } = decoded;
                kmProvider
                    .getParams(k)
                    .then((result) => {
                    worker.postMessage(Schema.encodeSync(GetParamsOperationResultSchema)({ op, k, result }));
                })
                    .catch((e) => {
                    worker.terminate();
                    reject(e);
                });
            }
            else if (op === 'result') {
                const { value } = decoded;
                worker.terminate();
                resolve(value);
            }
        });
        worker.addEventListener('error', (e) => {
            worker.terminate();
            reject(Error(e.message));
        });
        setTimeout(() => {
            worker.terminate();
            reject(new Error(`${op} action timed out`));
        }, MAX_TIME_TO_PROCESS);
    });
};
class WasmProverImpl {
    constructor(keyMaterialProvider) {
        this.keyMaterialProvider = keyMaterialProvider;
    }
    keyMaterialProvider;
    wasmProverProvider = (keyMaterialProvider) => ({
        check: async (serializedPreimage, _keyLocation) => callProverWorker({
            kmProvider: keyMaterialProvider ?? this.keyMaterialProvider,
            op: 'check',
            args: [serializedPreimage],
        }),
        prove: async (serializedPreimage, _keyLocation, overwriteBindingInput) => callProverWorker({
            kmProvider: keyMaterialProvider ?? this.keyMaterialProvider,
            op: 'prove',
            args: [serializedPreimage, overwriteBindingInput],
        }),
    });
    proveTransaction(transaction, costModel, keyMaterialProvider) {
        return pipe(Effect.succeed(this.wasmProverProvider(keyMaterialProvider)), Effect.flatMap((provider) => Effect.tryPromise({
            try: () => transaction.prove(provider, costModel),
            catch: (error) => error instanceof ClientError
                ? error
                : new ClientError({ message: 'Failed to prove transaction', cause: error }),
        })));
    }
    asProvingProvider() {
        return this.wasmProverProvider();
    }
}
export const makeDefaultKeyMaterialProvider = () => {
    const cache = new Map();
    const s3 = 'https://midnight-s3-fileshare-dev-eu-west-1.s3.eu-west-1.amazonaws.com';
    const ver = 9;
    const fetchWithRetry = async (url, retries = 5) => {
        for (let i = 0; i < retries; i++) {
            try {
                return await fetch(url);
            }
            catch (e) {
                // eslint-disable-next-line no-console
                console.error('Failed to fetch at attempt', i + 1, url, e);
                // cooldown a bit before retrying
                await new Promise((resolve) => setTimeout(resolve, 1000 * Math.pow(2, i + 1)));
            }
        }
        throw new Error(`Failed to fetch ${url} after ${retries} attempts`);
    };
    const keyMaterialProvider = {
        lookupKey: async (keyLocation) => {
            const pth = {
                'midnight/zswap/spend': `zswap/${ver}/spend`,
                'midnight/zswap/output': `zswap/${ver}/output`,
                'midnight/zswap/sign': `zswap/${ver}/sign`,
                'midnight/dust/spend': `dust/${ver}/spend`,
            }[keyLocation];
            if (pth === undefined) {
                return undefined;
            }
            if (cache.has(pth)) {
                return cache.get(pth);
            }
            const pk = await fetchWithRetry(`${s3}/${pth}.prover`);
            const vk = await fetchWithRetry(`${s3}/${pth}.verifier`);
            const ir = await fetchWithRetry(`${s3}/${pth}.bzkir`);
            const result = {
                proverKey: new Uint8Array(await pk.arrayBuffer()),
                verifierKey: new Uint8Array(await vk.arrayBuffer()),
                ir: new Uint8Array(await ir.arrayBuffer()),
            };
            cache.set(pth, result);
            return result;
        },
        getParams: async (k) => {
            const cacheKey = `params-${k}`;
            if (cache.has(cacheKey)) {
                return cache.get(cacheKey);
            }
            const data = await fetchWithRetry(`${s3}/bls_midnight_2p${k}`);
            const result = new Uint8Array(await data.arrayBuffer());
            cache.set(cacheKey, result);
            return result;
        },
    };
    return keyMaterialProvider;
};
