// @midnight-ntwrk/wallet-sdk-utilities/dist/EitherOps.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Effect, Either, Data } from 'effect';
import { dual } from 'effect/Function';
export const toEffect = (either) => {
    return Either.match(either, {
        onLeft: (l) => Effect.fail(l),
        onRight: (r) => Effect.succeed(r),
    });
};
export const flatMapLeft = dual(2, (either, cb) => {
    return Either.match(either, {
        onRight: (r) => Either.right(r),
        onLeft: cb,
    });
});
export class LeftError extends Data.TaggedError('LeftError') {
    constructor({ cause }) {
        super({ message: 'Unexpected left value', cause });
    }
}
export const getOrThrowLeft = (either) => {
    return Either.match(either, {
        onRight: (r) => r,
        onLeft: (l) => {
            throw new LeftError({ cause: l });
        },
    });
};
