// @midnight-ntwrk/wallet-sdk-indexer-client/dist/effect/Query.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Effect, Context, Effectable, Option, identity } from 'effect';
import { QueryClient } from './QueryClient.js';
/**
 * Constructs a new invocable GraphQL query.
 *
 * @param name The name of the tag to be associated with this query.
 * @param document A parsed GraphQL query document that represents the query.
 * @returns A {@link Query}.
 */
export const make = (name, document) => new QueryImpl(`${name}Query`, document);
class QueryImpl extends Effectable.Class {
    name;
    document;
    constructor(name, document) {
        super();
        this.document = document;
        this.name = name;
        this.tag = Context.GenericTag(name);
        this.run = ((variables) => Effect.flatMap(this, (f) => f(variables)));
    }
    tag;
    run;
    commit() {
        const self = this; // eslint-disable-line @typescript-eslint/no-this-alias
        return Effect.gen(function* () {
            return Option.match(yield* Effect.serviceOption(self.tag), {
                onSome: identity,
                onNone: () => self.defaultFn.bind(self),
            });
        });
    }
    defaultFn(variables) {
        return QueryClient.pipe(Effect.flatMap((client) => client.query(this.document, variables)));
    }
}
