// @midnight-ntwrk/wallet-sdk-indexer-client/dist/effect/Subscription.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Effect, Stream, Context, Effectable, Option, identity } from 'effect';
import { SubscriptionClient } from './SubscriptionClient.js';
export const make = (name, document) => new SubscriptionImpl(`${name}Subscription`, document);
class SubscriptionImpl extends Effectable.Class {
    name;
    document;
    constructor(name, document) {
        super();
        this.document = document;
        this.name = name;
        this.tag = Context.GenericTag(name);
        this.run = ((variables) => Stream.flatMap(this, (f) => f(variables)));
    }
    tag;
    run;
    commit() {
        const self = this; // eslint-disable-line @typescript-eslint/no-this-alias
        return Effect.gen(function* () {
            return Option.match(yield* Effect.serviceOption(self.tag), {
                onSome: identity,
                onNone: () => self.defaultFn.bind(self),
            });
        });
    }
    defaultFn(variables) {
        return SubscriptionClient.pipe(Stream.flatMap((client) => client.subscribe(this.document, variables)));
    }
}
