// @midnight-ntwrk/midnight-js-indexer-public-data-provider/dist/index.mjs
import { ApolloClient, InMemoryCache } from '@apollo/client/core/core.cjs';
import { from, split } from '@apollo/client/link/core/core.cjs';
import { createHttpLink } from '@apollo/client/link/http/http.cjs';
import { RetryLink } from '@apollo/client/link/retry/retry.cjs';
import { GraphQLWsLink } from '@apollo/client/link/subscriptions/subscriptions.cjs';
import { getMainDefinition } from '@apollo/client/utilities/utilities.cjs';
import { ContractState } from '@midnight-ntwrk/compact-runtime';
import { Transaction, ZswapChainState } from '@midnight-ntwrk/ledger-v7';
import { InvalidProtocolSchemeError, SucceedEntirely, FailFallible, FailEntirely, SegmentSuccess, SegmentFail } from '@midnight-ntwrk/midnight-js-types';
import { assertIsContractAddress } from '@midnight-ntwrk/midnight-js-utils';
import { Buffer } from 'buffer';
import fetch from 'cross-fetch';
import { createClient } from 'graphql-ws';
import * as ws from 'isomorphic-ws';
import * as Rx from 'rxjs';

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * An error describing the causes of error that occurred during server-side execution of
 * a query against the Indexer.
 */
class IndexerFormattedError extends Error {
    cause;
    /**
     * @param cause An array of GraphQL errors that occurred during the server-side execution.
     */
    constructor(cause) {
        super(`Indexer GraphQL error(s):\n${cause.reduce((acc, c, idx) => `${idx + 1}. ${c.message}:\n\t${acc}`, '')}`);
        this.cause = cause;
    }
}

const BlockHashQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "BLOCK_HASH_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "BlockOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "block" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }] } }] } }] };
const TxIdQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "TX_ID_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "TransactionOffset" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transactions" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "protocolVersion" } }, { "kind": "Field", "name": { "kind": "Name", "value": "raw" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedCreatedOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedSpentOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "block" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "author" } }, { "kind": "Field", "name": { "kind": "Name", "value": "timestamp" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "RegularTransaction" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "identifiers" } }, { "kind": "Field", "name": { "kind": "Name", "value": "fees" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "estimatedFees" } }, { "kind": "Field", "name": { "kind": "Name", "value": "paidFees" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "transactionResult" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "segments" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "success" } }] } }] } }] } }] } }] } }] };
const DeployTxQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "DEPLOY_TX_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractDeploy" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transaction" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "protocolVersion" } }, { "kind": "Field", "name": { "kind": "Name", "value": "raw" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "address" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "block" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "author" } }, { "kind": "Field", "name": { "kind": "Name", "value": "timestamp" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedCreatedOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedSpentOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "RegularTransaction" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "identifiers" } }, { "kind": "Field", "name": { "kind": "Name", "value": "fees" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "estimatedFees" } }, { "kind": "Field", "name": { "kind": "Name", "value": "paidFees" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "transactionResult" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "segments" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "success" } }] } }] } }] } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractUpdate" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transaction" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "protocolVersion" } }, { "kind": "Field", "name": { "kind": "Name", "value": "raw" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "address" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "block" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "author" } }, { "kind": "Field", "name": { "kind": "Name", "value": "timestamp" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedCreatedOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedSpentOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "RegularTransaction" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "identifiers" } }, { "kind": "Field", "name": { "kind": "Name", "value": "fees" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "estimatedFees" } }, { "kind": "Field", "name": { "kind": "Name", "value": "paidFees" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "transactionResult" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "segments" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "success" } }] } }] } }] } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractCall" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "deploy" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transaction" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "protocolVersion" } }, { "kind": "Field", "name": { "kind": "Name", "value": "raw" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "address" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "block" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "author" } }, { "kind": "Field", "name": { "kind": "Name", "value": "timestamp" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedCreatedOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "unshieldedSpentOutputs" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "owner" } }, { "kind": "Field", "name": { "kind": "Name", "value": "intentHash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "value" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "RegularTransaction" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "identifiers" } }, { "kind": "Field", "name": { "kind": "Name", "value": "fees" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "estimatedFees" } }, { "kind": "Field", "name": { "kind": "Name", "value": "paidFees" } }] } }, { "kind": "Field", "name": { "kind": "Name", "value": "transactionResult" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "status" } }, { "kind": "Field", "name": { "kind": "Name", "value": "segments" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "id" } }, { "kind": "Field", "name": { "kind": "Name", "value": "success" } }] } }] } }] } }] } }] } }] } }] } }] } }] };
const DeployContractStateTxQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "DEPLOY_CONTRACT_STATE_TX_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractDeploy" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractUpdate" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractCall" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "deploy" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transaction" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "address" } }, { "kind": "Field", "name": { "kind": "Name", "value": "state" } }] } }] } }] } }] } }] } }] } }] };
const LatestContractTxBlockHeightQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "LATEST_CONTRACT_TX_BLOCK_HEIGHT_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "transaction" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "block" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "height" } }] } }] } }] } }] } }] };
const TxsFromBlockSubDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "subscription", "name": { "kind": "Name", "value": "TXS_FROM_BLOCK_SUB" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "BlockOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "blocks" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "height" } }, { "kind": "Field", "name": { "kind": "Name", "value": "transactions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "hash" } }, { "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }, { "kind": "Field", "name": { "kind": "Name", "value": "address" } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "RegularTransaction" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "identifiers" } }] } }] } }] } }] } }] };
const ContractStateQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "CONTRACT_STATE_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractActionOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }] } }] } }] };
const ContractStateSubDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "subscription", "name": { "kind": "Name", "value": "CONTRACT_STATE_SUB" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "BlockOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }] } }] } }] };
const BothStateQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "BOTH_STATE_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractActionOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "state" } }, { "kind": "Field", "name": { "kind": "Name", "value": "zswapState" } }] } }] } }] };
const UnshieldedBalanceQueryDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "UNSHIELDED_BALANCE_QUERY" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractDeploy" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractUpdate" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractCall" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "deploy" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }] } }] } }] } }] };
const QueryUnshieldedBalancesWithOffsetDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "query", "name": { "kind": "Name", "value": "QUERY_UNSHIELDED_BALANCES_WITH_OFFSET" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractActionOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractAction" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractDeploy" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractUpdate" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractCall" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "deploy" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }] } }] } }] } }] };
const UnshieldedBalanceSubDocument = { "kind": "Document", "definitions": [{ "kind": "OperationDefinition", "operation": "subscription", "name": { "kind": "Name", "value": "UNSHIELDED_BALANCE_SUB" }, "variableDefinitions": [{ "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } }, "type": { "kind": "NonNullType", "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "HexEncoded" } } } }, { "kind": "VariableDefinition", "variable": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } }, "type": { "kind": "NamedType", "name": { "kind": "Name", "value": "BlockOffset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "contractActions" }, "arguments": [{ "kind": "Argument", "name": { "kind": "Name", "value": "address" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "address" } } }, { "kind": "Argument", "name": { "kind": "Name", "value": "offset" }, "value": { "kind": "Variable", "name": { "kind": "Name", "value": "offset" } } }], "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractDeploy" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractUpdate" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }, { "kind": "InlineFragment", "typeCondition": { "kind": "NamedType", "name": { "kind": "Name", "value": "ContractCall" } }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "deploy" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "unshieldedBalances" }, "selectionSet": { "kind": "SelectionSet", "selections": [{ "kind": "Field", "name": { "kind": "Name", "value": "tokenType" } }, { "kind": "Field", "name": { "kind": "Name", "value": "amount" } }] } }] } }] } }] } }] } }] };

/* eslint-disable */
const documents = {
    "\n  query BLOCK_HASH_QUERY($offset: BlockOffset) {\n    block(offset: $offset) {\n      height\n      hash\n    }\n  }": BlockHashQueryDocument,
    "\n  query TX_ID_QUERY($offset: TransactionOffset!) {\n    transactions(offset: $offset) {\n      id\n      protocolVersion\n      raw\n      hash\n      unshieldedCreatedOutputs {\n        owner\n        intentHash\n        tokenType\n        value\n      }\n      unshieldedSpentOutputs {\n        owner\n        intentHash\n        tokenType\n        value\n      }\n      block {\n        height\n        hash\n        author\n        timestamp\n      }\n      ... on RegularTransaction {\n        identifiers\n        fees {\n          estimatedFees\n          paidFees\n        }\n        transactionResult {\n          status\n          segments {\n            id\n            success\n          }\n        }\n      }\n    }\n  }": TxIdQueryDocument,
    "\n  query DEPLOY_TX_QUERY($address: HexEncoded!) {\n    contractAction(address: $address) {\n      ... on ContractDeploy {\n        transaction {\n          id\n          protocolVersion\n\t        raw\n          hash\n          contractActions {\n            address\n          }\n          block {\n            height\n            hash\n            author\n            timestamp\n          }\n          unshieldedCreatedOutputs {\n            owner\n            intentHash\n            tokenType\n            value\n          }\n          unshieldedSpentOutputs {\n            owner\n            intentHash\n            tokenType\n            value\n          }\n          ... on RegularTransaction {\n            identifiers\n            fees {\n              estimatedFees\n              paidFees\n            }\n            transactionResult {\n              status\n              segments {\n                id\n                success\n              }\n            }\n          }\n        }\n      }\n      ... on ContractUpdate {\n        transaction {\n          id\n          protocolVersion\n\t        raw\n          hash\n          contractActions {\n            address\n          }\n          block {\n            height\n            hash\n            author\n            timestamp\n          }\n          unshieldedCreatedOutputs {\n            owner\n            intentHash\n            tokenType\n            value\n          }\n          unshieldedSpentOutputs {\n            owner\n            intentHash\n            tokenType\n            value\n          }\n          ... on RegularTransaction {\n            identifiers\n            fees {\n              estimatedFees\n              paidFees\n            }\n            transactionResult {\n              status\n              segments {\n                id\n                success\n              }\n            }\n          }\n        }\n      }\n      ... on ContractCall {\n        deploy {\n          transaction {\n            id\n            protocolVersion\n\t          raw\n            hash\n            contractActions {\n              address\n            }\n            block {\n              height\n              hash\n              author\n              timestamp\n            }\n            unshieldedCreatedOutputs {\n              owner\n              intentHash\n              tokenType\n              value\n            }\n            unshieldedSpentOutputs {\n              owner\n              intentHash\n              tokenType\n              value\n            }\n            ... on RegularTransaction {\n              identifiers\n              fees {\n                estimatedFees\n                paidFees\n              }\n              transactionResult {\n                status\n                segments {\n                  id\n                  success\n                }\n              }\n            }\n          }\n        }\n      }\n    }\n  }": DeployTxQueryDocument,
    "\n  query DEPLOY_CONTRACT_STATE_TX_QUERY($address: HexEncoded!) {\n    contractAction(address: $address) {\n      ... on ContractDeploy {\n        state\n      }\n      ... on ContractUpdate {\n        state\n      }\n      ... on ContractCall {\n        deploy {\n          transaction {\n            contractActions {\n              address\n              state\n            }\n          }\n        }\n      }\n    }\n  }": DeployContractStateTxQueryDocument,
    "\n  query LATEST_CONTRACT_TX_BLOCK_HEIGHT_QUERY($address: HexEncoded!) {\n    contractAction(address: $address) {\n      transaction {\n        block {\n          height\n        }\n      }\n    }\n  }": LatestContractTxBlockHeightQueryDocument,
    "\n  subscription TXS_FROM_BLOCK_SUB($offset: BlockOffset) {\n    blocks(offset: $offset) {\n      hash,\n      height,\n      transactions {\n        hash\n        contractActions {\n          state\n          address\n        }\n        ... on RegularTransaction {\n          identifiers\n        }\n      }\n    }\n  }": TxsFromBlockSubDocument,
    "\n  query CONTRACT_STATE_QUERY($address: HexEncoded!, $offset: ContractActionOffset) {\n    contractAction(address: $address, offset: $offset) {\n      state\n    }\n  }": ContractStateQueryDocument,
    "\n  subscription CONTRACT_STATE_SUB($address: HexEncoded!, $offset: BlockOffset) {\n    contractActions(address: $address, offset: $offset) {\n      state\n    }\n  }": ContractStateSubDocument,
    "\n  query BOTH_STATE_QUERY($address: HexEncoded!, $offset: ContractActionOffset) {\n    contractAction(address: $address, offset: $offset) {\n      state\n      zswapState\n    }\n  }": BothStateQueryDocument,
    "\n  query UNSHIELDED_BALANCE_QUERY($address: HexEncoded!) {\n    contractAction(address: $address) {\n      ... on ContractDeploy {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractUpdate {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractCall {\n        deploy {\n          unshieldedBalances {\n            tokenType\n            amount\n          }\n        }\n      }\n    }\n  }": UnshieldedBalanceQueryDocument,
    "\n  query QUERY_UNSHIELDED_BALANCES_WITH_OFFSET($address: HexEncoded!, $offset: ContractActionOffset) {\n    contractAction(address: $address, offset: $offset) {\n      ... on ContractDeploy {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractUpdate {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractCall {\n        deploy {\n          unshieldedBalances {\n            tokenType\n            amount\n          }\n        }\n      }\n    }\n  }": QueryUnshieldedBalancesWithOffsetDocument,
    "\n  subscription UNSHIELDED_BALANCE_SUB($address: HexEncoded!, $offset: BlockOffset) {\n    contractActions(address: $address, offset: $offset) {\n      ... on ContractDeploy {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractUpdate {\n        unshieldedBalances {\n          tokenType\n          amount\n        }\n      }\n      ... on ContractCall {\n        deploy {\n          unshieldedBalances {\n            tokenType\n            amount\n          }\n        }\n      }\n    }\n  }": UnshieldedBalanceSubDocument,
};
function gql(source) {
    return documents[source] ?? {};
}

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const BLOCK_QUERY = gql(`
  query BLOCK_HASH_QUERY($offset: BlockOffset) {
    block(offset: $offset) {
      height
      hash
    }
  }`);
const TX_ID_QUERY = gql(`
  query TX_ID_QUERY($offset: TransactionOffset!) {
    transactions(offset: $offset) {
      id
      protocolVersion
      raw
      hash
      unshieldedCreatedOutputs {
        owner
        intentHash
        tokenType
        value
      }
      unshieldedSpentOutputs {
        owner
        intentHash
        tokenType
        value
      }
      block {
        height
        hash
        author
        timestamp
      }
      ... on RegularTransaction {
        identifiers
        fees {
          estimatedFees
          paidFees
        }
        transactionResult {
          status
          segments {
            id
            success
          }
        }
      }
    }
  }`);
const DEPLOY_TX_QUERY = gql(`
  query DEPLOY_TX_QUERY($address: HexEncoded!) {
    contractAction(address: $address) {
      ... on ContractDeploy {
        transaction {
          id
          protocolVersion
	        raw
          hash
          contractActions {
            address
          }
          block {
            height
            hash
            author
            timestamp
          }
          unshieldedCreatedOutputs {
            owner
            intentHash
            tokenType
            value
          }
          unshieldedSpentOutputs {
            owner
            intentHash
            tokenType
            value
          }
          ... on RegularTransaction {
            identifiers
            fees {
              estimatedFees
              paidFees
            }
            transactionResult {
              status
              segments {
                id
                success
              }
            }
          }
        }
      }
      ... on ContractUpdate {
        transaction {
          id
          protocolVersion
	        raw
          hash
          contractActions {
            address
          }
          block {
            height
            hash
            author
            timestamp
          }
          unshieldedCreatedOutputs {
            owner
            intentHash
            tokenType
            value
          }
          unshieldedSpentOutputs {
            owner
            intentHash
            tokenType
            value
          }
          ... on RegularTransaction {
            identifiers
            fees {
              estimatedFees
              paidFees
            }
            transactionResult {
              status
              segments {
                id
                success
              }
            }
          }
        }
      }
      ... on ContractCall {
        deploy {
          transaction {
            id
            protocolVersion
	          raw
            hash
            contractActions {
              address
            }
            block {
              height
              hash
              author
              timestamp
            }
            unshieldedCreatedOutputs {
              owner
              intentHash
              tokenType
              value
            }
            unshieldedSpentOutputs {
              owner
              intentHash
              tokenType
              value
            }
            ... on RegularTransaction {
              identifiers
              fees {
                estimatedFees
                paidFees
              }
              transactionResult {
                status
                segments {
                  id
                  success
                }
              }
            }
          }
        }
      }
    }
  }`);
const DEPLOY_CONTRACT_STATE_TX_QUERY = gql(`
  query DEPLOY_CONTRACT_STATE_TX_QUERY($address: HexEncoded!) {
    contractAction(address: $address) {
      ... on ContractDeploy {
        state
      }
      ... on ContractUpdate {
        state
      }
      ... on ContractCall {
        deploy {
          transaction {
            contractActions {
              address
              state
            }
          }
        }
      }
    }
  }`);
const LATEST_CONTRACT_TX_BLOCK_HEIGHT_QUERY = gql(`
  query LATEST_CONTRACT_TX_BLOCK_HEIGHT_QUERY($address: HexEncoded!) {
    contractAction(address: $address) {
      transaction {
        block {
          height
        }
      }
    }
  }`);
const TXS_FROM_BLOCK_SUB = gql(`
  subscription TXS_FROM_BLOCK_SUB($offset: BlockOffset) {
    blocks(offset: $offset) {
      hash,
      height,
      transactions {
        hash
        contractActions {
          state
          address
        }
        ... on RegularTransaction {
          identifiers
        }
      }
    }
  }`);
const CONTRACT_STATE_QUERY = gql(`
  query CONTRACT_STATE_QUERY($address: HexEncoded!, $offset: ContractActionOffset) {
    contractAction(address: $address, offset: $offset) {
      state
    }
  }`);
const CONTRACT_STATE_SUB = gql(`
  subscription CONTRACT_STATE_SUB($address: HexEncoded!, $offset: BlockOffset) {
    contractActions(address: $address, offset: $offset) {
      state
    }
  }`);
const CONTRACT_AND_ZSWAP_STATE_QUERY = gql(`
  query BOTH_STATE_QUERY($address: HexEncoded!, $offset: ContractActionOffset) {
    contractAction(address: $address, offset: $offset) {
      state
      zswapState
    }
  }`);
const UNSHIELDED_BALANCE_QUERY = gql(`
  query UNSHIELDED_BALANCE_QUERY($address: HexEncoded!) {
    contractAction(address: $address) {
      ... on ContractDeploy {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractUpdate {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractCall {
        deploy {
          unshieldedBalances {
            tokenType
            amount
          }
        }
      }
    }
  }`);
const QUERY_UNSHIELDED_BALANCES_WITH_OFFSET = gql(`
  query QUERY_UNSHIELDED_BALANCES_WITH_OFFSET($address: HexEncoded!, $offset: ContractActionOffset) {
    contractAction(address: $address, offset: $offset) {
      ... on ContractDeploy {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractUpdate {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractCall {
        deploy {
          unshieldedBalances {
            tokenType
            amount
          }
        }
      }
    }
  }`);
const UNSHIELDED_BALANCE_SUB = gql(`
  subscription UNSHIELDED_BALANCE_SUB($address: HexEncoded!, $offset: BlockOffset) {
    contractActions(address: $address, offset: $offset) {
      ... on ContractDeploy {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractUpdate {
        unshieldedBalances {
          tokenType
          amount
        }
      }
      ... on ContractCall {
        deploy {
          unshieldedBalances {
            tokenType
            amount
          }
        }
      }
    }
  }`);

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const isRegularTransaction = (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
tx) => {
    return 'identifiers' in tx && 'hash' in tx && Array.isArray(tx.identifiers);
};
const maybeThrowGraphQLErrors = (result) => {
    if (result.errors && result.errors.length > 0) {
        throw new IndexerFormattedError(result.errors);
    }
    return result;
};
const maybeThrowApolloError = (result) => {
    if (result.error) {
        throw new Error(result.error.message);
    }
    return result;
};
const maybeThrowErrors = (queryResult) => {
    maybeThrowApolloError(queryResult);
    return maybeThrowGraphQLErrors(queryResult);
};
const toByteArray = (s) => Buffer.from(s, 'hex');
const deserializeContractState = (s) => ContractState.deserialize(toByteArray(s));
const deserializeZswapState = (s) => ZswapChainState.deserialize(toByteArray(s));
const deserializeTransaction = (s) => Transaction.deserialize('signature', 'proof', 'binding', toByteArray(s));
const zenToRx = (zenObservable) => new Rx.Observable((subscriber) => zenObservable.subscribe(subscriber));
/**
 * The default time (in milliseconds) to wait between queries when polling.
 */
const DEFAULT_POLL_INTERVAL = 1000;
// Assumes that the block exists.
const blockOffsetToBlock$ = (apolloClient) => (offset) => zenToRx(apolloClient
    .subscribe({
    query: TXS_FROM_BLOCK_SUB,
    variables: {
        offset
    },
    fetchPolicy: 'no-cache'
})
    .map(maybeThrowGraphQLErrors)
    .map((fetchResult) => {
    const blocks = fetchResult.data.blocks;
    return {
        hash: blocks.hash,
        height: blocks.height,
        transactions: blocks.transactions
            .filter((tx) => 'identifiers' in tx)
            .map(tx => ({
            hash: tx.hash,
            identifiers: tx.identifiers,
            contractActions: tx.contractActions
        }))
    };
}));
const transactionIdToTransaction$ = (apolloClient) => (identifier) => zenToRx(apolloClient
    .watchQuery({
    query: TX_ID_QUERY,
    variables: {
        offset: { identifier }
    },
    pollInterval: DEFAULT_POLL_INTERVAL,
    fetchPolicy: 'no-cache',
    initialFetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache'
})
    .map(maybeThrowErrors)
    .filter((maybeQueryResult) => maybeQueryResult.data.transactions.length !== 0)
    .map((maybeQueryResult) => ({
    height: maybeQueryResult.data.transactions[0].block.height
}))).pipe(Rx.concatMap(blockOffsetToBlock$(apolloClient)), Rx.concatMap(({ transactions }) => Rx.from(transactions)));
const transactionToContractState$ = (transactionId) => ({ identifiers, contractActions }) => Rx.zip(identifiers, contractActions).pipe(Rx.skipWhile((pair) => pair[0] !== transactionId), Rx.map((pair) => deserializeContractState(pair[1].state)));
const toTxStatus = (transactionResult) => {
    const result = transactionResult.status;
    const map = {
        'FAILURE': FailEntirely,
        'PARTIAL_SUCCESS': FailFallible,
        'SUCCESS': SucceedEntirely
    };
    if (result === 'FAILURE' || result === 'PARTIAL_SUCCESS' || result === 'SUCCESS') {
        return map[result];
    }
    throw new Error(`Unexpected 'status' value ${result}`);
};
const toSegmentStatus = (success) => success ? SegmentSuccess : SegmentFail;
const toSegmentStatusMap = (transactionResult) => {
    if (transactionResult.status !== 'PARTIAL_SUCCESS') {
        return undefined;
    }
    if (!transactionResult.segments) {
        return undefined;
    }
    return new Map(transactionResult.segments.map((segment) => [segment.id, toSegmentStatus(segment.success)]));
};
const transformIndexerUtxoToUnshieldedUtxo = (indexerUtxo) => ({
    owner: indexerUtxo.owner,
    intentHash: indexerUtxo.intentHash,
    tokenType: indexerUtxo.tokenType,
    value: BigInt(indexerUtxo.value)
});
const toUnshieldedUtxos = (createdUtxo, spentUtxo) => ({
    created: createdUtxo.map(transformIndexerUtxoToUnshieldedUtxo),
    spent: spentUtxo.map(transformIndexerUtxoToUnshieldedUtxo)
});
const transformContractBalanceToUnshieldedBalance = (contractBalance) => ({
    balance: BigInt(contractBalance.amount),
    tokenType: contractBalance.tokenType
});
const toUnshieldedBalances = (contractBalances) => contractBalances.map(transformContractBalanceToUnshieldedBalance);
const blockToContractState$ = (contractAddress) => (block) => Rx.from(block.transactions).pipe(Rx.concatMap(({ contractActions }) => Rx.from(contractActions)), Rx.filter((call) => call.address === contractAddress), Rx.map((call) => deserializeContractState(call.state)));
const contractAddressToLatestBlockOffset$ = (apolloClient) => (contractAddress) => zenToRx(apolloClient
    .watchQuery({
    query: LATEST_CONTRACT_TX_BLOCK_HEIGHT_QUERY,
    variables: {
        address: contractAddress
    },
    pollInterval: DEFAULT_POLL_INTERVAL,
    fetchPolicy: 'no-cache',
    initialFetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache'
})
    .map(maybeThrowErrors)
    .filter((maybeQueryResult) => maybeQueryResult.data.contractAction !== null)
    .map((queryResult) => {
    const contract = queryResult.data.contractAction;
    return contract.transaction.block.height;
})).pipe(Rx.take(1), Rx.map((height) => ({ height })));
// Assumes block already exists
const blockOffsetToContractState$ = (apolloClient) => (contractAddress) => (offset) => zenToRx(apolloClient
    .subscribe({
    query: CONTRACT_STATE_SUB,
    variables: {
        address: contractAddress,
        offset
    },
    fetchPolicy: 'no-cache'
})
    .map(maybeThrowGraphQLErrors)
    .map((queryResult) => queryResult.data.contractActions.state)
    .map(deserializeContractState));
const waitForContractToAppear = (apolloClient) => (contractAddress) => (offset) => zenToRx(apolloClient
    .watchQuery({
    query: CONTRACT_STATE_QUERY,
    variables: {
        address: contractAddress,
        offset
    },
    pollInterval: DEFAULT_POLL_INTERVAL,
    fetchPolicy: 'no-cache',
    initialFetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache'
})
    .map(maybeThrowErrors)
    .filter((maybeQueryResult) => maybeQueryResult.data.contractAction !== null)
    .map((queryResult) => queryResult.data.contractAction.state)).pipe(Rx.take(1));
const waitForBlockToAppear = (apolloClient) => (offset) => zenToRx(apolloClient
    .watchQuery({
    query: BLOCK_QUERY,
    variables: {
        offset
    },
    pollInterval: DEFAULT_POLL_INTERVAL,
    fetchPolicy: 'no-cache',
    initialFetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache'
})
    .map(maybeThrowErrors)
    .filter((fetchResult) => fetchResult.data.block !== null)).pipe(Rx.take(1));
const waitForUnshieldedBalancesToAppear = (apolloClient) => (contractAddress) => zenToRx(apolloClient
    .watchQuery({
    query: UNSHIELDED_BALANCE_QUERY,
    variables: {
        address: contractAddress
    },
    pollInterval: DEFAULT_POLL_INTERVAL,
    fetchPolicy: 'no-cache',
    initialFetchPolicy: 'no-cache',
    nextFetchPolicy: 'no-cache'
})
    .map(maybeThrowErrors)
    .filter((maybeQueryResult) => maybeQueryResult.data.contractAction !== null)
    .map((queryResult) => {
    const contractAction = queryResult.data.contractAction;
    if ('unshieldedBalances' in contractAction) {
        return contractAction.unshieldedBalances;
    }
    if ('deploy' in contractAction) {
        return contractAction.deploy.unshieldedBalances;
    }
    return [];
})).pipe(Rx.take(1));
const blockOffsetToUnshieldedBalances$ = (apolloClient) => (contractAddress) => (offset) => zenToRx(apolloClient
    .subscribe({
    query: UNSHIELDED_BALANCE_SUB,
    variables: {
        address: contractAddress,
        offset
    },
    fetchPolicy: 'no-cache'
})
    .map(maybeThrowGraphQLErrors)
    .map((queryResult) => {
    const contractAction = queryResult.data.contractActions;
    if ('unshieldedBalances' in contractAction) {
        return contractAction.unshieldedBalances;
    }
    if ('deploy' in contractAction) {
        return contractAction.deploy.unshieldedBalances;
    }
    return [];
})
    .map(toUnshieldedBalances));
const indexerPublicDataProviderInternal = (queryURL, subscriptionURL, webSocketImpl = ws.WebSocket) => {
    const queryURLObj = new URL(queryURL);
    if (queryURLObj.protocol !== 'http:' && queryURLObj.protocol !== 'https:') {
        throw new InvalidProtocolSchemeError(queryURLObj.protocol, ['http:', 'https:']);
    }
    const subscriptionURLObj = new URL(subscriptionURL);
    if (subscriptionURLObj.protocol !== 'ws:' && subscriptionURLObj.protocol !== 'wss:') {
        throw new InvalidProtocolSchemeError(subscriptionURLObj.protocol, ['ws:', 'wss:']);
    }
    // Construct the Apollo client.
    const link = createHttpLink({ fetch, uri: queryURL });
    // Retry link with exponential backoff.
    const retryLink = new RetryLink({
        delay: {
            initial: 1000,
            max: 10000,
            jitter: true
        },
        attempts: {
            max: 5
        }
    });
    // Combine the retry link with the HTTP link to form the final link.
    const apolloLink = from([retryLink, link]);
    const apolloClient = new ApolloClient({
        link: split(({ query }) => {
            const definition = getMainDefinition(query);
            return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
        }, new GraphQLWsLink(createClient({ url: subscriptionURL, webSocketImpl })), apolloLink),
        cache: new InMemoryCache()
    });
    return {
        async queryContractState(address, config) {
            let offset;
            if (config) {
                offset = {
                    blockOffset: config.type === 'blockHeight' ? { height: config.blockHeight } : { hash: config.blockHash }
                };
            }
            else {
                offset = null;
            }
            const maybeContractState = await apolloClient
                .query({
                query: CONTRACT_STATE_QUERY,
                variables: {
                    address,
                    offset
                },
                fetchPolicy: 'no-cache'
            })
                .then(maybeThrowErrors)
                .then((queryResult) => queryResult.data?.contractAction?.state ?? null);
            return maybeContractState ? deserializeContractState(maybeContractState) : null;
        },
        async queryZSwapAndContractState(address, config) {
            let offset;
            if (config) {
                offset = {
                    blockOffset: config.type === 'blockHeight' ? { height: config.blockHeight } : { hash: config.blockHash }
                };
            }
            else {
                offset = null;
            }
            const maybeContractStates = await apolloClient
                .query({
                query: CONTRACT_AND_ZSWAP_STATE_QUERY,
                variables: {
                    address,
                    offset
                },
                fetchPolicy: 'no-cache'
            })
                .then(maybeThrowErrors)
                .then((queryResult) => queryResult.data.contractAction);
            return maybeContractStates
                ? [deserializeZswapState(maybeContractStates.zswapState), deserializeContractState(maybeContractStates.state)]
                : null;
        },
        async queryUnshieldedBalances(address, config) {
            let offset;
            if (config) {
                offset = {
                    blockOffset: config.type === 'blockHeight' ? { height: config.blockHeight } : { hash: config.blockHash }
                };
            }
            else {
                offset = null;
            }
            const maybeUnshieldedBalances = await apolloClient
                .query({
                query: QUERY_UNSHIELDED_BALANCES_WITH_OFFSET,
                variables: {
                    address,
                    offset
                },
                fetchPolicy: 'no-cache'
            })
                .then(maybeThrowErrors)
                .then((queryResult) => {
                const contractAction = queryResult.data.contractAction;
                if (!contractAction) {
                    return null;
                }
                if ('unshieldedBalances' in contractAction) {
                    return contractAction.unshieldedBalances;
                }
                if ('deploy' in contractAction) {
                    return contractAction.deploy.unshieldedBalances;
                }
                return [];
            });
            return maybeUnshieldedBalances ? toUnshieldedBalances(maybeUnshieldedBalances) : null;
        },
        async queryDeployContractState(contractAddress) {
            return apolloClient
                .query({
                query: DEPLOY_CONTRACT_STATE_TX_QUERY,
                variables: {
                    address: contractAddress
                },
                fetchPolicy: 'no-cache'
            })
                .then((queryResult) => {
                if (queryResult.data.contractAction) {
                    const contract = queryResult.data.contractAction;
                    return 'deploy' in contract
                        ? contract.deploy.transaction.contractActions.find(({ address }) => address === contractAddress).state
                        : contract.state;
                }
                return null;
            })
                .then((maybeContractState) => (maybeContractState ? deserializeContractState(maybeContractState) : null));
        },
        async watchForContractState(contractAddress) {
            return Rx.firstValueFrom(waitForContractToAppear(apolloClient)(contractAddress)(null).pipe(Rx.map(deserializeContractState)));
        },
        async watchForUnshieldedBalances(contractAddress) {
            return Rx.firstValueFrom(waitForUnshieldedBalancesToAppear(apolloClient)(contractAddress).pipe(Rx.map(toUnshieldedBalances)));
        },
        async watchForDeployTxData(contractAddress) {
            return Rx.firstValueFrom(zenToRx(apolloClient
                .watchQuery({
                query: DEPLOY_TX_QUERY,
                variables: {
                    address: contractAddress
                },
                pollInterval: DEFAULT_POLL_INTERVAL,
                fetchPolicy: 'no-cache',
                initialFetchPolicy: 'no-cache',
                nextFetchPolicy: 'no-cache'
            })
                .filter((maybeQueryResult) => maybeQueryResult.data.contractAction !== null)
                .map(maybeThrowErrors)
                .map((queryResults) => {
                const contract = queryResults.data.contractAction;
                return 'deploy' in contract ? contract.deploy.transaction : contract.transaction;
            })
                .filter(isRegularTransaction)
                .map((transaction) => ({
                tx: deserializeTransaction(transaction.raw),
                status: toTxStatus(transaction.transactionResult),
                txId: transaction.identifiers[transaction.contractActions.findIndex(({ address }) => address === contractAddress)],
                identifiers: transaction.identifiers,
                txHash: transaction.hash,
                blockHeight: transaction.block.height,
                blockHash: transaction.block.hash,
                blockTimestamp: transaction.block.timestamp,
                blockAuthor: transaction.block.author,
                segmentStatusMap: toSegmentStatusMap(transaction.transactionResult),
                unshielded: toUnshieldedUtxos(transaction.unshieldedCreatedOutputs, transaction.unshieldedSpentOutputs),
                indexerId: transaction.id,
                protocolVersion: transaction.protocolVersion,
                fees: {
                    estimatedFees: transaction.fees.estimatedFees,
                    paidFees: transaction.fees.paidFees
                },
            }))));
        },
        async watchForTxData(txId) {
            return Rx.firstValueFrom(zenToRx(apolloClient
                .watchQuery({
                query: TX_ID_QUERY,
                variables: { offset: { identifier: txId } },
                pollInterval: DEFAULT_POLL_INTERVAL,
                fetchPolicy: 'no-cache',
                initialFetchPolicy: 'no-cache',
                nextFetchPolicy: 'no-cache'
            })
                .map(maybeThrowErrors)
                .filter((maybeQueryResult) => maybeQueryResult.data.transactions.length !== 0)
                .map((queryResult) => queryResult.data.transactions[0])
                .filter(isRegularTransaction)
                .map((transaction) => ({
                tx: deserializeTransaction(transaction.raw),
                status: toTxStatus(transaction.transactionResult),
                txId,
                txHash: transaction.hash,
                identifiers: transaction.identifiers,
                blockHeight: transaction.block.height,
                blockHash: transaction.block.hash,
                segmentStatusMap: toSegmentStatusMap(transaction.transactionResult),
                unshielded: toUnshieldedUtxos(transaction.unshieldedCreatedOutputs, transaction.unshieldedSpentOutputs),
                blockTimestamp: transaction.block.timestamp,
                blockAuthor: transaction.block.author,
                indexerId: transaction.id,
                protocolVersion: transaction.protocolVersion,
                fees: {
                    paidFees: transaction.fees.paidFees,
                    estimatedFees: transaction.fees.estimatedFees
                }
            }))));
        },
        contractStateObservable(contractAddress, config = { type: 'latest' }) {
            if (config.type === 'txId') {
                const contractStates = transactionIdToTransaction$(apolloClient)(config.txId).pipe(Rx.filter(isRegularTransaction), Rx.concatMap(transactionToContractState$(config.txId)));
                return (config.inclusive ?? true) ? contractStates : contractStates.pipe(Rx.skip(1));
            }
            if (config.type === 'latest') {
                return contractAddressToLatestBlockOffset$(apolloClient)(contractAddress).pipe(Rx.concatMap(blockOffsetToBlock$(apolloClient)), Rx.concatMap(blockToContractState$(contractAddress)));
            }
            if (config.type === 'all') {
                return waitForContractToAppear(apolloClient)(contractAddress)(null).pipe(Rx.concatMap(() => blockOffsetToContractState$(apolloClient)(contractAddress)(null)));
            }
            const offset = config.type === 'blockHash' ? { hash: config.blockHash } : { height: config.blockHeight };
            const blocks = waitForBlockToAppear(apolloClient)(offset).pipe(Rx.filter(isRegularTransaction), Rx.concatMap(() => blockOffsetToBlock$(apolloClient)(offset)));
            const maybeShortenedBlocks = config.type === 'blockHeight' || config.type === 'blockHash'
                ? Rx.iif(() => config.inclusive ?? true, blocks, blocks.pipe(Rx.skip(1)))
                : blocks;
            return maybeShortenedBlocks.pipe(Rx.concatMap(blockToContractState$(contractAddress)));
        },
        unshieldedBalancesObservable(contractAddress, config = { type: 'latest' }) {
            if (config.type === 'txId') {
                throw new Error('txId configuration not supported for unshielded balances observable');
            }
            if (config.type === 'latest') {
                return contractAddressToLatestBlockOffset$(apolloClient)(contractAddress).pipe(Rx.concatMap(blockOffsetToUnshieldedBalances$(apolloClient)(contractAddress)));
            }
            if (config.type === 'all') {
                return waitForUnshieldedBalancesToAppear(apolloClient)(contractAddress).pipe(Rx.concatMap(() => blockOffsetToUnshieldedBalances$(apolloClient)(contractAddress)(null)));
            }
            const offset = config.type === 'blockHash' ? { hash: config.blockHash } : { height: config.blockHeight };
            const balances = waitForBlockToAppear(apolloClient)(offset).pipe(Rx.concatMap(() => blockOffsetToUnshieldedBalances$(apolloClient)(contractAddress)(offset)));
            return config.type === 'blockHeight' || config.type === 'blockHash'
                ? Rx.iif(() => config.inclusive ?? true, balances, balances.pipe(Rx.skip(1)))
                : balances;
        }
    };
};
/**
 * Constructs a {@link PublicDataProvider} based on an {@link ApolloClient}.
 *
 * @param queryURL The URL of a GraphQL server query endpoint.
 * @param subscriptionURL The URL of a GraphQL server subscription (websocket) endpoint.
 * @param webSocketImpl An optional websocket implementation for the Apollo client to use.
 *
 * TODO: Re-examine caching when 'ContractCall' and 'ContractDeploy' have transaction identifiers included.
 */
const indexerPublicDataProvider = (queryURL, subscriptionURL, webSocketImpl = ws.WebSocket) => {
    /**
     * This current object is a wrapper around the real implementation of the indexer client constructed
     * below. This wrapper just asserts that the input contract addresses are valid, and prepends the hex
     * representation of the network ID to all input contract addresses to work around a discrepancy
     * as of ledger 3.0.0 between the contract address representation on the indexer (with network ID)
     * and the address representation in the ledger WASM API (without network ID).
     */
    const publicDataProvider = indexerPublicDataProviderInternal(queryURL, subscriptionURL, webSocketImpl);
    return {
        contractStateObservable(contractAddress, config) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.contractStateObservable(contractAddress, config);
        },
        queryContractState(contractAddress, config) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.queryContractState(contractAddress, config);
        },
        queryDeployContractState(contractAddress) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.queryDeployContractState(contractAddress);
        },
        queryZSwapAndContractState(contractAddress, config) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.queryZSwapAndContractState(contractAddress, config);
        },
        queryUnshieldedBalances(contractAddress, config) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.queryUnshieldedBalances(contractAddress, config);
        },
        watchForContractState(contractAddress) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.watchForContractState(contractAddress);
        },
        watchForUnshieldedBalances(contractAddress) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.watchForUnshieldedBalances(contractAddress);
        },
        watchForDeployTxData(contractAddress) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.watchForDeployTxData(contractAddress);
        },
        watchForTxData(txId) {
            return publicDataProvider.watchForTxData(txId);
        },
        unshieldedBalancesObservable(contractAddress, config) {
            assertIsContractAddress(contractAddress);
            return publicDataProvider.unshieldedBalancesObservable(contractAddress, config);
        }
    };
};

export { IndexerFormattedError, indexerPublicDataProvider, toUnshieldedBalances, toUnshieldedUtxos };
