// @midnight-ntwrk/wallet-sdk-capabilities/dist/submission/submissionService.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Data, Deferred, Effect, Encoding, Exit, pipe, Scope } from 'effect';
import { NodeClient, PolkadotNodeClient, SubmissionEvent as SubmissionEventImported, } from '@midnight-ntwrk/wallet-sdk-node-client/effect';
import { SerializedTransaction } from '@midnight-ntwrk/wallet-sdk-abstractions';
export const SubmissionEvent = SubmissionEventImported;
export class SubmissionError extends Data.TaggedError('SubmissionError') {
}
export const makeDefaultSubmissionServiceEffect = (config) => {
    const scopeAndClientDeferred = Deferred.make().pipe(Effect.runSync);
    const makeScopeAndClient = Effect.gen(function* () {
        const scope = yield* Scope.make();
        const client = yield* PolkadotNodeClient.make({
            nodeURL: config.relayURL,
        }).pipe(Effect.provideService(Scope.Scope, scope));
        return { scope, client };
    });
    void pipe(scopeAndClientDeferred, Deferred.complete(makeScopeAndClient), Effect.runPromise);
    const submit = (transaction, waitForStatus = 'InBlock') => {
        return pipe(NodeClient.sendMidnightTransactionAndWait(SerializedTransaction.from(transaction), waitForStatus), Effect.provideServiceEffect(NodeClient.NodeClient, pipe(scopeAndClientDeferred, Deferred.await, Effect.map(({ client }) => client))), Effect.mapError((err) => new SubmissionError({ message: 'Transaction submission error', cause: err })));
    };
    return {
        submitTransaction: submit,
        close() {
            return pipe(scopeAndClientDeferred, Deferred.await, Effect.flatMap(({ scope }) => Scope.close(scope, Exit.void)), Effect.ignoreLogged);
        },
    };
};
export const makeDefaultSubmissionService = (config) => {
    const effectService = makeDefaultSubmissionServiceEffect(config);
    const submit = (transaction, waitForStatus = 'InBlock') => effectService.submitTransaction(transaction, waitForStatus).pipe(Effect.runPromise);
    return {
        submitTransaction: submit,
        close: () => effectService.close().pipe(Effect.runPromise),
    };
};
export const makeSimulatorSubmissionService = (waitForStatus = 'InBlock') => (config) => {
    const submit = (transaction) => {
        return config.simulator.submitTransaction(transaction).pipe(Effect.map((output) => {
            const serializedTransaction = SerializedTransaction.from(transaction);
            const fakeTxHash = Encoding.encodeHex(serializedTransaction).slice(0, 64);
            // Let's mimic node's client behavior here
            switch (waitForStatus) {
                case 'Submitted':
                    return SubmissionEvent.Submitted({
                        tx: serializedTransaction,
                        txHash: fakeTxHash,
                    });
                case 'InBlock':
                    return SubmissionEvent.InBlock({
                        tx: serializedTransaction,
                        blockHash: output.blockHash,
                        blockHeight: output.blockNumber,
                        txHash: fakeTxHash,
                    });
                case 'Finalized':
                    return SubmissionEvent.Finalized({
                        tx: serializedTransaction,
                        blockHash: output.blockHash,
                        blockHeight: output.blockNumber,
                        txHash: fakeTxHash,
                    });
            }
        }), Effect.mapError((err) => new SubmissionError({ message: 'Transaction submission error', cause: err })));
    };
    return {
        submitTransaction: submit,
        close: () => Effect.void,
    };
};
