// @midnight-ntwrk/wallet-sdk-runtime/dist/testing/utils.js
import { Chunk } from 'effect';
import { reduce } from 'rxjs';
/**
 * Utility function that takes state values from an RxJS observable until it completes or errors.
 *
 * @param observable The RxJS observable from which state values should be read.
 * @param onErrCallback An optional callback to invoke if an error is encountered reading a state value.
 * @returns A `Promise` that resolves with an array of state values that were received before encountering
 * any error.
 *
 * @internal
 */
export const toProtocolStateArray = (observable, onErrCallback) => new Promise((resolve) => {
    const receivedStates = [];
    observable.subscribe({
        next(value) {
            receivedStates.push(value);
        },
        complete() {
            resolve(receivedStates);
        },
        error(err) {
            onErrCallback?.call(undefined, err);
            resolve(receivedStates);
        },
    });
});
export const reduceToChunk = () => reduce((chunk, value) => Chunk.append(chunk, value), Chunk.empty());
export const isRange = (values) => {
    const firstDropped = Chunk.drop(values, 1);
    const lastDropped = Chunk.dropRight(values, 1);
    return Chunk.zip(lastDropped, firstDropped).pipe(Chunk.every(([l, r]) => r == l + 1));
};
