// @midnight-ntwrk/wallet-sdk-capabilities/dist/proving/provingService.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import * as ledger from '@midnight-ntwrk/ledger-v7';
import { HttpProverClient, WasmProver } from '@midnight-ntwrk/wallet-sdk-prover-client/effect';
import { ClientError, ServerError } from '@midnight-ntwrk/wallet-sdk-utilities/networking';
import { Data, Effect, pipe } from 'effect';
export class ProvingError extends Data.TaggedError('Wallet.Proving') {
}
const wrapEffectService = (effectService) => ({
    prove: (transaction) => Effect.runPromise(effectService.prove(transaction)),
});
export const fromProvingProviderEffect = (provider) => {
    return {
        prove(transaction) {
            return pipe(provider, Effect.flatMap((provider) => Effect.tryPromise({
                try: () => transaction.prove(provider, ledger.CostModel.initialCostModel()),
                catch: (error) => error instanceof ClientError || error instanceof ServerError
                    ? error
                    : new ClientError({ message: 'Failed to prove transaction', cause: error }),
            })), Effect.catchAll((error) => Effect.fail(new ProvingError({
                message: error.message,
                cause: error,
            }))));
        },
    };
};
export const fromProvingProvider = (provider) => {
    return fromProvingProviderEffect(Effect.succeed(provider));
};
export const makeServerProvingServiceEffect = (configuration) => {
    return pipe(HttpProverClient.create({
        url: configuration.provingServerUrl,
    }), Effect.map((client) => client.asProvingProvider()), fromProvingProviderEffect);
};
export const makeWasmProvingServiceEffect = (configuration) => {
    return pipe(WasmProver.create({
        keyMaterialProvider: configuration?.keyMaterialProvider ?? WasmProver.makeDefaultKeyMaterialProvider(),
    }), Effect.map((prover) => prover.asProvingProvider()), fromProvingProviderEffect);
};
export const makeSimulatorProvingServiceEffect = () => {
    return {
        prove(transaction) {
            return Effect.succeed(transaction.eraseProofs());
        },
    };
};
export const makeDefaultProvingServiceEffect = (configuration) => makeServerProvingServiceEffect(configuration);
export const makeDefaultProvingService = (configuration) => wrapEffectService(makeDefaultProvingServiceEffect(configuration));
export const makeServerProvingService = (configuration) => wrapEffectService(makeServerProvingServiceEffect(configuration));
export const makeWasmProvingService = (configuration) => wrapEffectService(makeWasmProvingServiceEffect(configuration));
export const makeSimulatorProvingService = () => wrapEffectService(makeSimulatorProvingServiceEffect());
