// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/CoreWallet.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { ProtocolVersion } from '@midnight-ntwrk/wallet-sdk-abstractions';
import { createSyncProgress } from './SyncProgress.js';
import { UnshieldedState } from './UnshieldedState.js';
import { Either, Array as Arr, pipe } from 'effect';
import { ApplyTransactionError, RollbackUtxoError, SpendUtxoError } from './WalletError.js';
export const CoreWallet = {
    init(publicKey, networkId) {
        return {
            state: UnshieldedState.empty(),
            publicKey,
            protocolVersion: ProtocolVersion.MinSupportedVersion,
            progress: createSyncProgress(),
            networkId,
        };
    },
    restore(state, publicKey, syncProgress, protocolVersion, networkId) {
        return {
            state,
            publicKey,
            protocolVersion,
            progress: createSyncProgress(syncProgress),
            networkId,
        };
    },
    updateProgress(wallet, { appliedId, highestTransactionId, isConnected }) {
        const progress = createSyncProgress({
            appliedId: appliedId ?? wallet.progress.appliedId,
            highestTransactionId: highestTransactionId ?? wallet.progress.highestTransactionId,
            isConnected: isConnected ?? wallet.progress.isConnected,
        });
        return { ...wallet, progress };
    },
    applyUpdate(coreWallet, update) {
        return UnshieldedState.applyUpdate(coreWallet.state, update).pipe(Either.map((state) => ({ ...coreWallet, state })), Either.mapLeft((error) => new ApplyTransactionError(error)));
    },
    applyFailedUpdate(coreWallet, update) {
        return UnshieldedState.applyFailedUpdate(coreWallet.state, update).pipe(Either.map((state) => ({ ...coreWallet, state })), Either.mapLeft((error) => new ApplyTransactionError(error)));
    },
    rollbackUtxo(coreWallet, utxo) {
        return UnshieldedState.rollbackSpendByUtxo(coreWallet.state, utxo).pipe(Either.map((state) => ({ ...coreWallet, state })), Either.mapLeft((error) => new RollbackUtxoError(error)));
    },
    spend(coreWallet, utxo) {
        return UnshieldedState.spendByUtxo(coreWallet.state, utxo).pipe(Either.map((state) => ({ ...coreWallet, state })), Either.mapLeft((error) => new SpendUtxoError(error)));
    },
    spendUtxos(wallet, utxos) {
        return pipe(utxos, Arr.reduce(Either.right([[], wallet.state]), (acc, utxoToSpend) => acc.pipe(Either.flatMap(([accUtxos, state]) => UnshieldedState.spendByUtxo(state, utxoToSpend).pipe(Either.map((nextState) => [accUtxos.concat([utxoToSpend]), nextState]), Either.mapLeft((error) => new SpendUtxoError(error)))))), Either.map(([spentUtxos, state]) => [spentUtxos, { ...wallet, state }]));
    },
};
