// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/Sync.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Stream, Schema, pipe, Either } from 'effect';
import { CoreWallet } from './CoreWallet.js';
import { UnshieldedTransactions } from '@midnight-ntwrk/wallet-sdk-indexer-client';
import { WsSubscriptionClient, ConnectionHelper } from '@midnight-ntwrk/wallet-sdk-indexer-client/effect';
import { SyncWalletError } from './WalletError.js';
import { WsURL } from '@midnight-ntwrk/wallet-sdk-utilities/networking';
import { EitherOps } from '@midnight-ntwrk/wallet-sdk-utilities';
import { WalletSyncUpdateSchema } from './SyncSchema.js';
export const makeDefaultSyncService = (config) => {
    return {
        updates: (state) => {
            const { indexerClientConnection } = config;
            const webSocketUrlResult = ConnectionHelper.createWebSocketUrl(indexerClientConnection.indexerHttpUrl, indexerClientConnection.indexerWsUrl);
            if (Either.isLeft(webSocketUrlResult)) {
                return Stream.fail(new SyncWalletError(new Error(`Could not derive WebSocket URL from indexer HTTP URL: ${webSocketUrlResult.left.message}`)));
            }
            const indexerWsUrlResult = WsURL.make(webSocketUrlResult.right);
            if (Either.isLeft(indexerWsUrlResult)) {
                return Stream.fail(new SyncWalletError(new Error(`Invalid indexer WS URL: ${indexerWsUrlResult.left.message}`)));
            }
            const indexerWsUrl = indexerWsUrlResult.right;
            const { appliedId } = state.progress;
            const { address } = state.publicKey;
            return pipe(UnshieldedTransactions.run({ address, transactionId: Number(appliedId) }), Stream.provideLayer(WsSubscriptionClient.layer({ url: indexerWsUrl, keepAlive: indexerClientConnection.keepAlive })), Stream.mapError((error) => new SyncWalletError(error)), Stream.mapEffect((subscription) => {
                const { unshieldedTransactions } = subscription;
                return pipe(Schema.decodeUnknownEither(WalletSyncUpdateSchema)(unshieldedTransactions), Either.mapLeft((err) => new SyncWalletError(err)), EitherOps.toEffect);
            }));
        },
    };
};
export const makeDefaultSyncCapability = (_config, getContext) => {
    return {
        applyUpdate: (state, update) => {
            if (update.type === 'UnshieldedTransactionsProgress') {
                return Either.right(CoreWallet.updateProgress(state, {
                    highestTransactionId: BigInt(update.highestTransactionId),
                    isConnected: true,
                }));
            }
            else {
                const updatePayload = {
                    createdUtxos: update.createdUtxos,
                    spentUtxos: update.spentUtxos,
                    status: update.status,
                };
                const stateAfterApplyingUpdate = update.status === 'FAILURE'
                    ? CoreWallet.applyFailedUpdate(state, updatePayload)
                    : CoreWallet.applyUpdate(state, updatePayload);
                return stateAfterApplyingUpdate.pipe(Either.map((wallet) => {
                    const stateAfterUpdatingProgress = CoreWallet.updateProgress(wallet, {
                        appliedId: BigInt(update.transaction.id),
                    });
                    const { transactionHistoryService } = getContext();
                    void transactionHistoryService.create(update);
                    return stateAfterUpdatingProgress;
                }));
            }
        },
    };
};
export const makeSimulatorSyncService = (config) => {
    return {
        updates: (_state) => config.simulator.state$.pipe(Stream.map((state) => ({ update: state }))),
    };
};
export const makeSimulatorSyncCapability = () => {
    return {
        applyUpdate: (state, _update) => {
            return Either.right(state);
            // return CoreWallet.replayEvents(state, secretKeys, events);
        },
    };
};
