// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/v1/TransactionHistory.js
const convertUpdateToEntry = ({ transaction, createdUtxos, spentUtxos, status, }) => {
    return {
        id: transaction.id,
        hash: transaction.hash,
        protocolVersion: transaction.protocolVersion,
        identifiers: transaction.identifiers ? transaction.identifiers : [],
        status,
        timestamp: transaction.block?.timestamp ?? null,
        fees: transaction.fees?.paidFees ?? null,
        createdUtxos: createdUtxos.map(({ utxo }) => ({
            value: utxo.value,
            owner: utxo.owner,
            tokenType: utxo.type,
            intentHash: utxo.intentHash,
            outputIndex: utxo.outputNo,
        })),
        spentUtxos: spentUtxos.map(({ utxo }) => ({
            value: utxo.value,
            owner: utxo.owner,
            tokenType: utxo.type,
            intentHash: utxo.intentHash,
            outputIndex: utxo.outputNo,
        })),
    };
};
export const makeDefaultTransactionHistoryService = (config, _getContext) => {
    const { txHistoryStorage } = config;
    return {
        create: async (update) => {
            const entry = convertUpdateToEntry(update);
            await txHistoryStorage.create(entry);
        },
        get: async (hash) => {
            return await txHistoryStorage.get(hash);
        },
        getAll: () => {
            return txHistoryStorage.getAll();
        },
        delete: async (hash) => {
            return txHistoryStorage.delete(hash);
        },
    };
};
