// @midnight-ntwrk/midnight-js-types/dist/index.cjs
'use strict';

var effect = require('@midnight-ntwrk/compact-js/effect');
var platformJs = require('@midnight-ntwrk/platform-js');
var Configuration = require('@midnight-ntwrk/platform-js/effect/Configuration');
var effect$1 = require('effect');
var ledgerV8 = require('@midnight-ntwrk/ledger-v8');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var Configuration__namespace = /*#__PURE__*/_interopNamespaceDefault(Configuration);

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Creates a ZK configuration reader by adapting a given {@link ZKConfigProvider}.
 *
 * @param zkConfigProvider The {@link ZKConfigProvider} that is to be adapted.
 * @returns A {@link ZKConfiguration.ZKConfiguration.Reader | ZKConfiguration.Reader} that reads from
 * `zkConfigProvider`.
 *
 * @internal
 */
const makeAdaptedReader = (zkConfigProvider) => (compiledContract) => effect$1.Effect.gen(function* () {
    // TODO: Consider implementing the logic used in Compact.js (look at the contract manifest to determine
    // if the circuit is verifiable). See PM-21376.
    const getVerifierKey = (provableCircuitId) => effect$1.Effect.tryPromise({
        try: () => zkConfigProvider.getVerifierKey(provableCircuitId).then((verifierKey) => effect$1.Option.some(effect.Contract.VerifierKey(verifierKey))),
        catch: (err) => effect.ZKConfigurationReadError.make(compiledContract.tag, provableCircuitId, 'verifier-key', err)
    });
    return {
        getVerifierKey,
        getVerifierKeys: (provableCircuitIds) => effect$1.Effect.forEach(provableCircuitIds, (provableCircuitId) => getVerifierKey(provableCircuitId).pipe(effect$1.Effect.map((verifierKey) => [provableCircuitId, verifierKey])), { concurrency: 'unbounded', discard: false })
    };
});
const makeAdaptedRuntimeLayer = (zkConfigProvider, configMap) => effect$1.Layer.mergeAll(effect$1.Layer.succeed(effect.ZKConfiguration.ZKConfiguration, effect.ZKConfiguration.ZKConfiguration.of({
    createReader: makeAdaptedReader(zkConfigProvider)
})), Configuration__namespace.layer).pipe(effect$1.Layer.provide(effect$1.Layer.setConfigProvider(effect$1.ConfigProvider.fromMap(configMap, { pathDelim: '_' }).pipe(effect$1.ConfigProvider.constantCase))));
/**
 * Constructs an Effect managed runtime configured to execute contract executables.
 *
 * @param zkConfigProvider The {@link ZKConfigProvider} that is to be adapted.
 * @param options Values that will be mapped into and made available within the constructed runtime.
 * @returns An Effect {@link ManagedRuntime} that can be used to execute {@link ContractExecutable} instances.
 */
const makeContractExecutableRuntime = (zkConfigProvider, options) => {
    let config = [['KEYS_COIN_PUBLIC', options.coinPublicKey]];
    if (options.signingKey) {
        config = config.concat([['KEYS_SIGNING', options.signingKey]]);
    }
    return effect.ContractExecutableRuntime.make(makeAdaptedRuntimeLayer(zkConfigProvider, new Map(config)));
};
/**
 * Unwraps an Effect `Exit` instance, returning its value if it is successful, or throwing the error contained
 * within it.
 *
 * @param exit The source Effect `Exit` instance.
 * @returns The value from `exit` if it is successful, otherwise throws the error contained within it.
 */
const exitResultOrError = (exit) => effect$1.Exit.match(exit, {
    onSuccess: (a) => a,
    onFailure: (cause) => {
        if (effect$1.Cause.isFailType(cause))
            throw cause.error;
        throw new Error(`Unexpected error: ${effect$1.Cause.pretty(cause)}`);
    }
});
/**
 * Wraps an object into an `Option.some`.
 *
 * @param obj The value that should be wrapped into an `Option`.
 * @returns An `Option.some` for `obj`.
 */
const asEffectOption = (obj) => {
    return effect$1.Option.some(obj);
};
/**
 * Constructs a branded contract address from a given string value.
 *
 * @param address A string value representing a contract address.
 * @returns A {@link ContractAddress.ContractAddress | ContractAddress} constructed from `address`.
 */
const asContractAddress = (address) => platformJs.ContractAddress.ContractAddress(address);

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * An error describing an invalid protocol scheme.
 */
class InvalidProtocolSchemeError extends Error {
    invalidScheme;
    allowableSchemes;
    /**
     * @param invalidScheme The invalid scheme.
     * @param allowableSchemes The valid schemes that are allowed.
     */
    constructor(invalidScheme, allowableSchemes) {
        super(`Invalid protocol scheme: '${invalidScheme}'. Allowable schemes are one of: ${allowableSchemes.join(',')}`);
        this.invalidScheme = invalidScheme;
        this.allowableSchemes = allowableSchemes;
    }
}
/**
 * An error thrown when exporting private states fails.
 */
class PrivateStateExportError extends Error {
    constructor(message) {
        super(message);
        this.name = 'PrivateStateExportError';
    }
}
/**
 * An error thrown when exporting signing keys fails.
 */
class SigningKeyExportError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SigningKeyExportError';
    }
}
/**
 * Base error thrown when importing private states fails.
 */
class PrivateStateImportError extends Error {
    cause;
    constructor(message, cause) {
        super(message);
        this.cause = cause;
        this.name = 'PrivateStateImportError';
    }
}
/**
 * Error thrown when decryption of export data fails.
 * This could be due to wrong password, corrupted data, or tampered content.
 * The specific cause is intentionally not disclosed to prevent oracle attacks.
 */
class ExportDecryptionError extends PrivateStateImportError {
    constructor() {
        super('Failed to decrypt export data. The password may be incorrect or the data may be corrupted.', 'decryption_failed');
        this.name = 'ExportDecryptionError';
    }
}
/**
 * Error thrown when the export data format is invalid.
 */
class InvalidExportFormatError extends PrivateStateImportError {
    constructor(message = 'Invalid export format') {
        super(message, 'invalid_format');
        this.name = 'InvalidExportFormatError';
    }
}
/**
 * Error thrown when import conflicts with existing data and conflictStrategy is 'error'.
 */
class ImportConflictError extends PrivateStateImportError {
    conflictCount;
    constructor(conflictCount, entityName = 'private state') {
        super(`Import conflicts with ${conflictCount} existing ${entityName}${conflictCount === 1 ? '' : 's'}`, 'conflict');
        this.conflictCount = conflictCount;
        this.name = 'ImportConflictError';
    }
}

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A valid named log level.
 */
exports.LogLevel = void 0;
(function (LogLevel) {
    /**
     * Log levels typically used by DAapp developers.
     */
    LogLevel["INFO"] = "info";
    LogLevel["WARN"] = "warn";
    LogLevel["ERROR"] = "error";
    LogLevel["FATAL"] = "fatal";
    /**
     * Log levels used by Midnight.JS to report internal state.
     */
    LogLevel["DEBUG"] = "debug";
    LogLevel["TRACE"] = "trace";
})(exports.LogLevel || (exports.LogLevel = {}));

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**

 * Creates a branded prover key representation from a prover key binary.
 *
 * @param uint8Array The prover key binary.
 */
const createProverKey = (uint8Array) => {
    return uint8Array;
};
/**
 * Creates a branded verifier key representation from a verifier key binary.
 *
 * @param uint8Array The verifier key binary.
 */
const createVerifierKey = (uint8Array) => {
    return uint8Array;
};
/**
 * Creates a branded ZKIR representation from a ZKIR binary.
 *
 * @param uint8Array The ZKIR binary.
 */
const createZKIR = (uint8Array) => {
    return uint8Array;
};
/**
 * Converts a ZKConfig object to ProvingKeyMaterial format.
 * @param zkConfig
 */
const zkConfigToProvingKeyMaterial = (zkConfig) => {
    return {
        proverKey: zkConfig.proverKey,
        verifierKey: zkConfig.verifierKey,
        ir: zkConfig.zkir,
    };
};
/**
 * Indicates that the segment update is invalid.
 */
const SegmentFail = 'SegmentFail';
/**
 * Indicates that the segment is valid.
 */
const SegmentSuccess = 'SegmentSuccess';
/**
 * Indicates that the transaction is invalid.
 */
const FailEntirely = 'FailEntirely';
/**
 * Indicates that the transaction is valid but the portion of the transcript
 * that is allowed to fail (the portion after a checkpoint) did fail. All effects
 * from the guaranteed part of the transaction are kept but the effects from the
 * fallible part of the transaction are discarded.
 */
const FailFallible = 'FailFallible';
/**
 * Indicates that the guaranteed and fallible portions of the transaction were
 * successful.
 */
const SucceedEntirely = 'SucceedEntirely';

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Maximum number of states that can be exported/imported.
 * This limit prevents memory exhaustion attacks.
 */
const MAX_EXPORT_STATES = 10000;
/**
 * Maximum number of signing keys that can be exported/imported.
 * This limit prevents memory exhaustion attacks.
 */
const MAX_EXPORT_SIGNING_KEYS = 10000;

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Creates a {@link ProofProvider} from a {@link ProvingProvider}.
 * The returned provider proves transactions using the initial cost model.
 *
 * @param provingProvider - The underlying proving provider used to generate proofs.
 * @param costModel - Optional cost model to use for proof generation. Defaults to the initial cost model if not provided.
 * @returns A {@link ProofProvider} that delegates proof generation to the given proving provider.
 */
const createProofProvider = (provingProvider, costModel = ledgerV8.CostModel.initialCostModel()) => ({
    async proveTx(unprovenTx) {
        return unprovenTx.prove(provingProvider, costModel);
    }
});

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A provider for zero-knowledge intermediate representations, prover keys, and verifier keys. All
 * three are used by the {@link ProofProvider} to create a proof for a call transaction. The implementation
 * of this provider depends on the runtime environment, since each environment has different conventions
 * for accessing static artifacts.
 * @typeParam K - The type of the circuit ID used by the provider.
 */
class ZKConfigProvider {
    /**
     * Retrieves the verifier keys produced by `compactc` compiler for the given circuits.
     * @param circuitIds The circuit IDs of the verifier keys to retrieve.
     */
    async getVerifierKeys(circuitIds) {
        return Promise.all(circuitIds.map(async (id) => {
            const key = await this.getVerifierKey(id);
            return [id, key];
        }));
    }
    /**
     * Retrieves all zero-knowledge artifacts produced by `compactc` compiler for the given circuit.
     * @param circuitId The circuit ID of the artifacts to retrieve.
     */
    async get(circuitId) {
        return {
            circuitId,
            proverKey: await this.getProverKey(circuitId),
            verifierKey: await this.getVerifierKey(circuitId),
            zkir: await this.getZKIR(circuitId)
        };
    }
    asKeyMaterialProvider() {
        return this;
    }
}

Object.defineProperty(exports, "Transaction", {
    enumerable: true,
    get: function () { return ledgerV8.Transaction; }
});
exports.ExportDecryptionError = ExportDecryptionError;
exports.FailEntirely = FailEntirely;
exports.FailFallible = FailFallible;
exports.ImportConflictError = ImportConflictError;
exports.InvalidExportFormatError = InvalidExportFormatError;
exports.InvalidProtocolSchemeError = InvalidProtocolSchemeError;
exports.MAX_EXPORT_SIGNING_KEYS = MAX_EXPORT_SIGNING_KEYS;
exports.MAX_EXPORT_STATES = MAX_EXPORT_STATES;
exports.PrivateStateExportError = PrivateStateExportError;
exports.PrivateStateImportError = PrivateStateImportError;
exports.SegmentFail = SegmentFail;
exports.SegmentSuccess = SegmentSuccess;
exports.SigningKeyExportError = SigningKeyExportError;
exports.SucceedEntirely = SucceedEntirely;
exports.ZKConfigProvider = ZKConfigProvider;
exports.asContractAddress = asContractAddress;
exports.asEffectOption = asEffectOption;
exports.createProofProvider = createProofProvider;
exports.createProverKey = createProverKey;
exports.createVerifierKey = createVerifierKey;
exports.createZKIR = createZKIR;
exports.exitResultOrError = exitResultOrError;
exports.makeContractExecutableRuntime = makeContractExecutableRuntime;
exports.zkConfigToProvingKeyMaterial = zkConfigToProvingKeyMaterial;
