// @midnight-ntwrk/wallet-sdk-unshielded-wallet/dist/KeyStore.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { UnshieldedAddress } from '@midnight-ntwrk/wallet-sdk-address-format';
import { addressFromKey, signData, signatureVerifyingKey, } from '@midnight-ntwrk/ledger-v7';
export const PublicKey = {
    fromKeyStore: (keystore) => {
        return {
            publicKey: keystore.getPublicKey(),
            addressHex: keystore.getAddress(),
            address: keystore.getBech32Address().asString(),
        };
    },
};
export const createKeystore = (secretKey, networkId) => {
    const keystore = {
        getSecretKey: () => Buffer.from(secretKey),
        getBech32Address: () => {
            const address = keystore.getAddress();
            const addressBuffer = Buffer.from(address, 'hex');
            return UnshieldedAddress.codec.encode(networkId, new UnshieldedAddress(addressBuffer));
        },
        getPublicKey: () => signatureVerifyingKey(keystore.getSecretKey().toString('hex')),
        getAddress: () => addressFromKey(keystore.getPublicKey()),
        signData: (data) => signData(keystore.getSecretKey().toString('hex'), data),
    };
    return keystore;
};
