// @midnight-ntwrk/wallet-sdk-utilities/dist/testing/test-containers.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Effect, identity } from 'effect';
import { GenericContainer, Network, Wait } from 'testcontainers';
import { getPortPromise } from 'portfinder';
export const createNetwork = () => Effect.acquireRelease(Effect.promise(() => new Network().start()), (net) => Effect.promise(() => net.stop()));
const startContainer = (container) => {
    return Effect.acquireRelease(Effect.promise(() => container.start()), (container) => Effect.promise(() => container.stop({ timeout: 5_000 })));
};
export const runNodeContainer = (adjustment = identity) => {
    const container = new GenericContainer('ghcr.io/midnight-ntwrk/midnight-node:0.22.0')
        .withEnvironment({
        CFG_PRESET: 'dev',
        SIDECHAIN_BLOCK_BENEFICIARY: '04bcf7ad3be7a5c790460be82a713af570f22e0f801f6659ab8e84a52be6969e',
    })
        .withExposedPorts(9944)
        .withWaitStrategy(Wait.forListeningPorts());
    return startContainer(adjustment(container));
};
export const runProofServerContainer = (adjustment = identity) => {
    const container = new GenericContainer('ghcr.io/midnight-ntwrk/proof-server:8.0.2')
        .withEnvironment({
        RUST_BACKTRACE: 'full',
    })
        .withExposedPorts(6300)
        .withCommand(['midnight-proof-server -v'])
        .withWaitStrategy(Wait.forListeningPorts());
    return startContainer(adjustment(container));
};
export const runTxGenerator = (config, adjustment = identity) => {
    const container = new GenericContainer('ghcr.io/midnight-ntwrk/midnight-node-toolkit:0.22.0').withCommand([
        'generate-txs',
        '--dust-warp',
        ['--src-url', config.nodeUrl],
        ['--proof-server', config.proofServerUrl],
        ['--dest-file', `/tmp/${config.fileName}`],
        ['batches'],
        ['--num-batches', String(config.batches)],
        ['--num-txs-per-batch', String(config.txsPerBatch)],
    ].flat());
    return startContainer(adjustment(container));
};
export const findAvailablePort = Effect.promise(() => getPortPromise());
