// @midnight-ntwrk/wallet-sdk-node-client/dist/index.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { NodeClient, PolkadotNodeClient as EffectNodeClient, } from './effect/index.js';
import { Effect, Exit, pipe, Scope } from 'effect';
import { ObservableOps } from '@midnight-ntwrk/wallet-sdk-utilities';
export { makeConfig, DEFAULT_CONFIG } from './effect/PolkadotNodeClient.js';
export class PolkadotNodeClient {
    static init(config) {
        return Effect.gen(function* () {
            const scope = yield* Scope.make();
            const client = yield* NodeClient.NodeClient.pipe(Effect.provide(EffectNodeClient.layer(config)), Effect.provideService(Scope.Scope, scope));
            return new PolkadotNodeClient(client, scope);
        }).pipe(Effect.runPromise);
    }
    #effectClient;
    #scope;
    constructor(effectClient, scope) {
        this.#effectClient = effectClient;
        this.#scope = scope;
    }
    sendMidnightTransaction(serializedTransaction) {
        return ObservableOps.fromStream(this.#effectClient.sendMidnightTransaction(serializedTransaction));
    }
    sendMidnightTransactionAndWait(serializedTransaction, waitFor) {
        const runRequest = (request) => pipe(request, Effect.provideService(NodeClient.NodeClient, this.#effectClient), Effect.runPromise);
        return NodeClient.sendMidnightTransactionAndWait(serializedTransaction, waitFor).pipe(runRequest);
    }
    close() {
        return Scope.close(this.#scope, Exit.void).pipe(Effect.runPromise);
    }
}
