// @midnight-ntwrk/platform-js/dist/esm/effect/CoinPublicKey.js
/*
 * This file is part of midnight-sdk.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Brand } from 'effect';
import * as Either from 'effect/Either';
import * as Hex_ from './Hex.js';
/**
 * Creates a coin public key from a plain hex-encoded source string ({@link Hex_.PlainHex | PlainHex}).
 *
 * @category constructors
 */
export const Hex = Brand.all(Brand.nominal(), Hex_.PlainHex);
/**
 * Create a coin public key from a source string formatted according to the Bech32m encoding scheme.
 *
 * @category constructors
 */
export const Bech32m = Brand.nominal();
/**
 * Creates a hex-encoded coin public key from a given source.
 *
 * @param self A source string that should become a {@link CoinPublicKey.Hex}, or a {@link CoinPublicKey} that
 * should become a {@link CoinPublicKey.Hex}.
 * @returns A {@link CoinPublicKey.Hex}.
 *
 * @category constructors
 */
export const asHex = (self) => {
    if (Hex.is(self))
        return self;
    if (Bech32m.is(self)) {
        return /* TODO: convert */ Hex(self);
    }
    return Hex(self);
};
/**
 * Creates a Bech32m formatted coin public key from a given source.
 *
 * @param self A source string that should become a {@link CoinPublicKey.Bech32m}, or a {@link CoinPublicKey} that
 * should become a {@link CoinPublicKey.Bech32m}.
 * @returns A {@link CoinPublicKey.Bech32m}.
 *
 * @category constructors
 */
export const asBech32m = (self) => {
    if (Bech32m.is(self))
        return self;
    if (Hex.is(self)) {
        return /* TODO: convert */ Bech32m(self);
    }
    return Bech32m(self);
};
/**
 * Creates a buffer representing the raw bytes of a coin public key.
 *
 * @param self The {@link CoinPublicKey} for which raw bytes are required.
 * @returns A `Uint8Array` representing the raw bytes of `self`.
 *
 * @category constructors
 */
export const asBytes = (self) => Buffer.from(asHex(self), 'hex');
/**
 * Create a coin public key from a source string.
 *
 * @param value The string value that is become a {@link CoinPublicKey}.
 * @returns A {@link CoinPublicKey} that is an instance of {@link CoinPublicKey.Hex} if `value` could be
 * parsed as a hex-encoded string; otherwise as an instance of {@link CoinPublicKey.Bech32m}.
 *
 * @category constructors
 */
export const make = (value) => {
    return Either.match(Hex_.parseHex(value), {
        onRight: (_) => Hex(value),
        onLeft: (_) => Bech32m(value)
    });
};
