// @midnight-ntwrk/platform-js/dist/esm/effect/NetworkId.js
/*
 * This file is part of midnight-sdk.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as Equal from 'effect/Equal';
import * as equivalence from 'effect/Equivalence';
import { dual } from 'effect/Function';
import { NodeInspectSymbol } from 'effect/Inspectable';
import { hasProperty } from 'effect/Predicate';
import * as NetworkIdMoniker from './NetworkIdMoniker.js';
const MAINNET_MONIKER = NetworkIdMoniker.NetworkIdMoniker('main');
const TypeId = Symbol.for('platform-js/effect/NetworkId');
const MonikerSymbol = Symbol.for('platform-js/effect/NetworkId#NetworkMoniker');
/**
 * Provides equivalence for {@link NetworkId} instances.
 *
 * @remarks
 * For two {@link NetworkId} instances to be considered equal, their underlying monikers must be equal.
 *
 * @category equivalence
 */
export const Equivalence = equivalence.mapInput((a, b) => Equal.equals(a, b), (_) => _[MonikerSymbol]);
/**
 * Determines if a value is a network identifier.
 *
 * @param u The value to check.
 * @returns `true` if `u` is a {@link NetworkId}; `false` otherwise.
 *
 * @category guards
 */
export const isNetworkId = (u) => hasProperty(u, TypeId);
/**
 * Determines if two network identifiers are equal.
 *
 * @category predicates
 */
export const equals = dual(2, (self, that) => Equivalence(self, that));
const NetworkIdProto = (networkMoniker) => ({
    [TypeId]: TypeId,
    [Equal.symbol](that) {
        return isNetworkId(that) && equals(this, that);
    },
    toString() {
        return this[MonikerSymbol] === true
            ? MAINNET_MONIKER
            : this[MonikerSymbol];
    },
    toJSON() {
        return {
            _id: 'NetworkId',
            moniker: this.toString()
        };
    },
    [NodeInspectSymbol]() {
        return this.toJSON();
    },
    [MonikerSymbol]: networkMoniker,
    isMainNet() {
        return this[MonikerSymbol] === true;
    }
});
/**
 * A network identifier that represents the Midnight MainNet.
 */
export const MainNet = Object.create(NetworkIdProto(true));
/**
 * Creates a network identifier.
 *
 * @param input The input to use when constructing the network identifier.
 * @returns A {@link NetworkId} derived from `input`.
 *
 * @category constructors
 */
export const make = (input) => Object.create(NetworkIdProto(isNetworkId(input) ? input[MonikerSymbol] : input));
