// @midnight-ntwrk/wallet-sdk-dust-wallet/dist/v1/CoreWallet.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { DustLocalState, } from '@midnight-ntwrk/ledger-v7';
import { ProtocolVersion, SyncProgress } from '@midnight-ntwrk/wallet-sdk-abstractions';
import { DateOps } from '@midnight-ntwrk/wallet-sdk-utilities';
import { Array as Arr, Option, pipe } from 'effect';
export const PublicKey = {
    fromSecretKey: (secretKey) => {
        return {
            publicKey: secretKey.publicKey,
        };
    },
};
export const CoreWallet = {
    init(localState, secretKey, networkId) {
        return CoreWallet.empty(localState, PublicKey.fromSecretKey(secretKey), networkId);
    },
    initEmpty(dustParameters, secretKey, networkId) {
        return CoreWallet.empty(new DustLocalState(dustParameters), PublicKey.fromSecretKey(secretKey), networkId);
    },
    empty(localState, publicKey, networkId) {
        return {
            state: localState,
            publicKey,
            networkId,
            pendingDust: [],
            progress: SyncProgress.createSyncProgress(),
            protocolVersion: ProtocolVersion.MinSupportedVersion,
        };
    },
    restore(localState, publicKey, pendingTokens, syncProgress, protocolVersion, networkId) {
        return {
            state: localState,
            publicKey,
            networkId,
            pendingDust: pendingTokens,
            progress: SyncProgress.createSyncProgress(syncProgress),
            protocolVersion: ProtocolVersion.ProtocolVersion(protocolVersion),
        };
    },
    applyEvents(wallet, secretKey, events, currentTime) {
        // TODO: replace currentTime with `updatedState.syncTime` introduced in ledger-6.2.0-rc.1
        const updatedState = wallet.state.replayEvents(secretKey, events).processTtls(currentTime);
        const availableNonces = updatedState.utxos.map((utxo) => utxo.nonce);
        return {
            ...wallet,
            state: updatedState,
            pendingDust: wallet.pendingDust.filter((t) => availableNonces.includes(t.nonce)),
        };
    },
    applyFailed(wallet, tx) {
        const pendingSpendsMap = CoreWallet.pendingDustToMap(wallet.pendingDust);
        const relevantSpends = pipe([...(tx.intents?.values() ?? [])], Arr.flatMap((intent) => {
            const spendTime = intent.dustActions?.ctime;
            return (intent.dustActions?.spends ?? []).map((spend) => ({ spend, spendTime }));
        }), Arr.filterMap(({ spend, spendTime }) => pipe(Option.fromNullable(pendingSpendsMap.get(spend.oldNullifier)), Option.map(() => ({ spend, spendTime })))));
        const [updatedState, removedNullifiers] = pipe(relevantSpends, Arr.reduce([wallet.state, []], ([state, removed], { spend, spendTime }) => [
            state.processTtls(DateOps.addSeconds(spendTime, wallet.state.params.dustGracePeriodSeconds)),
            Arr.append(removed, spend.oldNullifier),
        ]));
        return {
            ...wallet,
            state: updatedState,
            pendingDust: wallet.pendingDust.filter((coin) => !removedNullifiers.includes(coin.nullifier)),
        };
    },
    revertTransaction(wallet, tx) {
        return CoreWallet.applyFailed(wallet, tx);
    },
    updateProgress(wallet, { appliedIndex, highestRelevantWalletIndex, highestIndex, highestRelevantIndex, isConnected, }) {
        const updatedProgress = SyncProgress.createSyncProgress({
            appliedIndex: appliedIndex ?? wallet.progress.appliedIndex,
            highestRelevantWalletIndex: highestRelevantWalletIndex ?? wallet.progress.highestRelevantWalletIndex,
            highestIndex: highestIndex ?? wallet.progress.highestIndex,
            highestRelevantIndex: highestRelevantIndex ?? wallet.progress.highestRelevantIndex,
            isConnected: isConnected ?? wallet.progress.isConnected,
        });
        return { ...wallet, progress: updatedProgress };
    },
    spendCoins(wallet, secretKey, coins, currentTime) {
        const [output, newState, newPending] = pipe(coins, Arr.reduce([[], wallet.state, wallet.pendingDust], ([spends, localState, pending], { token: coinToSpend, value: takeFee }) => {
            const [newState, dustSpend] = localState.spend(secretKey, coinToSpend, takeFee, currentTime);
            const newPending = [...pending, { ...coinToSpend, nullifier: dustSpend.oldNullifier }];
            return [Arr.append(spends, dustSpend), newState, newPending];
        }));
        return [output, { ...wallet, state: newState, pendingDust: newPending }];
    },
    pendingDustToMap(coins) {
        return new Map(coins.map(({ nullifier, ...coins }) => [nullifier, coins]));
    },
};
