// @midnight-ntwrk/wallet-sdk-dust-wallet/dist/v1/Transacting.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Either, pipe, BigInt as BigIntOps, Iterable as IterableOps, Option } from 'effect';
import { DustActions, DustRegistration, Intent, SignatureEnabled, Transaction, UnshieldedOffer, addressFromKey, nativeToken, } from '@midnight-ntwrk/ledger-v7';
import { OtherWalletError, TransactingError } from './WalletError.js';
import { LedgerOps } from '@midnight-ntwrk/wallet-sdk-utilities';
import { CoreWallet } from './CoreWallet.js';
import { BindingMarker, ProofMarker, SignatureMarker } from './Utils.js';
export const makeDefaultTransactingCapability = (config, getContext) => {
    return new TransactingCapabilityImplementation(config.networkId, config.costParameters, () => getContext().coinSelection, () => getContext().coinsAndBalancesCapability, () => getContext().keysCapability);
};
export const makeSimulatorTransactingCapability = (config, getContext) => {
    return new TransactingCapabilityImplementation(config.networkId, config.costParameters, () => getContext().coinSelection, () => getContext().coinsAndBalancesCapability, () => getContext().keysCapability);
};
export class TransactingCapabilityImplementation {
    networkId;
    costParams;
    getCoinSelection;
    getCoins;
    getKeys;
    constructor(networkId, costParams, getCoinSelection, getCoins, getKeys) {
        this.getCoins = getCoins;
        this.networkId = networkId;
        this.costParams = costParams;
        this.getCoinSelection = getCoinSelection;
        this.getKeys = getKeys;
    }
    createDustGenerationTransaction(currentTime, ttl, nightUtxos, nightVerifyingKey, dustReceiverAddress) {
        const makeOffer = (utxos) => {
            if (utxos.length === 0) {
                return Option.none();
            }
            const totalValue = pipe(utxos, IterableOps.map((coin) => coin.utxo.value), BigIntOps.sumAll);
            const inputs = utxos.map(({ utxo }) => ({
                ...utxo,
                owner: nightVerifyingKey,
            }));
            const output = {
                owner: addressFromKey(nightVerifyingKey),
                type: nativeToken().raw,
                value: totalValue,
            };
            return Option.some(UnshieldedOffer.new(inputs, [output], []));
        };
        return Either.gen(this, function* () {
            const receiver = dustReceiverAddress ? dustReceiverAddress.data : undefined;
            return yield* LedgerOps.ledgerTry(() => {
                const network = this.networkId;
                const splitResult = this.getCoins().splitNightUtxos(nightUtxos);
                // if receiver is `undefined`, it means the coin(s) are being deregistered so the allowFeePayment should not be used
                const feePayment = receiver
                    ? pipe(splitResult.guaranteed, IterableOps.filter((coin) => !coin.utxo.registeredForDustGeneration), IterableOps.map((coin) => coin.dust.generatedNow), BigIntOps.sumAll)
                    : 0n;
                const maybeGuaranteedOffer = makeOffer(splitResult.guaranteed);
                const maybeFallibleOffer = makeOffer(splitResult.fallible);
                const dustRegistration = new DustRegistration(SignatureMarker.signature, nightVerifyingKey, receiver, feePayment);
                const dustActions = new DustActions(SignatureMarker.signature, ProofMarker.preProof, currentTime, [], [dustRegistration]);
                const intent = pipe(Intent.new(ttl), (intent) => Option.match(maybeGuaranteedOffer, {
                    onNone: () => intent,
                    onSome: (guaranteedOffer) => {
                        intent.guaranteedUnshieldedOffer = guaranteedOffer;
                        return intent;
                    },
                }), (intent) => Option.match(maybeFallibleOffer, {
                    onNone: () => intent,
                    onSome: (fallibleOffer) => {
                        intent.fallibleUnshieldedOffer = fallibleOffer;
                        return intent;
                    },
                }), (intent) => {
                    intent.dustActions = dustActions;
                    return intent;
                });
                return Transaction.fromParts(network, undefined, undefined, intent);
            });
        });
    }
    addDustGenerationSignature(transaction, signatureData) {
        return Either.gen(this, function* () {
            const intent = transaction.intents?.get(1);
            if (!intent) {
                return yield* Either.left(new TransactingError({ message: 'No intent found in the transaction intents with segment = 1' }));
            }
            const { dustActions, guaranteedUnshieldedOffer, fallibleUnshieldedOffer } = intent;
            if (!dustActions) {
                return yield* Either.left(new TransactingError({ message: 'No dustActions found in intent' }));
            }
            if (!guaranteedUnshieldedOffer) {
                return yield* Either.left(new TransactingError({ message: 'No guaranteedUnshieldedOffer found in intent' }));
            }
            const [registration, ...restRegistrations] = dustActions.registrations;
            if (!registration) {
                return yield* Either.left(new TransactingError({ message: 'No registrations found in dustActions' }));
            }
            return yield* LedgerOps.ledgerTry(() => {
                const signature = new SignatureEnabled(signatureData);
                const registrationWithSignature = new DustRegistration(signature.instance, registration.nightKey, registration.dustAddress, registration.allowFeePayment, signature);
                const newDustActions = new DustActions(signature.instance, ProofMarker.preProof, dustActions.ctime, dustActions.spends, [registrationWithSignature, ...restRegistrations]);
                // make a copy of intent to avoid mutation
                const newIntent = Intent.deserialize(signature.instance, ProofMarker.preProof, BindingMarker.preBinding, intent.serialize());
                newIntent.dustActions = newDustActions;
                const inputsLen = guaranteedUnshieldedOffer.inputs.length;
                const signatures = [];
                for (let i = 0; i < inputsLen; ++i) {
                    signatures.push(guaranteedUnshieldedOffer.signatures.at(i) ?? signatureData);
                }
                newIntent.guaranteedUnshieldedOffer = guaranteedUnshieldedOffer.addSignatures(signatures);
                if (fallibleUnshieldedOffer) {
                    const inputsLen = fallibleUnshieldedOffer.inputs.length;
                    const signatures = [];
                    for (let i = 0; i < inputsLen; ++i) {
                        signatures.push(fallibleUnshieldedOffer.signatures.at(i) ?? signatureData);
                    }
                    newIntent.fallibleUnshieldedOffer = fallibleUnshieldedOffer.addSignatures(signatures);
                }
                // make a copy of transaction to avoid mutation
                const newTransaction = Transaction.deserialize(signature.instance, ProofMarker.preProof, BindingMarker.preBinding, transaction.serialize());
                newTransaction.intents = newTransaction.intents.set(1, newIntent);
                return newTransaction;
            });
        });
    }
    calculateFee(transaction, ledgerParams) {
        return (transaction.feesWithMargin(ledgerParams, this.costParams.feeBlocksMargin) + this.costParams.additionalFeeOverhead);
    }
    static feeImbalance(transaction, totalFee) {
        const dustImbalance = transaction
            .imbalances(0, totalFee)
            .entries()
            .find(([tt, _]) => tt.tag === 'dust');
        return dustImbalance ? -dustImbalance[1] : totalFee;
    }
    balanceTransactions(secretKey, state, transactions, ttl, currentTime, ledgerParams) {
        const network = this.networkId;
        const feeLeft = transactions.reduce((total, transaction) => total +
            TransactingCapabilityImplementation.feeImbalance(transaction, this.calculateFee(transaction, ledgerParams)), 0n);
        const dust = this.getCoins().getAvailableCoinsWithGeneratedDust(state, currentTime);
        const selectedCoins = this.getCoinSelection()(dust, feeLeft);
        if (!selectedCoins.length) {
            return Either.left(new TransactingError({ message: 'No dust found in the wallet state' }));
        }
        const totalFeeInSelected = selectedCoins.reduce((total, { value }) => total + value, 0n);
        const feeDiff = totalFeeInSelected - feeLeft;
        if (feeDiff < 0n) {
            return Either.left(new TransactingError({ message: 'Not enough dust generated to pay the fee' }));
        }
        // reduce the largest token's value by `feeDiff`
        const coinsWithFeeToTake = selectedCoins.toSorted((a, b) => Number(b.value - a.value));
        if (feeDiff > 0n) {
            const highestByValue = coinsWithFeeToTake[0];
            coinsWithFeeToTake[0] = {
                value: highestByValue.value - feeDiff,
                token: highestByValue.token,
            };
        }
        return LedgerOps.ledgerTry(() => {
            const intent = Intent.new(ttl);
            const [spends, updatedState] = CoreWallet.spendCoins(state, secretKey, coinsWithFeeToTake, currentTime);
            intent.dustActions = new DustActions(SignatureMarker.signature, ProofMarker.preProof, currentTime, [...spends], []);
            const feeTransaction = Transaction.fromPartsRandomized(network, undefined, undefined, intent);
            return [feeTransaction, updatedState];
        });
    }
    revertTransaction(state, transaction) {
        return Either.try({
            try: () => CoreWallet.revertTransaction(state, transaction),
            catch: (err) => {
                return new OtherWalletError({
                    message: `Error while reverting transaction ${transaction.identifiers().at(0)}`,
                    cause: err,
                });
            },
        });
    }
}
