// @midnight-ntwrk/platform-js/dist/esm/effect/Hex.js
/*
 * This file is part of midnight-sdk.
 * Copyright (C) 2025 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as internal from './internal/hex.js';
/**
 * Creates a hex-encoded string, of some arbitrary byte length, that has a `'0x'` prefix.
 *
 * @category constructors
 */
export const PrefixedHex = internal.make({ requirePrefix: true });
/**
 * Creates a hex-encoded string, from some given constraints, that has a `'0x'` prefix.
 *
 * @param constraints The {@link HexConstraints} to apply when parsing a received hex-encoded string.
 * @returns A function that creates a {@link Hex.PrefixedHex | PrefixedHex} instance from a received string
 * ensuring that it meets `constraints`.
 *
 * @category constructors
 */
export const ConstrainedPrefixedHex = (constraints) => internal.make({ requirePrefix: true, ...constraints });
/**
 * Creates a plain hex-encoded string, of some arbitrary byte length.
 *
 * @category constructors
 */
export const PlainHex = internal.make({ requirePrefix: false });
/**
 * Creates a plain hex-encoded string, from some given constraints.
 *
 * @param constraints The {@link HexConstraints} to apply when parsing a received hex-encoded string.
 * @returns A function that creates a {@link Hex.PlainHex | PlainHex} instance from a received string
 * ensuring that it meets `constraints`.
 *
 * @category constructors
 */
export const ConstrainedPlainHex = (constraints) => internal.make({ requirePrefix: false, ...constraints });
;
/**
 * Parses a hex-encoded string.
 *
 * @param source The source string to parse.
 * @returns An `Either` with a `Right` value of {@link ParsedHexString} describing the parsed elements of `source`,
 * or a `Left` value of {@link ParseError.ParseError | ParseError} if parsing fails.
 *
 * @category utilities
 */
export const parseHex = internal.parseHex;
