// @midnight-ntwrk/wallet-sdk-shielded/dist/v1/WalletError.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { Data } from 'effect';
export const WalletError = {
    other(err) {
        let message;
        if (err) {
            if (typeof err == 'object' && 'message' in err) {
                message = String(err.message);
            }
            else if (typeof err == 'string') {
                message = err;
            }
            else {
                message = '';
            }
        }
        else {
            message = '';
        }
        return new OtherWalletError({ message: `Other wallet error: ${message}`, cause: err });
    },
    submission(err) {
        const message = err && typeof err == 'object' && 'message' in err ? String(err.message) : '';
        return new SubmissionError({ message: `Transaction submission error: ${message}`, cause: err });
    },
};
export class OtherWalletError extends Data.TaggedError('Wallet.Other') {
}
export class SyncWalletError extends Data.TaggedError('Wallet.Sync') {
}
export class SubmissionError extends Data.TaggedError('Wallet.SubmissionWalletError') {
}
export class InsufficientFundsError extends Data.TaggedError('Wallet.InsufficientFunds') {
}
export class AddressError extends Data.TaggedError('Wallet.Address') {
}
export class InvalidCoinHashesError extends Data.TaggedError('Wallet.InvalidCoinHashes') {
}
export class TransactingError extends Data.TaggedError('Wallet.Transacting') {
}
