// @midnight-ntwrk/midnight-js-http-client-proof-provider/dist/index.cjs
'use strict';

var ledgerV8 = require('@midnight-ntwrk/ledger-v8');
var midnightJsTypes = require('@midnight-ntwrk/midnight-js-types');
var fetch = require('cross-fetch');
var fetchBuilder = require('fetch-retry');

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const retryOptions = {
    retries: 3,
    retryDelay: (attempt) => 2 ** attempt * 1_000,
    retryOn: [500, 503]
};
const fetchRetry = fetchBuilder(fetch, retryOptions);
const CHECK_PATH = '/check';
const PROVE_PATH = '/prove';
const buildEndpointUrl = (baseUrl, endpoint) => {
    const url = new URL(baseUrl);
    url.pathname = url.pathname.replace(/\/$/, '') + endpoint;
    return url;
};
const DEFAULT_TIMEOUT = 300000;
const getKeyMaterial = async (zkConfigProvider, keyLocation) => {
    try {
        const zkConfig = await zkConfigProvider.get(keyLocation);
        return midnightJsTypes.zkConfigToProvingKeyMaterial(zkConfig);
    }
    catch {
        return undefined;
    }
};
const makeHttpRequest = async (url, payload, timeout, headers = {}) => {
    const response = await fetchRetry(url, {
        method: 'POST',
        body: new Uint8Array(payload),
        headers: { 'Content-Type': 'application/octet-stream', ...headers },
        signal: AbortSignal.timeout(timeout)
    });
    if (!response.ok) {
        throw new Error(`Failed Proof Server response: url="${response.url}", code="${response.status}", status="${response.statusText}"`);
    }
    return new Uint8Array(await response.arrayBuffer());
};
const httpClientProvingProvider = (url, zkConfigProvider, config) => {
    const checkUrl = buildEndpointUrl(url, CHECK_PATH);
    const proveUrl = buildEndpointUrl(url, PROVE_PATH);
    if (checkUrl.protocol !== 'http:' && checkUrl.protocol !== 'https:') {
        throw new midnightJsTypes.InvalidProtocolSchemeError(checkUrl.protocol, ['http:', 'https:']);
    }
    if (proveUrl.protocol !== 'http:' && proveUrl.protocol !== 'https:') {
        throw new midnightJsTypes.InvalidProtocolSchemeError(proveUrl.protocol, ['http:', 'https:']);
    }
    const timeout = config?.timeout ?? DEFAULT_TIMEOUT;
    const headers = config?.headers ?? {};
    return {
        async check(serializedPreimage, keyLocation) {
            const keyMaterial = await getKeyMaterial(zkConfigProvider, keyLocation);
            const payload = ledgerV8.createCheckPayload(serializedPreimage, keyMaterial?.ir);
            const result = await makeHttpRequest(checkUrl, payload, timeout, headers);
            return ledgerV8.parseCheckResult(result);
        },
        async prove(serializedPreimage, keyLocation, overwriteBindingInput) {
            const keyMaterial = await getKeyMaterial(zkConfigProvider, keyLocation);
            const payload = ledgerV8.createProvingPayload(serializedPreimage, overwriteBindingInput, keyMaterial);
            return makeHttpRequest(proveUrl, payload, timeout, headers);
        }
    };
};

/*
 * This file is part of midnight-js.
 * Copyright (C) 2025-2026 Midnight Foundation
 * SPDX-License-Identifier: Apache-2.0
 * Licensed under the Apache License, Version 2.0 (the "License");
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const DEFAULT_CONFIG = {
    timeout: 300000,
    zkConfig: undefined
};
/**
 * Creates a high-level {@link ProofProvider} that implements transaction-level proving
 * using the low-level circuit-by-circuit {@link ProvingProvider} as its foundation.
 *
 * This adapter bridges the gap between:
 * - High-level ProofProvider interface (works with complete transactions)
 * - Low-level ProvingProvider interface (works with individual circuits)
 *
 * @param url The URL of the proof server
 * @param zkConfigProvider Provider for zero-knowledge configuration artifacts
 * @param config Optional configuration for the underlying ProvingProvider
 * @returns A ProofProvider instance that uses ProvingProvider internally
 *
 * @remarks
 * **Architecture:**
 * ```
 * ProofProvider (Transaction-level)
 *     ↓ (adapter)
 * ProvingProvider (Circuit-level)
 *     ↓ (HTTP client)
 * Proof Server (/check, /prove endpoints)
 * ```
 *
 * **Note:** The /prove-tx endpoint is NOT used. All proving is done through
 * individual circuit operations using /check and /prove endpoints.
 */
const httpClientProofProvider = (url, zkConfigProvider, config) => {
    const baseProvingProvider = httpClientProvingProvider(url, zkConfigProvider, config);
    return {
        async proveTx(unprovenTx, _partialProveTxConfig) {
            const costModel = ledgerV8.CostModel.initialCostModel();
            return unprovenTx.prove(baseProvingProvider, costModel);
        }
    };
};

exports.DEFAULT_CONFIG = DEFAULT_CONFIG;
exports.DEFAULT_TIMEOUT = DEFAULT_TIMEOUT;
exports.httpClientProofProvider = httpClientProofProvider;
exports.httpClientProvingProvider = httpClientProvingProvider;
