// @midnight-ntwrk/wallet-sdk-facade/dist/index.js
// This file is part of MIDNIGHT-WALLET-SDK.
// Copyright (C) Midnight Foundation
// SPDX-License-Identifier: Apache-2.0
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import * as ledger from '@midnight-ntwrk/ledger-v8';
import { makeDefaultSubmissionService, } from '@midnight-ntwrk/wallet-sdk-capabilities';
import { makeDefaultProvingService, } from '@midnight-ntwrk/wallet-sdk-capabilities/proving';
import { FetchTermsAndConditions as FetchTermsAndConditionsQuery } from '@midnight-ntwrk/wallet-sdk-indexer-client';
import { QueryRunner } from '@midnight-ntwrk/wallet-sdk-indexer-client/effect';
import { Array as Arr, pipe } from 'effect';
import { combineLatest, map, firstValueFrom, concatMap } from 'rxjs';
import { PendingTransactions, PendingTransactionsServiceImpl, } from '@midnight-ntwrk/wallet-sdk-capabilities';
import { finalizedTransactionTrait } from './transaction.js';
const TokenKindsToBalance = new (class {
    allTokenKinds = ['shielded', 'unshielded', 'dust'];
    toFlags = (tokenKinds) => {
        return pipe(tokenKinds, (kinds) => (kinds === 'all' ? this.allTokenKinds : kinds), (kinds) => ({
            shouldBalanceUnshielded: kinds.includes('unshielded'),
            shouldBalanceShielded: kinds.includes('shielded'),
            shouldBalanceDust: kinds.includes('dust'),
        }));
    };
})();
export const BalancingRecipe = {
    isRecipe: (value) => {
        return (typeof value === 'object' &&
            value !== null &&
            'type' in value &&
            typeof value.type === 'string' &&
            ['FINALIZED_TRANSACTION', 'UNBOUND_TRANSACTION', 'UNPROVEN_TRANSACTION'].includes(value.type));
    },
    getTransactions: (recipe) => {
        switch (recipe.type) {
            case 'FINALIZED_TRANSACTION': {
                return [recipe.originalTransaction, recipe.balancingTransaction];
            }
            case 'UNBOUND_TRANSACTION': {
                const balancingPart = recipe.balancingTransaction ? [recipe.balancingTransaction] : [];
                return [recipe.baseTransaction, ...balancingPart];
            }
            case 'UNPROVEN_TRANSACTION': {
                return [recipe.transaction];
            }
        }
    },
};
export class FacadeState {
    shielded;
    unshielded;
    dust;
    pending;
    get isSynced() {
        return (this.shielded.state.progress.isStrictlyComplete() &&
            this.dust.state.progress.isStrictlyComplete() &&
            this.unshielded.progress.isStrictlyComplete());
    }
    constructor(shielded, unshielded, dust, pending) {
        this.shielded = shielded;
        this.unshielded = unshielded;
        this.dust = dust;
        this.pending = pending;
    }
}
const DEFAULT_TTL_MS = 60 * 60 * 1000; // 1 hour
export class WalletFacade {
    static makeDefaultSubmissionService(config) {
        return makeDefaultSubmissionService(config);
    }
    static makeDefaultPendingTransactionsService(config) {
        return PendingTransactionsServiceImpl.init({
            configuration: config,
            txTrait: finalizedTransactionTrait,
        });
    }
    static makeDefaultProvingService(config) {
        if (config.provingServerUrl) {
            return makeDefaultProvingService({
                provingServerUrl: config.provingServerUrl,
            });
        }
        else {
            throw new Error("Missing required configuration: 'provingServerUrl' must be set in config, or provide a custom provingService in init parameters.");
        }
    }
    /**
     * Fetches the current Terms and Conditions from the network indexer.
     *
     * This is a static, pre-initialization utility — no wallet instance is required.
     * Wallet builders should call this before or independently of wallet initialization
     * to display the current T&C to end users and obtain the hash for content verification.
     *
     * The returned `hash` is the hex-encoded SHA-256 hash of the document at `url`.
     * Wallet builders are responsible for fetching and rendering the document content
     * via `url` in whatever manner suits their application.
     *
     * @param configuration - An object with an `indexerClientConnection.indexerHttpUrl`.
     *   Any wallet configuration that satisfies {@link FetchTermsAndConditionsConfiguration} can be passed directly.
     * @returns A promise resolving to the current {@link TermsAndConditions}, or rejecting if
     *   no Terms and Conditions have been set on the network yet.
     */
    static async fetchTermsAndConditions(configuration) {
        const result = await QueryRunner.runPromise(FetchTermsAndConditionsQuery, {}, {
            url: configuration.indexerClientConnection.indexerHttpUrl,
        });
        const tc = result.block?.systemParameters?.termsAndConditions;
        if (!tc) {
            throw new Error('Terms and Conditions are not currently set on the network.');
        }
        return tc;
    }
    static async init(initParams) {
        const submissionService = await Promise.resolve(initParams.submissionService
            ? initParams.submissionService(initParams.configuration)
            : WalletFacade.makeDefaultSubmissionService(initParams.configuration));
        const pendingTransactionsService = await Promise.resolve(initParams.pendingTransactionsService
            ? initParams.pendingTransactionsService(initParams.configuration)
            : WalletFacade.makeDefaultPendingTransactionsService(initParams.configuration));
        const provingService = await Promise.resolve(initParams.provingService
            ? initParams.provingService(initParams.configuration)
            : WalletFacade.makeDefaultProvingService(initParams.configuration));
        const shielded = await Promise.resolve(initParams.shielded(initParams.configuration));
        const unshielded = await Promise.resolve(initParams.unshielded(initParams.configuration));
        const dust = await Promise.resolve(initParams.dust(initParams.configuration));
        return new WalletFacade(shielded, unshielded, dust, submissionService, pendingTransactionsService, provingService);
    }
    shielded;
    unshielded;
    dust;
    submissionService;
    pendingTransactionsService;
    provingService;
    #pendingSubscription;
    constructor(shieldedWallet, unshieldedWallet, dustWallet, submissionService, pendingTransactionsService, provingService) {
        this.shielded = shieldedWallet;
        this.unshielded = unshieldedWallet;
        this.dust = dustWallet;
        this.submissionService = submissionService;
        this.pendingTransactionsService = pendingTransactionsService;
        this.provingService = provingService;
        this.#pendingSubscription = this.pendingTransactionsService
            .state()
            .pipe(concatMap((pending) => PendingTransactions.allFailed(pending)), concatMap((item) => this.revert(item.tx)))
            .subscribe();
    }
    defaultTtl() {
        return new Date(Date.now() + DEFAULT_TTL_MS);
    }
    mergeUnprovenTransactions(a, b) {
        if (a && b)
            return a.merge(b);
        return a ?? b;
    }
    async createDustActionTransaction(action, nightUtxos, nightVerifyingKey, signDustRegistration) {
        const ttl = this.defaultTtl();
        const transaction = await this.dust.createDustGenerationTransaction(undefined, ttl, nightUtxos.map(({ utxo, meta }) => ({
            ...utxo,
            ctime: meta.ctime,
            registeredForDustGeneration: meta.registeredForDustGeneration,
        })), nightVerifyingKey, action.type === 'registration' ? action.dustReceiverAddress : undefined);
        const intent = transaction.intents?.get(1);
        if (!intent) {
            throw Error('Dust generation transaction is missing intent segment 1.');
        }
        const signatureData = intent.signatureData(1);
        const signature = await Promise.resolve(signDustRegistration(signatureData));
        return await this.dust.addDustGenerationSignature(transaction, signature);
    }
    state() {
        return combineLatest([
            this.shielded.state,
            this.unshielded.state,
            this.dust.state,
            this.pendingTransactionsService.state(),
        ]).pipe(map(([shieldedState, unshieldedState, dustState, pending]) => new FacadeState(shieldedState, unshieldedState, dustState, pending)));
    }
    async waitForSyncedState() {
        const [shieldedState, unshieldedState, dustState, pending] = await Promise.all([
            this.shielded.waitForSyncedState(),
            this.unshielded.waitForSyncedState(),
            this.dust.waitForSyncedState(),
            firstValueFrom(this.pendingTransactionsService.state()),
        ]);
        return new FacadeState(shieldedState, unshieldedState, dustState, pending);
    }
    async submitTransaction(tx) {
        try {
            await this.pendingTransactionsService.addPendingTransaction(tx);
            await this.submissionService.submitTransaction(tx, 'Finalized');
            return tx.identifiers().at(-1);
        }
        catch (error) {
            await this.revert(tx);
            throw error;
        }
    }
    async balanceFinalizedTransaction(tx, secretKeys, options) {
        const { shieldedSecretKeys, dustSecretKey } = secretKeys;
        const { ttl, tokenKindsToBalance = 'all' } = options;
        const { shouldBalanceDust, shouldBalanceShielded, shouldBalanceUnshielded } = TokenKindsToBalance.toFlags(tokenKindsToBalance);
        // Step 1: Run unshielded and shielded balancing
        const unshieldedBalancingTx = shouldBalanceUnshielded
            ? await this.unshielded.balanceFinalizedTransaction(tx)
            : undefined;
        const shieldedBalancingTx = shouldBalanceShielded
            ? await this.shielded.balanceTransaction(shieldedSecretKeys, tx)
            : undefined;
        // Step 2: Merge unshielded and shielded balancing
        const mergedBalancingTx = this.mergeUnprovenTransactions(shieldedBalancingTx, unshieldedBalancingTx);
        // Step 3: Conditionally add dust/fee balancing
        const feeBalancingTx = shouldBalanceDust
            ? await this.dust.balanceTransactions(dustSecretKey, mergedBalancingTx ? [tx, mergedBalancingTx] : [tx], ttl)
            : undefined;
        // Step 4: Merge fee balancing and create final recipe
        const balancingTx = this.mergeUnprovenTransactions(mergedBalancingTx, feeBalancingTx);
        if (!balancingTx) {
            throw new Error('No balancing transaction was created. Please check your transaction.');
        }
        return {
            type: 'FINALIZED_TRANSACTION',
            originalTransaction: tx,
            balancingTransaction: balancingTx,
        };
    }
    async balanceUnboundTransaction(tx, secretKeys, options) {
        const { shieldedSecretKeys, dustSecretKey } = secretKeys;
        const { ttl, tokenKindsToBalance = 'all' } = options;
        const { shouldBalanceDust, shouldBalanceShielded, shouldBalanceUnshielded } = TokenKindsToBalance.toFlags(tokenKindsToBalance);
        // Step 1: Run unshielded and shielded balancing
        const shieldedBalancingTx = shouldBalanceShielded
            ? await this.shielded.balanceTransaction(shieldedSecretKeys, tx)
            : undefined;
        // For unbound transactions, unshielded balancing happens in place not with a balancing transaction
        const balancedUnshieldedTx = shouldBalanceUnshielded
            ? await this.unshielded.balanceUnboundTransaction(tx)
            : undefined;
        // Step 2: Unbound unshielded tx are balanced in place, use it as base tx if present
        const baseTx = balancedUnshieldedTx ?? tx;
        // Step 3: Conditionally add dust/fee balancing
        const feeBalancingTransaction = shouldBalanceDust
            ? await this.dust.balanceTransactions(dustSecretKey, shieldedBalancingTx ? [baseTx, shieldedBalancingTx] : [baseTx], ttl)
            : undefined;
        // Step 4: Create the final balancing transaction
        const balancingTransaction = this.mergeUnprovenTransactions(shieldedBalancingTx, feeBalancingTransaction);
        // if there is no balancingTransaction and there was no unshielded tx balancing (in place) throw an error.
        if (!balancingTransaction && !balancedUnshieldedTx) {
            throw new Error('No balancing transaction was created. Please check your transaction.');
        }
        return {
            type: 'UNBOUND_TRANSACTION',
            baseTransaction: baseTx,
            balancingTransaction: balancingTransaction ?? undefined,
        };
    }
    async balanceUnprovenTransaction(tx, secretKeys, options) {
        const { shieldedSecretKeys, dustSecretKey } = secretKeys;
        const { ttl, tokenKindsToBalance = 'all' } = options;
        const { shouldBalanceDust, shouldBalanceShielded, shouldBalanceUnshielded } = TokenKindsToBalance.toFlags(tokenKindsToBalance);
        // Step 1: Run unshielded and shielded balancing
        const shieldedBalancingTx = shouldBalanceShielded
            ? await this.shielded.balanceTransaction(shieldedSecretKeys, tx)
            : undefined;
        // For unproven transactions, unshielded balancing happens in place
        const balancedUnshieldedTx = shouldBalanceUnshielded
            ? await this.unshielded.balanceUnprovenTransaction(tx)
            : undefined;
        // Step 2: Use the balanced unshielded tx if present, otherwise use the original tx
        const baseTx = balancedUnshieldedTx ?? tx;
        // Step 3: Merge shielded balancing into base tx if present
        const mergedTx = this.mergeUnprovenTransactions(baseTx, shieldedBalancingTx);
        // Step 4: Conditionally add dust/fee balancing
        const feeBalancingTx = shouldBalanceDust
            ? await this.dust.balanceTransactions(dustSecretKey, [mergedTx], ttl)
            : undefined;
        // Step 5: Merge fee balancing if present
        const balancedTx = this.mergeUnprovenTransactions(mergedTx, feeBalancingTx);
        return {
            type: 'UNPROVEN_TRANSACTION',
            transaction: balancedTx,
        };
    }
    async finalizeRecipe(recipe) {
        return Promise.resolve(recipe)
            .then(async (recipe) => {
            switch (recipe.type) {
                case 'FINALIZED_TRANSACTION': {
                    const finalizedBalancing = await this.finalizeTransaction(recipe.balancingTransaction);
                    return recipe.originalTransaction.merge(finalizedBalancing);
                }
                case 'UNBOUND_TRANSACTION': {
                    const finalizedBalancingTx = recipe.balancingTransaction
                        ? await this.finalizeTransaction(recipe.balancingTransaction)
                        : undefined;
                    const finalizedTransaction = recipe.baseTransaction.bind();
                    return finalizedBalancingTx ? finalizedTransaction.merge(finalizedBalancingTx) : finalizedTransaction;
                }
                case 'UNPROVEN_TRANSACTION': {
                    return await this.finalizeTransaction(recipe.transaction);
                }
            }
        })
            .then(async (finalizedTx) => {
            await this.pendingTransactionsService.addPendingTransaction(finalizedTx);
            return finalizedTx;
        });
    }
    async signRecipe(recipe, signSegment) {
        switch (recipe.type) {
            case 'FINALIZED_TRANSACTION': {
                const signedBalancingTx = await this.signUnprovenTransaction(recipe.balancingTransaction, signSegment);
                return {
                    type: 'FINALIZED_TRANSACTION',
                    originalTransaction: recipe.originalTransaction,
                    balancingTransaction: signedBalancingTx,
                };
            }
            case 'UNBOUND_TRANSACTION': {
                const signedBalancingTx = recipe.balancingTransaction
                    ? await this.signUnprovenTransaction(recipe.balancingTransaction, signSegment)
                    : undefined;
                const signedBaseTx = await this.signUnboundTransaction(recipe.baseTransaction, signSegment);
                return {
                    type: 'UNBOUND_TRANSACTION',
                    baseTransaction: signedBaseTx,
                    balancingTransaction: signedBalancingTx,
                };
            }
            case 'UNPROVEN_TRANSACTION': {
                const signedTx = await this.signUnprovenTransaction(recipe.transaction, signSegment);
                return {
                    type: 'UNPROVEN_TRANSACTION',
                    transaction: signedTx,
                };
            }
        }
    }
    async signUnprovenTransaction(tx, signSegment) {
        return await this.unshielded.signUnprovenTransaction(tx, signSegment);
    }
    async signUnboundTransaction(tx, signSegment) {
        return await this.unshielded.signUnboundTransaction(tx, signSegment);
    }
    async finalizeTransaction(tx) {
        try {
            const unboundTx = await this.provingService.prove(tx);
            const finalizedTx = unboundTx.bind();
            await this.pendingTransactionsService.addPendingTransaction(finalizedTx);
            return finalizedTx;
        }
        catch (error) {
            await Promise.allSettled([
                this.shielded.revertTransaction(tx),
                this.unshielded.revertTransaction(tx),
                this.dust.revertTransaction(tx),
            ]);
            throw error;
        }
    }
    /** Estimates the fee for the given transaction only. This lacks the fees of the balancing transaction. */
    async calculateTransactionFee(tx) {
        return await this.dust.calculateFee([tx]);
    }
    /** Calculates the total fee for the given transaction plus the fee of the balancing transaction. */
    async estimateTransactionFee(tx, secretKey, options) {
        const ttl = options?.ttl ?? this.defaultTtl();
        return await this.dust.estimateFee(secretKey, [tx], ttl, options?.currentTime);
    }
    async transferTransaction(outputs, secretKeys, options) {
        const { shieldedSecretKeys, dustSecretKey } = secretKeys;
        const { ttl, payFees = true } = options;
        const unshieldedOutputs = outputs
            .filter((output) => output.type === 'unshielded')
            .flatMap((output) => output.outputs);
        const shieldedOutputs = outputs.filter((output) => output.type === 'shielded').flatMap((output) => output.outputs);
        if (unshieldedOutputs.length === 0 && shieldedOutputs.length === 0) {
            throw Error('At least one shielded or unshielded output is required.');
        }
        const shieldedTx = shieldedOutputs.length > 0
            ? await this.shielded.transferTransaction(shieldedSecretKeys, shieldedOutputs)
            : undefined;
        const unshieldedTx = unshieldedOutputs.length > 0 ? await this.unshielded.transferTransaction(unshieldedOutputs, ttl) : undefined;
        const mergedTxs = this.mergeUnprovenTransactions(shieldedTx, unshieldedTx);
        // Add fee payment
        const feeBalancingTx = payFees ? await this.dust.balanceTransactions(dustSecretKey, [mergedTxs], ttl) : undefined;
        const finalTx = this.mergeUnprovenTransactions(mergedTxs, feeBalancingTx);
        return {
            type: 'UNPROVEN_TRANSACTION',
            transaction: finalTx,
        };
    }
    /**
     * Provides estimate of the fee of issuing registration transaction with provided UTxOs
     * @param nightUtxos - Night UTxOs to use for the registration
     * @returns And object informing about fee at the moment, as well as estimation of dust generation of the UTxO(s), that would be used for paying the fee. These include data that allows to compute when the fee could be paid
     */
    async estimateRegistration(nightUtxos) {
        const now = new Date();
        const dustState = await this.dust.waitForSyncedState();
        const dustGenerationEstimations = pipe(nightUtxos, Arr.map(({ utxo, meta }) => ({
            ...utxo,
            ctime: meta.ctime,
            registeredForDustGeneration: meta.registeredForDustGeneration,
        })), (utxosWithMeta) => dustState.estimateDustGeneration(utxosWithMeta, now), (estimatedUtxos) => dustState.capabilities.coinsAndBalances.splitNightUtxos(estimatedUtxos), (split) => split.guaranteed);
        const fakeSigningKey = ledger.sampleSigningKey();
        const fakeVerifyingKey = ledger.signatureVerifyingKey(fakeSigningKey);
        const fakeRegistrationRecipe = await this.registerNightUtxosForDustGeneration(nightUtxos, fakeVerifyingKey, (payload) => ledger.signData(fakeSigningKey, payload), dustState.address);
        const finalizedFakeTx = fakeRegistrationRecipe.transaction.mockProve().bind();
        const fee = await this.calculateTransactionFee(finalizedFakeTx);
        return {
            fee,
            dustGenerationEstimations,
        };
    }
    async initSwap(desiredInputs, desiredOutputs, secretKeys, options) {
        const { shieldedSecretKeys, dustSecretKey } = secretKeys;
        const { ttl, payFees = false } = options;
        const { shielded: shieldedInputs, unshielded: unshieldedInputs } = desiredInputs;
        const shieldedOutputs = desiredOutputs
            .filter((output) => output.type === 'shielded')
            .flatMap((output) => output.outputs);
        const unshieldedOutputs = desiredOutputs
            .filter((output) => output.type === 'unshielded')
            .flatMap((output) => output.outputs);
        const hasShieldedPart = (shieldedInputs && Object.keys(shieldedInputs).length > 0) || shieldedOutputs.length > 0;
        const hasUnshieldedPart = (unshieldedInputs && Object.keys(unshieldedInputs).length > 0) || unshieldedOutputs.length > 0;
        if (!hasShieldedPart && !hasUnshieldedPart) {
            throw Error('At least one shielded or unshielded swap is required.');
        }
        const shieldedTx = hasShieldedPart && shieldedInputs !== undefined
            ? await this.shielded.initSwap(shieldedSecretKeys, shieldedInputs, shieldedOutputs)
            : undefined;
        const unshieldedTx = hasUnshieldedPart && unshieldedInputs !== undefined
            ? await this.unshielded.initSwap(unshieldedInputs, unshieldedOutputs, ttl)
            : undefined;
        const combinedTx = this.mergeUnprovenTransactions(shieldedTx, unshieldedTx);
        if (!combinedTx) {
            throw Error('Unexpected transaction state.');
        }
        const feeBalancingTx = payFees ? await this.dust.balanceTransactions(dustSecretKey, [combinedTx], ttl) : undefined;
        const finalTx = this.mergeUnprovenTransactions(combinedTx, feeBalancingTx);
        return {
            type: 'UNPROVEN_TRANSACTION',
            transaction: finalTx,
        };
    }
    async registerNightUtxosForDustGeneration(nightUtxos, nightVerifyingKey, signDustRegistration, dustReceiverAddress) {
        if (nightUtxos.length === 0) {
            throw Error('At least one Night UTXO is required.');
        }
        const receiverAddress = dustReceiverAddress ?? (await this.dust.getAddress());
        const dustRegistrationTx = await this.createDustActionTransaction({ type: 'registration', dustReceiverAddress: receiverAddress }, nightUtxos, nightVerifyingKey, signDustRegistration);
        return {
            type: 'UNPROVEN_TRANSACTION',
            transaction: dustRegistrationTx,
        };
    }
    async deregisterFromDustGeneration(nightUtxos, nightVerifyingKey, signDustRegistration) {
        const dustDeregistrationTx = await this.createDustActionTransaction({ type: 'deregistration' }, nightUtxos, nightVerifyingKey, signDustRegistration);
        return {
            type: 'UNPROVEN_TRANSACTION',
            transaction: dustDeregistrationTx,
        };
    }
    async revert(txOrRecipe) {
        // avoid instanceof check
        const transactionsToRevert = BalancingRecipe.isRecipe(txOrRecipe)
            ? BalancingRecipe.getTransactions(txOrRecipe)
            : [txOrRecipe];
        await Promise.all(transactionsToRevert.map((tx) => this.revertTransaction(tx)));
    }
    async revertTransaction(tx) {
        await Promise.all([
            this.shielded.revertTransaction(tx),
            this.unshielded.revertTransaction(tx),
            this.dust.revertTransaction(tx),
        ]).then(() => this.pendingTransactionsService.clear(tx));
    }
    async start(shieldedSecretKeys, dustSecretKey) {
        await Promise.all([
            this.shielded.start(shieldedSecretKeys),
            this.unshielded.start(),
            this.dust.start(dustSecretKey),
            this.pendingTransactionsService.start(),
        ]);
    }
    async stop() {
        await Promise.all([
            this.shielded.stop(),
            this.unshielded.stop(),
            this.dust.stop(),
            this.submissionService.close(),
            this.pendingTransactionsService.stop(),
            Promise.resolve(this.#pendingSubscription?.unsubscribe()),
        ]);
    }
}
